{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "191510b4-5c34-448d-ab60-e1beb0ca649a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Before we start, remember to activate your conda environment.\n",
    "## Write 'conda create --name myenv numpy pandas scipy matplotlib' into your terminal\n",
    "## Then activate the environment: 'conda activate myenv'\n",
    "## Remember to check that your Python interpreter is correct\n",
    "\n",
    "# First step: Import the libraries we're going to use (PANDAS, NUMPY, MATPLOTLIB)\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib as plt\n",
    "import os "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f1e8d419-84f5-41e0-980e-26e9a9850130",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DOI</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SNP ID</th>\n",
       "      <th>P-VALUE (GWAS)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [DOI, GENE, SNP ID, P-VALUE (GWAS)]\n",
       "Index: []"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Second step: We're going to read our Resources.csv and test to see if we can see it.\n",
    "SNP_file = pd.read_csv('Resources/Template.csv')    #You should always open the Data_Extraction_Code \n",
    "                                                       #folder to have a relative path similar to this one\n",
    "SNP_file.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7cbf96bf-ed42-4524-a648-a2069bb6b777",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DOI</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SNP ID</th>\n",
       "      <th>P-VALUE (GWAS)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [DOI, GENE, SNP ID, P-VALUE (GWAS)]\n",
       "Index: []"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Third step: Now let's create a second dataframe (to keep from editing the original)\n",
    "SNP_editable = SNP_file\n",
    "SNP_editable.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f8e5a280-4c04-4991-8c73-de47397fbea7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DOI</th>\n",
       "      <th>STUDY</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SNP ID</th>\n",
       "      <th>HGVS NOMENCLATURE</th>\n",
       "      <th>PROTECTIVE EFFECT (0=YES, 1=NO)</th>\n",
       "      <th>MAF</th>\n",
       "      <th>ALLELE</th>\n",
       "      <th>P-VALUE (GWAS)</th>\n",
       "      <th>SAMPLE SIZE</th>\n",
       "      <th>POPULATION / ETHNICITY</th>\n",
       "      <th>TYPE OF ANEURYSM\\n(0=UNDEFINED, 1=SACCULAR, 2=FUSIFORM, 3=MULTIPLE ANEURYSMS (UNDEFINED), 4= MULTIPLE ANEURYSMS (SACCULAR), 5=MULTIPLE ANEURYSMS (FUSIFORM), 6=MULTIPLE ANEURYSMS (MIXED)</th>\n",
       "      <th>ANEURYSM SITE</th>\n",
       "      <th>NOTES</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/37845353</td>\n",
       "      <td>CDKN2B-AS1</td>\n",
       "      <td>rs4977574</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.000000e-114</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Abdominal Aorta</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/37845353</td>\n",
       "      <td>LPA</td>\n",
       "      <td>rs10455872</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.000000e-55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Abdominal Aorta</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/37845353</td>\n",
       "      <td>ZNF335</td>\n",
       "      <td>rs3827066</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000e-48</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Abdominal Aorta</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/37845353</td>\n",
       "      <td>CELSR2</td>\n",
       "      <td>rs12740374</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.000000e-42</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Abdominal Aorta</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/37845353</td>\n",
       "      <td>LINC00540 - FTH1P7</td>\n",
       "      <td>rs7994761</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000e-39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Abdominal Aorta</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   DOI                                 STUDY                GENE      SNP ID  \\\n",
       "0  NaN  www.ncbi.nlm.nih.gov/pubmed/37845353          CDKN2B-AS1   rs4977574   \n",
       "1  NaN  www.ncbi.nlm.nih.gov/pubmed/37845353                 LPA  rs10455872   \n",
       "2  NaN  www.ncbi.nlm.nih.gov/pubmed/37845353              ZNF335   rs3827066   \n",
       "3  NaN  www.ncbi.nlm.nih.gov/pubmed/37845353              CELSR2  rs12740374   \n",
       "4  NaN  www.ncbi.nlm.nih.gov/pubmed/37845353  LINC00540 - FTH1P7   rs7994761   \n",
       "\n",
       "   HGVS NOMENCLATURE  PROTECTIVE EFFECT (0=YES, 1=NO)  MAF  ALLELE  \\\n",
       "0                NaN                              NaN  NaN     NaN   \n",
       "1                NaN                              NaN  NaN     NaN   \n",
       "2                NaN                              NaN  NaN     NaN   \n",
       "3                NaN                              NaN  NaN     NaN   \n",
       "4                NaN                              NaN  NaN     NaN   \n",
       "\n",
       "   P-VALUE (GWAS)  SAMPLE SIZE  POPULATION / ETHNICITY  \\\n",
       "0   2.000000e-114          NaN                     NaN   \n",
       "1    2.000000e-55          NaN                     NaN   \n",
       "2    1.000000e-48          NaN                     NaN   \n",
       "3    2.000000e-42          NaN                     NaN   \n",
       "4    1.000000e-39          NaN                     NaN   \n",
       "\n",
       "   TYPE OF ANEURYSM\\n(0=UNDEFINED, 1=SACCULAR, 2=FUSIFORM, 3=MULTIPLE ANEURYSMS (UNDEFINED), 4= MULTIPLE ANEURYSMS (SACCULAR), 5=MULTIPLE ANEURYSMS (FUSIFORM), 6=MULTIPLE ANEURYSMS (MIXED)  \\\n",
       "0                                                NaN                                                                                                                                           \n",
       "1                                                NaN                                                                                                                                           \n",
       "2                                                NaN                                                                                                                                           \n",
       "3                                                NaN                                                                                                                                           \n",
       "4                                                NaN                                                                                                                                           \n",
       "\n",
       "     ANEURYSM SITE  NOTES  \n",
       "0  Abdominal Aorta    NaN  \n",
       "1  Abdominal Aorta    NaN  \n",
       "2  Abdominal Aorta    NaN  \n",
       "3  Abdominal Aorta    NaN  \n",
       "4  Abdominal Aorta    NaN  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Fourth step: Let's merge the dataframe and use SNP ID as our merge.\n",
    "# Let's read our second .csv file\n",
    "Our_SNP_List = pd.read_csv('Resources/Input_Files/AA_SNP_Extraction.csv')\n",
    "Our_SNP_List.head()\n",
    "\n",
    "# And..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "551e8452-ecf7-4d5f-9b8f-db04c2b1db8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MERGE! \n",
    "merge_data = pd.merge(Our_SNP_List, SNP_file, on=[\"SNP ID\", \"GENE\", \"DOI\", \"P-VALUE (GWAS)\"], how='left')\n",
    "merge_data.head(10)\n",
    "\n",
    "# Now we have our own database set up as a dataframe and filled it with our columns. \n",
    "# If you merged two databases, great! If you did not, boo! Let's output the merged_data (if you did not merge two databases,\n",
    "# then ignore the output.\n",
    "output_path = 'Resources/OutputFromExtractionCode'\n",
    "merge_data.to_csv(os.path.join(output_path, 'Merged_data_1.csv'), index=False)\n",
    "\n",
    "\n",
    "#merge_data.to_csv('Merged_data_1.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "dd82b8c2-e8c2-4e49-902e-0409cd30961d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>SNP ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDKN2B-AS1</td>\n",
       "      <td>rs4977574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LPA</td>\n",
       "      <td>rs10455872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZNF335</td>\n",
       "      <td>rs3827066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CELSR2</td>\n",
       "      <td>rs12740374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LINC00540 - FTH1P7</td>\n",
       "      <td>rs7994761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>APOE</td>\n",
       "      <td>rs429358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SMARCA4</td>\n",
       "      <td>rs73015011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ADAMTS8</td>\n",
       "      <td>rs4936098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CHRNA5</td>\n",
       "      <td>rs17486278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CDKN2B-AS1</td>\n",
       "      <td>rs2891168</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GENE      SNP ID\n",
       "0          CDKN2B-AS1   rs4977574\n",
       "1                 LPA  rs10455872\n",
       "2              ZNF335   rs3827066\n",
       "3              CELSR2  rs12740374\n",
       "4  LINC00540 - FTH1P7   rs7994761\n",
       "5                APOE    rs429358\n",
       "6             SMARCA4  rs73015011\n",
       "7             ADAMTS8   rs4936098\n",
       "8              CHRNA5  rs17486278\n",
       "9          CDKN2B-AS1   rs2891168"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Now let's drop all the data we collected and just stay with the two things that matter most to us: \n",
    "# the SNP ID and the Gene:\n",
    "SNP_ID_And_Gene_From_Merged_Data = merge_data[['GENE', 'SNP ID']]\n",
    "SNP_ID_And_Gene_From_Merged_Data.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cfd151fe-b8d4-468b-a7b2-6e8a1a9687c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "### We'll be using this for our RENTREZ module in R'. Let's output it twice (one for keeps and one we can modify safely)\n",
    "SNP_ID_And_Gene_From_Merged_Data.to_csv(os.path.join(output_path, 'Original_Gene_And_dbSNP_ID_Merged.csv'), index=False)\n",
    "SNP_ID_And_Gene_From_Merged_Data.to_csv(os.path.join(output_path, 'Editable_Gene_And_dbSNP_ID_Merged.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3f1d9c91-ad09-4fc8-9a3f-e61027e45130",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>SNP ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>ABCA6</td>\n",
       "      <td>rs117753190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>ADAM10</td>\n",
       "      <td>rs11856657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>ADAMTS8</td>\n",
       "      <td>rs4936098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>ADAMTS8</td>\n",
       "      <td>rs7936928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ADAMTS8</td>\n",
       "      <td>rs4936098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>ADH1C - ADH7</td>\n",
       "      <td>rs1229849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>ANKRD44-IT1, ANKRD44</td>\n",
       "      <td>rs919433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>ANTXR1</td>\n",
       "      <td>rs13384676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>APOA5 - LNC-RHL1</td>\n",
       "      <td>rs662799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>APOE</td>\n",
       "      <td>rs429358</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     GENE       SNP ID\n",
       "119                 ABCA6  rs117753190\n",
       "61                 ADAM10   rs11856657\n",
       "40                ADAMTS8    rs4936098\n",
       "142               ADAMTS8    rs7936928\n",
       "7                 ADAMTS8    rs4936098\n",
       "147          ADH1C - ADH7    rs1229849\n",
       "149  ANKRD44-IT1, ANKRD44     rs919433\n",
       "114                ANTXR1   rs13384676\n",
       "118      APOA5 - LNC-RHL1     rs662799\n",
       "44                   APOE     rs429358"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### We can already do something pretty cool, which is determine which gene had more pathogenic SNPs.\n",
    "### We can sort it:\n",
    "SNP_Sorted_By_Gene =  SNP_ID_And_Gene_From_Merged_Data.sort_values(by='GENE')\n",
    "SNP_Sorted_By_Gene.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8d3a0793-bc82-49fd-ad95-c142054d2c3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>Reported SNPs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDKN2B-AS1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LPA</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DAB2IP</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LINC00540 - FTH1P7</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ADAMTS8</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ERG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>RPS4XP18 - RNU6-1032P</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>IL6R</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GDF7</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>LDLR</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>LRP1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>LINC02775 - SMYD2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CRISPLD2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>LPL</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>PCSK9</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SUGCT</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>TRIB1AL</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>LIPA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ZPR1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CELSR2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     GENE  Reported SNPs\n",
       "0              CDKN2B-AS1              4\n",
       "1                     LPA              4\n",
       "2                  DAB2IP              3\n",
       "3      LINC00540 - FTH1P7              3\n",
       "4                 ADAMTS8              3\n",
       "5                     ERG              3\n",
       "6   RPS4XP18 - RNU6-1032P              3\n",
       "7                    IL6R              3\n",
       "8                    GDF7              2\n",
       "9                    LDLR              2\n",
       "10                   LRP1              2\n",
       "11      LINC02775 - SMYD2              2\n",
       "12               CRISPLD2              2\n",
       "13                    LPL              2\n",
       "14                  PCSK9              2\n",
       "15                  SUGCT              2\n",
       "16                TRIB1AL              2\n",
       "17                   LIPA              2\n",
       "18                   ZPR1              2\n",
       "19                 CELSR2              2"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### And get the unique count for each gene:\n",
    "Unique_Genes = SNP_Sorted_By_Gene['GENE'].value_counts().reset_index()\n",
    "Unique_Genes.columns = ['GENE', 'Reported SNPs']\n",
    "Unique_Genes.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "493db22e-ff98-465f-b2ef-966c4fa59d04",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>SNP ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCA6</td>\n",
       "      <td>[rs117753190]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADAM10</td>\n",
       "      <td>[rs11856657]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADAMTS8</td>\n",
       "      <td>[rs4936098, rs7936928, rs4936098]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ADH1C - ADH7</td>\n",
       "      <td>[rs1229849]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ANKRD44-IT1, ANKRD44</td>\n",
       "      <td>[rs919433]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ANTXR1</td>\n",
       "      <td>[rs13384676]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>APOA5 - LNC-RHL1</td>\n",
       "      <td>[rs662799]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>APOE</td>\n",
       "      <td>[rs429358, rs429358]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ATOH8</td>\n",
       "      <td>[rs113626898]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ATP7BP1 - RPS4XP18</td>\n",
       "      <td>[rs7240701]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>C20orf181 - ABHD16B</td>\n",
       "      <td>[rs73149487]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CCDC157</td>\n",
       "      <td>[rs59731356]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CDK6</td>\n",
       "      <td>[rs42235]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CDKN1A</td>\n",
       "      <td>[rs3176336, rs3176336]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CDKN2B-AS1</td>\n",
       "      <td>[rs4977574, rs10757274, rs2891168, rs7866503]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CELSR2</td>\n",
       "      <td>[rs12740374, rs12740374]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CELSR2 - PSRC1</td>\n",
       "      <td>[rs602633]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CFDP1</td>\n",
       "      <td>[rs11149831]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CHRNA3 - CHRNB4</td>\n",
       "      <td>[rs55958997]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CHRNA5</td>\n",
       "      <td>[rs17486278]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    GENE                                         SNP ID\n",
       "0                  ABCA6                                  [rs117753190]\n",
       "1                 ADAM10                                   [rs11856657]\n",
       "2                ADAMTS8              [rs4936098, rs7936928, rs4936098]\n",
       "3           ADH1C - ADH7                                    [rs1229849]\n",
       "4   ANKRD44-IT1, ANKRD44                                     [rs919433]\n",
       "5                 ANTXR1                                   [rs13384676]\n",
       "6       APOA5 - LNC-RHL1                                     [rs662799]\n",
       "7                   APOE                           [rs429358, rs429358]\n",
       "8                  ATOH8                                  [rs113626898]\n",
       "9     ATP7BP1 - RPS4XP18                                    [rs7240701]\n",
       "10   C20orf181 - ABHD16B                                   [rs73149487]\n",
       "11               CCDC157                                   [rs59731356]\n",
       "12                  CDK6                                      [rs42235]\n",
       "13                CDKN1A                         [rs3176336, rs3176336]\n",
       "14            CDKN2B-AS1  [rs4977574, rs10757274, rs2891168, rs7866503]\n",
       "15                CELSR2                       [rs12740374, rs12740374]\n",
       "16        CELSR2 - PSRC1                                     [rs602633]\n",
       "17                 CFDP1                                   [rs11149831]\n",
       "18       CHRNA3 - CHRNB4                                   [rs55958997]\n",
       "19                CHRNA5                                   [rs17486278]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### And we should group them by gene so we can see which SNPs are actually related:\n",
    "grouped_SNPs = SNP_Sorted_By_Gene.groupby('GENE')['SNP ID'].apply(list).reset_index()\n",
    "grouped_SNPs.head(20)\n",
    "\n",
    "# groupby('category') groups the data by the 'category' column.\n",
    "# apply(list) applies the list function to each group, resulting in a list of strings for each unique category.\n",
    "# reset_index() is used to turn the result back into a DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f26c4003-dff5-4b2d-92bb-48e238852392",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Now let's output them as CSV files:\n",
    "SNP_Sorted_By_Gene.to_csv(os.path.join(output_path, 'SNP_Sorted_By_Gene.csv'), index=False)\n",
    "Unique_Genes.to_csv(os.path.join(output_path, 'Unique_Genes.csv'), index=False)\n",
    "grouped_SNPs.to_csv(os.path.join(output_path, 'grouped_SNPs.csv'), index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6033b25-548d-411d-9dbf-8aaa7a7eef34",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Now let's turn to our python R file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c79a7766-782e-426c-a8f8-be7dc2972d10",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
