{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "9d9b8fd4-0498-44fc-9d30-bd878bb7cfa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Now since this is a new file we need to import other modules (YAY!) including our R' emulator\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects.packages import importr\n",
    "from rpy2.robjects import pandas2ri\n",
    "\n",
    "### and let's not forget the usual gang:\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib as plt\n",
    "import re\n",
    "import os "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "9b8c2979-866d-405e-a5f1-fd872a08cdb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Before we get started in an R' environment. We need to talk about something important.\n",
    "#### Rentrez does not use SNP ID to search. It uses UID. As in, they don't use 'rs2000' \n",
    "#### but rather '2000'. So we need to use our previous output and split the function.\n",
    "Output_GeneSNP_file = pd.read_csv('Resources/OutputFromExtractionCode/Editable_Gene_And_dbSNP_ID_Merged.csv')\n",
    "Output_GeneSNP_file.head(10)\n",
    "\n",
    "output_path = 'Resources/OutputFromExtractionCode'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "751ff5a8-be40-41ee-b44d-ece195aa51da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>SNP ID</th>\n",
       "      <th>UID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDKN2B-AS1</td>\n",
       "      <td>rs4977574</td>\n",
       "      <td>4977574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LPA</td>\n",
       "      <td>rs10455872</td>\n",
       "      <td>10455872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZNF335</td>\n",
       "      <td>rs3827066</td>\n",
       "      <td>3827066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CELSR2</td>\n",
       "      <td>rs12740374</td>\n",
       "      <td>12740374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LINC00540 - FTH1P7</td>\n",
       "      <td>rs7994761</td>\n",
       "      <td>7994761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>APOE</td>\n",
       "      <td>rs429358</td>\n",
       "      <td>429358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SMARCA4</td>\n",
       "      <td>rs73015011</td>\n",
       "      <td>73015011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ADAMTS8</td>\n",
       "      <td>rs4936098</td>\n",
       "      <td>4936098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CHRNA5</td>\n",
       "      <td>rs17486278</td>\n",
       "      <td>17486278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CDKN2B-AS1</td>\n",
       "      <td>rs2891168</td>\n",
       "      <td>2891168</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GENE      SNP ID       UID\n",
       "0          CDKN2B-AS1   rs4977574   4977574\n",
       "1                 LPA  rs10455872  10455872\n",
       "2              ZNF335   rs3827066   3827066\n",
       "3              CELSR2  rs12740374  12740374\n",
       "4  LINC00540 - FTH1P7   rs7994761   7994761\n",
       "5                APOE    rs429358    429358\n",
       "6             SMARCA4  rs73015011  73015011\n",
       "7             ADAMTS8   rs4936098   4936098\n",
       "8              CHRNA5  rs17486278  17486278\n",
       "9          CDKN2B-AS1   rs2891168   2891168"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Now we split it. \n",
    "Split_dbSNP_ID = Output_GeneSNP_file\n",
    "Split_dbSNP_ID[['rs', 'UID']] = Output_GeneSNP_file['SNP ID'].str.split('rs', expand = True)\n",
    "Split_dbSNP_ID.head(10)\n",
    "\n",
    "#And delete the rs column:\n",
    "UID_SNP = Split_dbSNP_ID.drop(columns=['rs'])\n",
    "UID_SNP.head(10)\n",
    "\n",
    "## Now we have the UID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "d59ca7d3-ffa1-4e16-b92b-8d440f53957c",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Let's import RENTREZ to our python environment\n",
    "rentrez = importr('rentrez')\n",
    "\n",
    "### You'll need to provide an email address to use Entrez tools. BTW Hot tip: \n",
    "### If you have multiple lines, using triple quotes is cleaner and easier to read:\n",
    "\n",
    "robjects.r('''\n",
    "library(rentrez)\n",
    "entrez_email <- \"kenji.ponsm@gmail.com\"\n",
    "''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "a111e1f2-502d-44da-9872-d5a970fe2d4e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Searchable fields for database 'snp'\n",
      "  ALL \t All terms from all searchable fields \n",
      "  UID \t Unique number assigned to publication \n",
      "  FILT \t Limits the records \n",
      "  RS \t Clustered SNP ID (rs) \n",
      "  CHR \t chromosomes \n",
      "  GENE \t locus link symbol \n",
      "  HAN \t Submitter Handle \n",
      "  ACCN \t nucleotide accessions \n",
      "  GENE_ID \t Gene ID \n",
      "  FXN \t dbSNP Functional consequence class \n",
      "  GTYP \t Genotype info \n",
      "  SS \t Submitter ID \n",
      "  VARI \t Allele \n",
      "  SCLS \t SNP class \n",
      "  CPOS \t Chromosome base position \n",
      "  WORD \t Free text associated with record \n",
      "  SIDX \t SNP Index \n",
      "  CLIN \t Variations with clinical effects or significances \n",
      "  GMAF \t Minor Allele Frequency derived from global population (ie. 1000G) \n",
      "  VALI \t Validation status \n",
      "  CPOS_GRCH37 \t Chromosome base position on previous assembly version \n",
      "  ORGN \t Organism \n",
      "  ALFA_EUR \t ALFA European Minor Allele Frequency \n",
      "  ALFA_AFR \t ALFA African population Minor Allele Frequency \n",
      "  ALFA_ASN \t ALFA Asian population Minor Allele Frequency \n",
      "  ALFA_LAC \t ALFA Latin American 1 population Minor Allele Frequency \n",
      "  ALFA_LEN \t ALFA Latin American 2 population Minor Allele Frequency \n",
      "  ALFA_SAS \t ALFA South Asian population Minor Allele Frequency \n",
      "  ALFA_OTR \t ALFA Other population Minor Allele Frequency \n",
      "\n"
     ]
    }
   ],
   "source": [
    "### What can we look for with entrez\n",
    "snp_database = robjects.r('entrez_db_searchable(\"snp\")')\n",
    "print(snp_database)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "2b86d25f-7e00-496f-90dd-3ba9b8db7d8c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4977574\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>allele_origin</th>\n",
       "      <th>global_mafs</th>\n",
       "      <th>global_population</th>\n",
       "      <th>global_samplesize</th>\n",
       "      <th>suspected</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>genes</th>\n",
       "      <th>acc</th>\n",
       "      <th>...</th>\n",
       "      <th>allele</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>chrpos_prev_assm</th>\n",
       "      <th>text</th>\n",
       "      <th>snp_id_sort</th>\n",
       "      <th>clinical_sort</th>\n",
       "      <th>cited_sort</th>\n",
       "      <th>chrpos_sort</th>\n",
       "      <th>merged_sort</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4977574</td>\n",
       "      <td>4977574</td>\n",
       "      <td></td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>risk-factor</td>\n",
       "      <td>[1] \"CDKN2B-AS1\"\\n</td>\n",
       "      <td>NC_000009.12</td>\n",
       "      <td>...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>9:22098575</td>\n",
       "      <td>9:22098574</td>\n",
       "      <td></td>\n",
       "      <td>0004977574</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>0022098575</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       uid   snp_id allele_origin  \\\n",
       "0  4977574  4977574                 \n",
       "\n",
       "                                         global_mafs global_population  \\\n",
       "0   [1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...                     \n",
       "\n",
       "  global_samplesize suspected clinical_significance               genes  \\\n",
       "0                                       risk-factor  [1] \"CDKN2B-AS1\"\\n   \n",
       "\n",
       "            acc  ... allele snp_class      chrpos chrpos_prev_assm text  \\\n",
       "0  NC_000009.12  ...      D       snv  9:22098575       9:22098574        \n",
       "\n",
       "  snp_id_sort clinical_sort cited_sort chrpos_sort merged_sort  \n",
       "0  0004977574             1             0022098575           0  \n",
       "\n",
       "[1 rows x 32 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Let's give it a test shall we? Let's say I want the information of rs429358 (GENE: APOE)\n",
    "print(UID_SNP.iloc[0,2])\n",
    "test_snp = UID_SNP.iloc[0,2]\n",
    "search_test = rentrez.entrez_summary(db='snp', id=test_snp)\n",
    "search_test_dict = {key: str(search_test.rx2(key) [0]) for key in search_test.names}\n",
    "\n",
    "search_test_df = pd.DataFrame([search_test_dict])\n",
    "\n",
    "search_test_df.head(10)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "42a5e8ad-01e1-4263-ac18-3c39726932e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [1] \"uid\"                   \"snp_id\"                \"allele_origin\"        \n",
      " [4] \"global_mafs\"           \"global_population\"     \"global_samplesize\"    \n",
      " [7] \"suspected\"             \"clinical_significance\" \"genes\"                \n",
      "[10] \"acc\"                   \"chr\"                   \"handle\"               \n",
      "[13] \"spdi\"                  \"fxn_class\"             \"validated\"            \n",
      "[16] \"docsum\"                \"tax_id\"                \"orig_build\"           \n",
      "[19] \"upd_build\"             \"createdate\"            \"updatedate\"           \n",
      "[22] \"ss\"                    \"allele\"                \"snp_class\"            \n",
      "[25] \"chrpos\"                \"chrpos_prev_assm\"      \"text\"                 \n",
      "[28] \"snp_id_sort\"           \"clinical_sort\"         \"cited_sort\"           \n",
      "[31] \"chrpos_sort\"           \"merged_sort\"          \n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(search_test.names)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "c48732cd-c951-44bb-9104-227be736a212",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>global_mafs</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>genes</th>\n",
       "      <th>acc</th>\n",
       "      <th>chr</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>allele</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>chrpos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4977574</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...</td>\n",
       "      <td>risk-factor</td>\n",
       "      <td>[1] \"CDKN2B-AS1\"\\n</td>\n",
       "      <td>NC_000009.12</td>\n",
       "      <td>9</td>\n",
       "      <td>genic_downstream_transcript_variant,intron_var...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>9:22098575</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       uid                                        global_mafs  \\\n",
       "0  4977574   [1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...   \n",
       "\n",
       "  clinical_significance               genes           acc chr  \\\n",
       "0           risk-factor  [1] \"CDKN2B-AS1\"\\n  NC_000009.12   9   \n",
       "\n",
       "                                           fxn_class allele snp_class  \\\n",
       "0  genic_downstream_transcript_variant,intron_var...      D       snv   \n",
       "\n",
       "       chrpos  \n",
       "0  9:22098575  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### And let's filter for the actual values we want to extract: \n",
    "desired_dbSNP_values = (['genes', 'uid', 'snp_class', \n",
    "                         'fxn_class', 'chr', 'chrpos', 'global_mafs', \n",
    "                         'allele','clinical_significance', 'acc'] \n",
    "                         )\n",
    "filtered_search_test_dict = {key: str(search_test.rx2(key)[0]) for key in search_test.names if key in desired_dbSNP_values}\n",
    "filtered_search_test = pd.DataFrame([filtered_search_test_dict])\n",
    "filtered_search_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "fc3d7c7d-291c-4ee1-b728-9fd3dc5c9971",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>uid</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>chr</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>global_mafs</th>\n",
       "      <th>allele</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>acc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[1] \"CDKN2B-AS1\"\\n</td>\n",
       "      <td>4977574</td>\n",
       "      <td>snv</td>\n",
       "      <td>genic_downstream_transcript_variant,intron_var...</td>\n",
       "      <td>9</td>\n",
       "      <td>9:22098575</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...</td>\n",
       "      <td>D</td>\n",
       "      <td>risk-factor</td>\n",
       "      <td>NC_000009.12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                genes      uid snp_class  \\\n",
       "0  [1] \"CDKN2B-AS1\"\\n  4977574       snv   \n",
       "\n",
       "                                           fxn_class chr      chrpos  \\\n",
       "0  genic_downstream_transcript_variant,intron_var...   9  9:22098575   \n",
       "\n",
       "                                         global_mafs allele  \\\n",
       "0   [1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...      D   \n",
       "\n",
       "  clinical_significance           acc  \n",
       "0           risk-factor  NC_000009.12  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_and_ordered_search_test = filtered_search_test[[col for col in desired_dbSNP_values if col in filtered_search_test.columns]]\n",
    "filtered_and_ordered_search_test\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9827951a-19a9-4922-b977-d9bf27b7921c",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_snp_data = []\n",
    "\n",
    "# Open a text file in write mode to log skipped rows\n",
    "with open(\"skipped_rows_log.txt\", \"w\") as log_file:\n",
    "    log_file.write(\"Details of rows skipped due to None UID:\\n\\n\")\n",
    "\n",
    "    for index, row in UID_SNP.iterrows():\n",
    "        test_snp = row['UID']\n",
    "        \n",
    "        # Check if the UID is not None\n",
    "        if test_snp is None:\n",
    "            # Print to console and write to log file\n",
    "            log_message = f\"Skipping row {index} because UID is None. Row details:\\n{row}\\n\\n\"\n",
    "            print(log_message)\n",
    "            log_file.write(log_message)\n",
    "            continue  # Skip this iteration if UID is None\n",
    "\n",
    "        try: \n",
    "            # Retrieve SNP data from the SNP database\n",
    "            search_test = rentrez.entrez_summary(db='snp', id=test_snp)\n",
    "            \n",
    "            # Filter and safely retrieve desired fields\n",
    "            filtered_search_test_dict = {\n",
    "                key: str(search_test.rx2(key)[0]) if len(search_test.rx2(key)) > 0 else None\n",
    "                for key in search_test.names if key in desired_dbSNP_values\n",
    "            }\n",
    "            \n",
    "            # Order the data to match the desired column order\n",
    "            ordered_data = {key: filtered_search_test_dict.get(key, None) for key in desired_dbSNP_values}\n",
    "            \n",
    "            # Append ordered data to the results list\n",
    "            all_snp_data.append(ordered_data)\n",
    "\n",
    "        except Exception as e:\n",
    "            error_message = f\"Error retrieving data for UID {test_snp}: {e}\\n\"\n",
    "            print(error_message)\n",
    "            log_file.write(error_message)\n",
    "\n",
    "# Convert results to DataFrame\n",
    "final_df = pd.DataFrame(all_snp_data)\n",
    "\n",
    "# Display the first 50 rows\n",
    "final_df.head(50)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6520708-7ff6-44ce-8cf3-55a4169a0a38",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Now let's make a copy of our RAW results:\n",
    "final_df.to_csv(os.path.join(output_path, 'compiled_SNPs_RENTREZ_RAW.csv'), index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "281f19ee-bb6b-49ab-9700-8092436a009b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### And let's make it beautiful by rearranging and sorting each column:\n",
    "#### First, let's delete all variants that are not SNVs:\n",
    "filtered_final_df = final_df.loc[final_df['snp_class'] == 'snv'] \n",
    "filtered_final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e3773e8-af4e-4ae4-b3a0-61bc87730794",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Now let's clean up la mugre of each cell in the genes column\n",
    "\n",
    "# Define a function to clean up each cell\n",
    "def clean_cell(value):\n",
    "    if isinstance(value, str):\n",
    "        # Remove `[1]`, quotation marks, and newline characters\n",
    "        cleaned_value = re.sub(r'^\\[\\d+\\]\\s*\"', '', value).replace('\"', '').strip()\n",
    "        return cleaned_value\n",
    "    return value\n",
    "\n",
    "# Apply the cleaning function to the 'genes' column using .loc\n",
    "filtered_final_df.loc[:, 'genes'] = filtered_final_df['genes'].apply(clean_cell)\n",
    "\n",
    "filtered_final_df.head(200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91e34fd6-15e9-4f15-ad5e-2f8c15b1e9e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a copy to avoid the SettingWithCopyWarning\n",
    "filtered_final_df = filtered_final_df.copy()\n",
    "\n",
    "# Rename columns\n",
    "filtered_final_df.rename(columns={'genes': 'Gene ID (GENE_ID)', 'uid': 'dbSNP ID (RS)',\n",
    "                                  'snp_class': 'SNP Class (SCLS)', 'fxn_class': 'Functional consequence class (FXN)', \n",
    "                                  'chr': 'Chromosome', 'chrpos': 'Chromosome base position (CPOS)', \n",
    "                                  'global_mafs': 'Minor Allele Frequency from global population (GMAF)',\n",
    "                                  'clinical_significance': 'Associated clinical phenotypes (CLIN)',\n",
    "                                  'acc': 'Nucleotide Accessions (ACCN)', 'allele': 'Allele'                                 \n",
    "                                 }, inplace=True)\n",
    "\n",
    "# IUPAC mapping dictionary for ambiguity codes\n",
    "iupac_mapping = {\n",
    "    \"R\": \"A/G\", \"Y\": \"C/T\", \"S\": \"G/C\", \"W\": \"A/T\",\n",
    "    \"K\": \"G/T\", \"M\": \"A/C\", \"B\": \"C/G/T\", \"D\": \"A/G/T\",\n",
    "    \"H\": \"A/C/T\", \"V\": \"A/C/G\", \"N\": \"A/T/C/G\"\n",
    "}\n",
    "\n",
    "# Define a function to translate IUPAC codes in the 'allele' column\n",
    "def translate_allele(allele):\n",
    "    if isinstance(allele, str):  # Check if the value is a string\n",
    "        # Replace each IUPAC code with its corresponding alleles\n",
    "        return \"/\".join([iupac_mapping.get(a, a) for a in allele])\n",
    "    return allele\n",
    "\n",
    "# Assume filtered_final_df is your DataFrame and 'allele' is the column with IUPAC codes\n",
    "# Apply the translation function to the 'allele' column\n",
    "filtered_final_df['Allele'] = filtered_final_df['Allele'].apply(translate_allele)\n",
    "\n",
    "# Display the updated DataFrame\n",
    "filtered_final_df.head(5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "6498f193-b198-486d-a815-a39041c0796c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene Symbol: CDKN2B-AS1, Ensembl Gene ID: ENSG00000240498, Gene Type: lncRNA\n",
      "Gene Symbol: LPA, Ensembl Gene ID: ENSG00000198670, Gene Type: protein_coding\n",
      "Gene Symbol: ZNF335, Ensembl Gene ID: ENSG00000198026, Gene Type: protein_coding\n",
      "Gene Symbol: CELSR2, Ensembl Gene ID: ENSG00000143126, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: APOE, Ensembl Gene ID: ENSG00000130203, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: ADAMTS8, Ensembl Gene ID: ENSG00000134917, Gene Type: protein_coding\n",
      "Gene Symbol: CHRNA5, Ensembl Gene ID: ENSG00000169684, Gene Type: protein_coding\n",
      "Gene Symbol: CDKN2B-AS1, Ensembl Gene ID: ENSG00000240498, Gene Type: lncRNA\n",
      "Gene Symbol: LPA, Ensembl Gene ID: ENSG00000198670, Gene Type: protein_coding\n",
      "Gene Symbol: CDKN2B-AS1, Ensembl Gene ID: ENSG00000240498, Gene Type: lncRNA\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: ZPR1, Ensembl Gene ID: ENSG00000109917, Gene Type: protein_coding\n",
      "Gene Symbol: DAB2IP, Ensembl Gene ID: ENSG00000136848, Gene Type: protein_coding\n",
      "Gene Symbol: PLCE1, Ensembl Gene ID: ENSG00000138193, Gene Type: protein_coding\n",
      "Gene Symbol: IL6R, Ensembl Gene ID: ENSG00000160712, Gene Type: protein_coding\n",
      "Gene Symbol: LRP1, Ensembl Gene ID: ENSG00000123384, Gene Type: protein_coding\n",
      "Gene Symbol: SUGCT, Ensembl Gene ID: ENSG00000175600, Gene Type: protein_coding\n",
      "Gene Symbol: GDF7, Ensembl Gene ID: ENSG00000143869, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: LPA, Ensembl Gene ID: ENSG00000198670, Gene Type: protein_coding\n",
      "Gene Symbol: ERG, Ensembl Gene ID: ENSG00000157554, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: ZPR1, Ensembl Gene ID: ENSG00000109917, Gene Type: protein_coding\n",
      "Gene Symbol: LPL, Ensembl Gene ID: ENSG00000175445, Gene Type: protein_coding\n",
      "Gene Symbol: LIPA, Ensembl Gene ID: ENSG00000107798, Gene Type: protein_coding\n",
      "Gene Symbol: CELSR2, Ensembl Gene ID: ENSG00000143126, Gene Type: protein_coding\n",
      "Gene Symbol: LPA, Ensembl Gene ID: ENSG00000198670, Gene Type: protein_coding\n",
      "Gene Symbol: PLCE1, Ensembl Gene ID: ENSG00000138193, Gene Type: protein_coding\n",
      "Gene Symbol: PLAUR, Ensembl Gene ID: ENSG00000011422, Gene Type: protein_coding\n",
      "Gene Symbol: PCIF1, Ensembl Gene ID: ENSG00000100982, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: VMP1, Ensembl Gene ID: ENSG00000062716, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: ADAMTS8, Ensembl Gene ID: ENSG00000134917, Gene Type: protein_coding\n",
      "Gene Symbol: PCSK9, Ensembl Gene ID: ENSG00000169174, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: LTBP4, Ensembl Gene ID: ENSG00000090006, Gene Type: protein_coding\n",
      "Gene Symbol: APOE, Ensembl Gene ID: ENSG00000130203, Gene Type: protein_coding\n",
      "Gene Symbol: SUGCT, Ensembl Gene ID: ENSG00000175600, Gene Type: protein_coding\n",
      "No gene found for symbol 'LOC105378327 LOC124902436' in Ensembl.\n",
      "Gene Symbol: DNM3, Ensembl Gene ID: ENSG00000197959, Gene Type: protein_coding\n",
      "No gene found for symbol 'LOC124904262' in Ensembl.\n",
      "Gene Symbol: LINC02732, Ensembl Gene ID: ENSG00000254416, Gene Type: lncRNA\n",
      "No gene found for symbol 'LDLR     LDLR-AS1' in Ensembl.\n",
      "Gene Symbol: CDK6, Ensembl Gene ID: ENSG00000105810, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: CDKN2B-AS1, Ensembl Gene ID: ENSG00000240498, Gene Type: lncRNA\n",
      "No gene found for symbol 'SF3A2   PLEKHJ1 MIR6789' in Ensembl.\n",
      "Gene Symbol: COL4A2, Ensembl Gene ID: ENSG00000134871, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: LHFPL2, Ensembl Gene ID: ENSG00000145685, Gene Type: protein_coding\n",
      "Gene Symbol: IL6R, Ensembl Gene ID: ENSG00000160712, Gene Type: protein_coding\n",
      "Gene Symbol: ADAM10, Ensembl Gene ID: ENSG00000137845, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: LOXL1, Ensembl Gene ID: ENSG00000129038, Gene Type: protein_coding\n",
      "Gene Symbol: GDF7, Ensembl Gene ID: ENSG00000143869, Gene Type: protein_coding\n",
      "Gene Symbol: VCAN, Ensembl Gene ID: ENSG00000038427, Gene Type: protein_coding\n",
      "Gene Symbol: STN1, Ensembl Gene ID: ENSG00000107960, Gene Type: protein_coding\n",
      "No gene found for symbol 'LOC100506444' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: ZNRF3, Ensembl Gene ID: ENSG00000183579, Gene Type: protein_coding\n",
      "No gene found for symbol 'PLAU     C10orf55' in Ensembl.\n",
      "Gene Symbol: MCF2L, Ensembl Gene ID: ENSG00000126217, Gene Type: protein_coding\n",
      "No gene found for symbol 'TGFB2     TGFB2-OT1' in Ensembl.\n",
      "Gene Symbol: DGKZ, Ensembl Gene ID: ENSG00000149091, Gene Type: protein_coding\n",
      "Gene Symbol: IQCH, Ensembl Gene ID: ENSG00000103599, Gene Type: protein_coding\n",
      "Gene Symbol: PREX1, Ensembl Gene ID: ENSG00000124126, Gene Type: protein_coding\n",
      "No gene found for symbol 'KIAA1656 CCDC157' in Ensembl.\n",
      "Gene Symbol: CRISPLD2, Ensembl Gene ID: ENSG00000103196, Gene Type: protein_coding\n",
      "Gene Symbol: PRKD2, Ensembl Gene ID: ENSG00000105287, Gene Type: protein_coding\n",
      "Gene Symbol: ERG, Ensembl Gene ID: ENSG00000157554, Gene Type: protein_coding\n",
      "Gene Symbol: TGFBR3, Ensembl Gene ID: ENSG00000069702, Gene Type: protein_coding\n",
      "Gene Symbol: DAB2IP, Ensembl Gene ID: ENSG00000136848, Gene Type: protein_coding\n",
      "No gene found for symbol 'LINC02544' in Ensembl.\n",
      "Gene Symbol: PDGFRA, Ensembl Gene ID: ENSG00000134853, Gene Type: protein_coding\n",
      "No gene found for symbol 'ACYP1   ZC2HC1C' in Ensembl.\n",
      "Gene Symbol: LDB2, Ensembl Gene ID: ENSG00000169744, Gene Type: protein_coding\n",
      "Gene Symbol: SCARB1, Ensembl Gene ID: ENSG00000073060, Gene Type: protein_coding\n",
      "No gene found for symbol 'LOC105373724' in Ensembl.\n",
      "Gene Symbol: CFDP1, Ensembl Gene ID: ENSG00000153774, Gene Type: protein_coding\n",
      "Gene Symbol: PRPF3, Ensembl Gene ID: ENSG00000117360, Gene Type: protein_coding\n",
      "Gene Symbol: TSPAN14, Ensembl Gene ID: ENSG00000108219, Gene Type: protein_coding\n",
      "Gene Symbol: PCSK9, Ensembl Gene ID: ENSG00000169174, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: CDKN1A, Ensembl Gene ID: ENSG00000124762, Gene Type: protein_coding\n",
      "Gene Symbol: FBN1, Ensembl Gene ID: ENSG00000166147, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: CDKN1A, Ensembl Gene ID: ENSG00000124762, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: LIPA, Ensembl Gene ID: ENSG00000107798, Gene Type: protein_coding\n",
      "Gene Symbol: PRDM6, Ensembl Gene ID: ENSG00000061455, Gene Type: protein_coding\n",
      "No gene found for symbol 'POLR1C       LOC105375069 LOC105375070' in Ensembl.\n",
      "No gene found for symbol 'LDLR     LDLR-AS1' in Ensembl.\n",
      "Gene Symbol: MRC2, Ensembl Gene ID: ENSG00000011028, Gene Type: protein_coding\n",
      "No gene found for symbol 'ZBTB46       LOC124904956' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: ZBTB38, Ensembl Gene ID: ENSG00000177311, Gene Type: protein_coding\n",
      "Gene Symbol: LRP1, Ensembl Gene ID: ENSG00000123384, Gene Type: protein_coding\n",
      "Gene Symbol: DAB2IP, Ensembl Gene ID: ENSG00000136848, Gene Type: protein_coding\n",
      "Gene Symbol: LPL, Ensembl Gene ID: ENSG00000175445, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: ANTXR1, Ensembl Gene ID: ENSG00000169604, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: APOA5, Ensembl Gene ID: ENSG00000110243, Gene Type: protein_coding\n",
      "Gene Symbol: ABCA6, Ensembl Gene ID: ENSG00000154262, Gene Type: protein_coding\n",
      "Gene Symbol: TAFA2, Ensembl Gene ID: ENSG00000198673, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: COL21A1, Ensembl Gene ID: ENSG00000124749, Gene Type: protein_coding\n",
      "Gene Symbol: CRISPLD2, Ensembl Gene ID: ENSG00000103196, Gene Type: protein_coding\n",
      "Gene Symbol: HHIPL1, Ensembl Gene ID: ENSG00000182218, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: TERT, Ensembl Gene ID: ENSG00000164362, Gene Type: protein_coding\n",
      "Gene Symbol: NUCB2, Ensembl Gene ID: ENSG00000070081, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'RANBP2 SH3RF3' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'LOC105377621' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'BDNF-AS   LINC00678' in Ensembl.\n",
      "Gene Symbol: ERG, Ensembl Gene ID: ENSG00000157554, Gene Type: protein_coding\n",
      "Gene Symbol: IST1, Ensembl Gene ID: ENSG00000182149, Gene Type: protein_coding\n",
      "Gene Symbol: CMIP, Ensembl Gene ID: ENSG00000153815, Gene Type: protein_coding\n",
      "Gene Symbol: IL6R, Ensembl Gene ID: ENSG00000160712, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: ADAMTS8, Ensembl Gene ID: ENSG00000134917, Gene Type: protein_coding\n",
      "Gene Symbol: HMGCR, Ensembl Gene ID: ENSG00000113161, Gene Type: protein_coding\n",
      "Gene Symbol: C20orf181, Ensembl Gene ID: ENSG00000279610, Gene Type: TEC\n",
      "Gene Symbol: ATOH8, Ensembl Gene ID: ENSG00000168874, Gene Type: protein_coding\n",
      "Gene Symbol: ZNF827, Ensembl Gene ID: ENSG00000151612, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: C5orf67, Ensembl Gene ID: ENSG00000225940, Gene Type: lncRNA\n",
      "No gene found for symbol 'ANKRD44     ANKRD44-IT1' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: MON1A, Ensembl Gene ID: ENSG00000164077, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "\n",
    "# Function to get Gene ID and Gene Type from Ensembl REST API\n",
    "def get_gene_info_via_ensembl(gene_symbol, species=\"homo_sapiens\"):\n",
    "    # Construct the URL for Ensembl's REST API\n",
    "    url = f\"https://rest.ensembl.org/lookup/symbol/{species}/{gene_symbol}\"\n",
    "    \n",
    "    # Set the headers to request JSON format\n",
    "    headers = {\"Content-Type\": \"application/json\"}\n",
    "    \n",
    "    try:\n",
    "        # Send the GET request to Ensembl's API\n",
    "        response = requests.get(url, headers=headers)\n",
    "        \n",
    "        # Check if the response is successful\n",
    "        if response.status_code == 200:\n",
    "            data = response.json()\n",
    "            \n",
    "            # Extract Gene ID and Gene Type\n",
    "            gene_id = data.get(\"id\", \"Gene ID not available\")\n",
    "            gene_type = data.get(\"biotype\", \"Gene type not available\")\n",
    "            \n",
    "            # Check if Gene Type is None and replace it\n",
    "            if gene_type is None:\n",
    "                gene_type = \"No gene reported\"\n",
    "            \n",
    "            print(f\"Gene Symbol: {gene_symbol}, Ensembl Gene ID: {gene_id}, Gene Type: {gene_type}\")\n",
    "            return gene_id, gene_type\n",
    "        elif response.status_code == 400:\n",
    "            print(f\"No gene found for symbol '{gene_symbol}' in Ensembl.\")\n",
    "            return \"No gene found\", \"Intergenic (Review)\"\n",
    "        else:\n",
    "            print(f\"Failed to retrieve data for {gene_symbol}. HTTP status: {response.status_code}\")\n",
    "            return None, \"Error retrieving data\"\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred for gene {gene_symbol}: {e}\")\n",
    "        return None, \"Error retrieving data\"\n",
    "\n",
    "# Assume filtered_final_df is your existing DataFrame with the gene symbols in a column named 'Gene Symbol'\n",
    "# Replace 'Gene Symbol' with the name of your actual column containing the gene symbols\n",
    "filtered_final_df[['Ensembl Gene ID', 'Gene Type']] = filtered_final_df['Gene ID (GENE_ID)'].apply(\n",
    "    lambda gene: pd.Series(get_gene_info_via_ensembl(gene))\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "34268edb-3400-4069-887d-34e3b2dc92ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "####Now let's output this table.\n",
    "filtered_final_df.to_csv(os.path.join(output_path, 'RENTREZ_NCBI_Compiled.csv'), index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "27fdd931-a77e-4d29-923f-a4e5e6f2a07d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID (GENE_ID)</th>\n",
       "      <th>Reported SNPs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDKN2B-AS1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LPA</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADAMTS8</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DAB2IP</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IL6R</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ERG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GDF7</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>LDLR     LDLR-AS1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CDKN1A</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PCSK9</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CRISPLD2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>LPL</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>LIPA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SUGCT</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>PLCE1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ZPR1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>APOE</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>LRP1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CELSR2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>POLR1C       LOC105375069 LOC105375070</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Gene ID (GENE_ID)  Reported SNPs\n",
       "0                               CDKN2B-AS1              4\n",
       "1                                      LPA              4\n",
       "2                                  ADAMTS8              3\n",
       "3                                   DAB2IP              3\n",
       "4                                     IL6R              3\n",
       "5                                      ERG              3\n",
       "6                                     GDF7              2\n",
       "7                        LDLR     LDLR-AS1              2\n",
       "8                                   CDKN1A              2\n",
       "9                                    PCSK9              2\n",
       "10                                CRISPLD2              2\n",
       "11                                     LPL              2\n",
       "12                                    LIPA              2\n",
       "13                                   SUGCT              2\n",
       "14                                   PLCE1              2\n",
       "15                                    ZPR1              2\n",
       "16                                    APOE              2\n",
       "17                                    LRP1              2\n",
       "18                                  CELSR2              2\n",
       "19  POLR1C       LOC105375069 LOC105375070              1"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Again unique genes\n",
    "Unique_Genes = filtered_final_df['Gene ID (GENE_ID)'].value_counts().reset_index()\n",
    "Unique_Genes.columns = ['Gene ID (GENE_ID)', 'Reported SNPs']\n",
    "Unique_Genes.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "56fe4f9f-7b10-400d-8c3c-d48dd5a2fa9a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID (GENE_ID)</th>\n",
       "      <th>dbSNP ID (RS)</th>\n",
       "      <th>SNP Class (SCLS)</th>\n",
       "      <th>Functional consequence class (FXN)</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Chromosome base position (CPOS)</th>\n",
       "      <th>Minor Allele Frequency from global population (GMAF)</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Associated clinical phenotypes (CLIN)</th>\n",
       "      <th>Nucleotide Accessions (ACCN)</th>\n",
       "      <th>Ensembl Gene ID</th>\n",
       "      <th>Gene Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LPA</td>\n",
       "      <td>10455872</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>6</td>\n",
       "      <td>6:160589086</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>A/G</td>\n",
       "      <td>benign</td>\n",
       "      <td>NC_000006.12</td>\n",
       "      <td>ENSG00000198670</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZNF335</td>\n",
       "      <td>3827066</td>\n",
       "      <td>snv</td>\n",
       "      <td>downstream_transcript_variant,intron_variant,g...</td>\n",
       "      <td>20</td>\n",
       "      <td>20:45957384</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>C/T</td>\n",
       "      <td>benign</td>\n",
       "      <td>NC_000020.11</td>\n",
       "      <td>ENSG00000198026</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CELSR2</td>\n",
       "      <td>12740374</td>\n",
       "      <td>snv</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>1</td>\n",
       "      <td>1:109274968</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>G/T</td>\n",
       "      <td>association</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>ENSG00000143126</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>APOE</td>\n",
       "      <td>429358</td>\n",
       "      <td>snv</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>19</td>\n",
       "      <td>19:44908684</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>C/T</td>\n",
       "      <td>pathogenic,risk-factor,association,conflicting...</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>ENSG00000130203</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ADAMTS8</td>\n",
       "      <td>4936098</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>11</td>\n",
       "      <td>11:130410772</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>A/G/T</td>\n",
       "      <td></td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>ENSG00000134917</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CHRNA5</td>\n",
       "      <td>17486278</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>15</td>\n",
       "      <td>15:78575140</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>A/C</td>\n",
       "      <td></td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>ENSG00000169684</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>LPA</td>\n",
       "      <td>140570886</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>6</td>\n",
       "      <td>6:160591981</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>C/T</td>\n",
       "      <td></td>\n",
       "      <td>NC_000006.12</td>\n",
       "      <td>ENSG00000198670</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ZPR1</td>\n",
       "      <td>964184</td>\n",
       "      <td>snv</td>\n",
       "      <td>3_prime_UTR_variant,genic_downstream_transcrip...</td>\n",
       "      <td>11</td>\n",
       "      <td>11:116778201</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>G/C</td>\n",
       "      <td></td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>ENSG00000109917</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>DAB2IP</td>\n",
       "      <td>7025486</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant,genic_upstream_transcript_variant</td>\n",
       "      <td>9</td>\n",
       "      <td>9:121660124</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...</td>\n",
       "      <td>A/G</td>\n",
       "      <td></td>\n",
       "      <td>NC_000009.12</td>\n",
       "      <td>ENSG00000136848</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>PLCE1</td>\n",
       "      <td>731141</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>10</td>\n",
       "      <td>10:94138924</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...</td>\n",
       "      <td>A/G</td>\n",
       "      <td></td>\n",
       "      <td>NC_000010.11</td>\n",
       "      <td>ENSG00000138193</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Gene ID (GENE_ID) dbSNP ID (RS) SNP Class (SCLS)  \\\n",
       "1                LPA      10455872              snv   \n",
       "2             ZNF335       3827066              snv   \n",
       "3             CELSR2      12740374              snv   \n",
       "5               APOE        429358              snv   \n",
       "7            ADAMTS8       4936098              snv   \n",
       "8             CHRNA5      17486278              snv   \n",
       "10               LPA     140570886              snv   \n",
       "13              ZPR1        964184              snv   \n",
       "14            DAB2IP       7025486              snv   \n",
       "15             PLCE1        731141              snv   \n",
       "\n",
       "                   Functional consequence class (FXN) Chromosome  \\\n",
       "1                                      intron_variant          6   \n",
       "2   downstream_transcript_variant,intron_variant,g...         20   \n",
       "3                                 3_prime_UTR_variant          1   \n",
       "5            missense_variant,coding_sequence_variant         19   \n",
       "7                                      intron_variant         11   \n",
       "8                                      intron_variant         15   \n",
       "10                                     intron_variant          6   \n",
       "13  3_prime_UTR_variant,genic_downstream_transcrip...         11   \n",
       "14   intron_variant,genic_upstream_transcript_variant          9   \n",
       "15                                     intron_variant         10   \n",
       "\n",
       "   Chromosome base position (CPOS)  \\\n",
       "1                      6:160589086   \n",
       "2                      20:45957384   \n",
       "3                      1:109274968   \n",
       "5                      19:44908684   \n",
       "7                     11:130410772   \n",
       "8                      15:78575140   \n",
       "10                     6:160591981   \n",
       "13                    11:116778201   \n",
       "14                     9:121660124   \n",
       "15                     10:94138924   \n",
       "\n",
       "   Minor Allele Frequency from global population (GMAF) Allele  \\\n",
       "1    [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      A/G   \n",
       "2    [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      C/T   \n",
       "3    [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      G/T   \n",
       "5    [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      C/T   \n",
       "7    [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...    A/G/T   \n",
       "8    [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      A/C   \n",
       "10   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      C/T   \n",
       "13   [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      G/C   \n",
       "14   [1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...      A/G   \n",
       "15   [1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...      A/G   \n",
       "\n",
       "                Associated clinical phenotypes (CLIN)  \\\n",
       "1                                              benign   \n",
       "2                                              benign   \n",
       "3                                         association   \n",
       "5   pathogenic,risk-factor,association,conflicting...   \n",
       "7                                                       \n",
       "8                                                       \n",
       "10                                                      \n",
       "13                                                      \n",
       "14                                                      \n",
       "15                                                      \n",
       "\n",
       "   Nucleotide Accessions (ACCN)  Ensembl Gene ID       Gene Type  \n",
       "1                  NC_000006.12  ENSG00000198670  protein_coding  \n",
       "2                  NC_000020.11  ENSG00000198026  protein_coding  \n",
       "3                  NC_000001.11  ENSG00000143126  protein_coding  \n",
       "5                  NC_000019.10  ENSG00000130203  protein_coding  \n",
       "7                  NC_000011.10  ENSG00000134917  protein_coding  \n",
       "8                  NC_000015.10  ENSG00000169684  protein_coding  \n",
       "10                 NC_000006.12  ENSG00000198670  protein_coding  \n",
       "13                 NC_000011.10  ENSG00000109917  protein_coding  \n",
       "14                 NC_000009.12  ENSG00000136848  protein_coding  \n",
       "15                 NC_000010.11  ENSG00000138193  protein_coding  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Assuming filtered_final_df is your DataFrame\n",
    "# Filter rows where Gene Type is 'protein-coding'\n",
    "protein_coding_df = filtered_final_df[filtered_final_df['Gene Type'] == 'protein_coding']\n",
    "\n",
    "# Display the filtered DataFrame\n",
    "protein_coding_df.head(10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "ece973ac-c826-40f2-bc67-4c6425468950",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Protein-coding genes have been saved to 'Resources/OutputFromExtractionCode'\n",
      "['LPA', 'ZNF335', 'CELSR2', 'APOE', 'ADAMTS8']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Assuming filtered_final_df is your DataFrame with gene symbols and gene types\n",
    "\n",
    "# Filter rows with 'protein-coding' in the Gene Type column\n",
    "protein_coding_df = filtered_final_df[filtered_final_df['Gene Type'] == 'protein_coding']\n",
    "\n",
    "# Extract the 'Gene ID (GENE_ID)' column as a list\n",
    "protein_coding_genes = protein_coding_df['Gene ID (GENE_ID)'].tolist()\n",
    "\n",
    "# Define the output path for the text file\n",
    "output_path_txt = f'{output_path}/protein_coding_genes.txt'  # You can set this to any desired path\n",
    "\n",
    "# Save the list to a text file with each gene on a new line\n",
    "with open(output_path_txt, \"w\") as file:\n",
    "    for gene in protein_coding_genes:\n",
    "        file.write(f\"{gene}\\n\")\n",
    "\n",
    "print(f\"Protein-coding genes have been saved to '{output_path}'\")\n",
    "\n",
    "# Display the first few items in the list for verification\n",
    "print(protein_coding_genes[:5])  # Shows the first 5 genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a96b6db4-e627-4137-9f16-6c570948ea9d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd6a42f0-ec8f-45cd-bfe8-8a23f2aeb042",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
