{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "9d9b8fd4-0498-44fc-9d30-bd878bb7cfa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Now since this is a new file we need to import other modules (YAY!) including our R' emulator\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects.packages import importr\n",
    "from rpy2.robjects import pandas2ri\n",
    "\n",
    "### and let's not forget the usual gang:\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib as plt\n",
    "import re\n",
    "import os "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "9b8c2979-866d-405e-a5f1-fd872a08cdb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Before we get started in an R' environment. We need to talk about something important.\n",
    "#### Rentrez does not use SNP ID to search. It uses UID. As in, they don't use 'rs2000' \n",
    "#### but rather '2000'. So we need to use our previous output and split the function.\n",
    "Output_GeneSNP_file = pd.read_csv('Resources/OutputFromExtractionCode/Editable_Gene_And_dbSNP_ID_Merged.csv')\n",
    "Output_GeneSNP_file.head(10)\n",
    "\n",
    "output_path = 'Resources/OutputFromExtractionCode'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "751ff5a8-be40-41ee-b44d-ece195aa51da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>SNP ID</th>\n",
       "      <th>UID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDKN2B-AS1</td>\n",
       "      <td>rs4977574</td>\n",
       "      <td>4977574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LPA</td>\n",
       "      <td>rs10455872</td>\n",
       "      <td>10455872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZNF335</td>\n",
       "      <td>rs3827066</td>\n",
       "      <td>3827066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CELSR2</td>\n",
       "      <td>rs12740374</td>\n",
       "      <td>12740374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LINC00540 - FTH1P7</td>\n",
       "      <td>rs7994761</td>\n",
       "      <td>7994761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>APOE</td>\n",
       "      <td>rs429358</td>\n",
       "      <td>429358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SMARCA4</td>\n",
       "      <td>rs73015011</td>\n",
       "      <td>73015011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ADAMTS8</td>\n",
       "      <td>rs4936098</td>\n",
       "      <td>4936098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CHRNA5</td>\n",
       "      <td>rs17486278</td>\n",
       "      <td>17486278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CDKN2B-AS1</td>\n",
       "      <td>rs2891168</td>\n",
       "      <td>2891168</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GENE      SNP ID       UID\n",
       "0          CDKN2B-AS1   rs4977574   4977574\n",
       "1                 LPA  rs10455872  10455872\n",
       "2              ZNF335   rs3827066   3827066\n",
       "3              CELSR2  rs12740374  12740374\n",
       "4  LINC00540 - FTH1P7   rs7994761   7994761\n",
       "5                APOE    rs429358    429358\n",
       "6             SMARCA4  rs73015011  73015011\n",
       "7             ADAMTS8   rs4936098   4936098\n",
       "8              CHRNA5  rs17486278  17486278\n",
       "9          CDKN2B-AS1   rs2891168   2891168"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Now we split it. \n",
    "Split_dbSNP_ID = Output_GeneSNP_file\n",
    "Split_dbSNP_ID[['rs', 'UID']] = Output_GeneSNP_file['SNP ID'].str.split('rs', expand = True)\n",
    "Split_dbSNP_ID.head(10)\n",
    "\n",
    "#And delete the rs column:\n",
    "UID_SNP = Split_dbSNP_ID.drop(columns=['rs'])\n",
    "UID_SNP.head(10)\n",
    "\n",
    "## Now we have the UID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "d59ca7d3-ffa1-4e16-b92b-8d440f53957c",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Let's import RENTREZ to our python environment\n",
    "rentrez = importr('rentrez')\n",
    "\n",
    "### You'll need to provide an email address to use Entrez tools. BTW Hot tip: \n",
    "### If you have multiple lines, using triple quotes is cleaner and easier to read:\n",
    "\n",
    "robjects.r('''\n",
    "library(rentrez)\n",
    "entrez_email <- \"kenji.ponsm@gmail.com\"\n",
    "''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "2b86d25f-7e00-496f-90dd-3ba9b8db7d8c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4977574\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>allele_origin</th>\n",
       "      <th>global_mafs</th>\n",
       "      <th>global_population</th>\n",
       "      <th>global_samplesize</th>\n",
       "      <th>suspected</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>genes</th>\n",
       "      <th>acc</th>\n",
       "      <th>...</th>\n",
       "      <th>allele</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>chrpos_prev_assm</th>\n",
       "      <th>text</th>\n",
       "      <th>snp_id_sort</th>\n",
       "      <th>clinical_sort</th>\n",
       "      <th>cited_sort</th>\n",
       "      <th>chrpos_sort</th>\n",
       "      <th>merged_sort</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4977574</td>\n",
       "      <td>4977574</td>\n",
       "      <td></td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>risk-factor</td>\n",
       "      <td>[1] \"CDKN2B-AS1\"\\n</td>\n",
       "      <td>NC_000009.12</td>\n",
       "      <td>...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>9:22098575</td>\n",
       "      <td>9:22098574</td>\n",
       "      <td></td>\n",
       "      <td>0004977574</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>0022098575</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       uid   snp_id allele_origin  \\\n",
       "0  4977574  4977574                 \n",
       "\n",
       "                                         global_mafs global_population  \\\n",
       "0   [1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...                     \n",
       "\n",
       "  global_samplesize suspected clinical_significance               genes  \\\n",
       "0                                       risk-factor  [1] \"CDKN2B-AS1\"\\n   \n",
       "\n",
       "            acc  ... allele snp_class      chrpos chrpos_prev_assm text  \\\n",
       "0  NC_000009.12  ...      D       snv  9:22098575       9:22098574        \n",
       "\n",
       "  snp_id_sort clinical_sort cited_sort chrpos_sort merged_sort  \n",
       "0  0004977574             1             0022098575           0  \n",
       "\n",
       "[1 rows x 32 columns]"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Let's give it a test shall we? Let's say I want the information of rs429358 (GENE: APOE)\n",
    "print(UID_SNP.iloc[0,2])\n",
    "test_snp = UID_SNP.iloc[0,2]\n",
    "search_test = rentrez.entrez_summary(db='snp', id=test_snp)\n",
    "search_test_dict = {key: str(search_test.rx2(key) [0]) for key in search_test.names}\n",
    "\n",
    "search_test_df = pd.DataFrame([search_test_dict])\n",
    "\n",
    "search_test_df.head(10)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "42a5e8ad-01e1-4263-ac18-3c39726932e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [1] \"uid\"                   \"snp_id\"                \"allele_origin\"        \n",
      " [4] \"global_mafs\"           \"global_population\"     \"global_samplesize\"    \n",
      " [7] \"suspected\"             \"clinical_significance\" \"genes\"                \n",
      "[10] \"acc\"                   \"chr\"                   \"handle\"               \n",
      "[13] \"spdi\"                  \"fxn_class\"             \"validated\"            \n",
      "[16] \"docsum\"                \"tax_id\"                \"orig_build\"           \n",
      "[19] \"upd_build\"             \"createdate\"            \"updatedate\"           \n",
      "[22] \"ss\"                    \"allele\"                \"snp_class\"            \n",
      "[25] \"chrpos\"                \"chrpos_prev_assm\"      \"text\"                 \n",
      "[28] \"snp_id_sort\"           \"clinical_sort\"         \"cited_sort\"           \n",
      "[31] \"chrpos_sort\"           \"merged_sort\"          \n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(search_test.names)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "c48732cd-c951-44bb-9104-227be736a212",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>global_population</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>genes</th>\n",
       "      <th>acc</th>\n",
       "      <th>chr</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>allele</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>chrpos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4977574</td>\n",
       "      <td>4977574</td>\n",
       "      <td></td>\n",
       "      <td>risk-factor</td>\n",
       "      <td>[1] \"CDKN2B-AS1\"\\n</td>\n",
       "      <td>NC_000009.12</td>\n",
       "      <td>9</td>\n",
       "      <td>genic_downstream_transcript_variant,intron_var...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>9:22098575</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       uid   snp_id global_population clinical_significance  \\\n",
       "0  4977574  4977574                             risk-factor   \n",
       "\n",
       "                genes           acc chr  \\\n",
       "0  [1] \"CDKN2B-AS1\"\\n  NC_000009.12   9   \n",
       "\n",
       "                                           fxn_class allele snp_class  \\\n",
       "0  genic_downstream_transcript_variant,intron_var...      D       snv   \n",
       "\n",
       "       chrpos  \n",
       "0  9:22098575  "
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### And let's filter for the actual values we want to extract: \n",
    "desired_dbSNP_values = (['genes', 'snp_id', 'uid', 'snp_class', \n",
    "                         'fxn_class', 'chr', 'chrpos', 'global_population', \n",
    "                         'allele','clinical_significance', 'acc'] \n",
    "                         )\n",
    "filtered_search_test_dict = {key: str(search_test.rx2(key)[0]) for key in search_test.names if key in desired_dbSNP_values}\n",
    "filtered_search_test = pd.DataFrame([filtered_search_test_dict])\n",
    "filtered_search_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "fc3d7c7d-291c-4ee1-b728-9fd3dc5c9971",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>uid</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>chr</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>global_population</th>\n",
       "      <th>allele</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>acc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[1] \"CDKN2B-AS1\"\\n</td>\n",
       "      <td>4977574</td>\n",
       "      <td>4977574</td>\n",
       "      <td>snv</td>\n",
       "      <td>genic_downstream_transcript_variant,intron_var...</td>\n",
       "      <td>9</td>\n",
       "      <td>9:22098575</td>\n",
       "      <td></td>\n",
       "      <td>D</td>\n",
       "      <td>risk-factor</td>\n",
       "      <td>NC_000009.12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                genes   snp_id      uid snp_class  \\\n",
       "0  [1] \"CDKN2B-AS1\"\\n  4977574  4977574       snv   \n",
       "\n",
       "                                           fxn_class chr      chrpos  \\\n",
       "0  genic_downstream_transcript_variant,intron_var...   9  9:22098575   \n",
       "\n",
       "  global_population allele clinical_significance           acc  \n",
       "0                        D           risk-factor  NC_000009.12  "
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_and_ordered_search_test = filtered_search_test[[col for col in desired_dbSNP_values if col in filtered_search_test.columns]]\n",
    "filtered_and_ordered_search_test\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "9827951a-19a9-4922-b977-d9bf27b7921c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping row 53 because UID is None. Row details:\n",
      "GENE                 NaN\n",
      "SNP ID    chr6:112398675\n",
      "UID                 None\n",
      "Name: 53, dtype: object\n",
      "\n",
      "\n",
      "Skipping row 116 because UID is None. Row details:\n",
      "GENE                NaN\n",
      "SNP ID    chr2:20679060\n",
      "UID                None\n",
      "Name: 116, dtype: object\n",
      "\n",
      "\n",
      "Skipping row 140 because UID is None. Row details:\n",
      "GENE                  NaN\n",
      "SNP ID    chr11:130410772\n",
      "UID                  None\n",
      "Name: 140, dtype: object\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>uid</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>chr</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>global_population</th>\n",
       "      <th>allele</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>acc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[1] \"CDKN2B-AS1\"\\n</td>\n",
       "      <td>4977574</td>\n",
       "      <td>4977574</td>\n",
       "      <td>snv</td>\n",
       "      <td>genic_downstream_transcript_variant,intron_var...</td>\n",
       "      <td>9</td>\n",
       "      <td>9:22098575</td>\n",
       "      <td></td>\n",
       "      <td>D</td>\n",
       "      <td>risk-factor</td>\n",
       "      <td>NC_000009.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[1] \"LPA\"\\n</td>\n",
       "      <td>10455872</td>\n",
       "      <td>10455872</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>6</td>\n",
       "      <td>6:160589086</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td>benign</td>\n",
       "      <td>NC_000006.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[1] \"ZNF335\"\\n</td>\n",
       "      <td>3827066</td>\n",
       "      <td>3827066</td>\n",
       "      <td>snv</td>\n",
       "      <td>downstream_transcript_variant,intron_variant,g...</td>\n",
       "      <td>20</td>\n",
       "      <td>20:45957384</td>\n",
       "      <td></td>\n",
       "      <td>Y</td>\n",
       "      <td>benign</td>\n",
       "      <td>NC_000020.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[1] \"CELSR2\"\\n</td>\n",
       "      <td>12740374</td>\n",
       "      <td>12740374</td>\n",
       "      <td>snv</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>1</td>\n",
       "      <td>1:109274968</td>\n",
       "      <td></td>\n",
       "      <td>K</td>\n",
       "      <td>association</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>None</td>\n",
       "      <td>7994761</td>\n",
       "      <td>7994761</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>13</td>\n",
       "      <td>13:22297307</td>\n",
       "      <td></td>\n",
       "      <td>D</td>\n",
       "      <td></td>\n",
       "      <td>NC_000013.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>[1] \"APOE\"\\n</td>\n",
       "      <td>429358</td>\n",
       "      <td>429358</td>\n",
       "      <td>snv</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>19</td>\n",
       "      <td>19:44908684</td>\n",
       "      <td></td>\n",
       "      <td>Y</td>\n",
       "      <td>pathogenic,risk-factor,association,conflicting...</td>\n",
       "      <td>NC_000019.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>None</td>\n",
       "      <td>73015011</td>\n",
       "      <td>73015011</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>19</td>\n",
       "      <td>19:11079088</td>\n",
       "      <td></td>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>NC_000019.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>[1] \"ADAMTS8\"\\n</td>\n",
       "      <td>4936098</td>\n",
       "      <td>4936098</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>11</td>\n",
       "      <td>11:130410772</td>\n",
       "      <td></td>\n",
       "      <td>D</td>\n",
       "      <td></td>\n",
       "      <td>NC_000011.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>[1] \"CHRNA5\"\\n</td>\n",
       "      <td>17486278</td>\n",
       "      <td>17486278</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>15</td>\n",
       "      <td>15:78575140</td>\n",
       "      <td></td>\n",
       "      <td>M</td>\n",
       "      <td></td>\n",
       "      <td>NC_000015.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>[1] \"CDKN2B-AS1\"\\n</td>\n",
       "      <td>2891168</td>\n",
       "      <td>2891168</td>\n",
       "      <td>snv</td>\n",
       "      <td>genic_downstream_transcript_variant,intron_var...</td>\n",
       "      <td>9</td>\n",
       "      <td>9:22098620</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000009.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>[1] \"LPA\"\\n</td>\n",
       "      <td>140570886</td>\n",
       "      <td>140570886</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>6</td>\n",
       "      <td>6:160591981</td>\n",
       "      <td></td>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>NC_000006.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>[1] \"CDKN2B-AS1\"\\n</td>\n",
       "      <td>10757274</td>\n",
       "      <td>10757274</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant,genic_downstream_transcript_var...</td>\n",
       "      <td>9</td>\n",
       "      <td>9:22096056</td>\n",
       "      <td></td>\n",
       "      <td>N</td>\n",
       "      <td>risk-factor</td>\n",
       "      <td>NC_000009.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>None</td>\n",
       "      <td>434182</td>\n",
       "      <td>434182</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>18</td>\n",
       "      <td>18:22702893</td>\n",
       "      <td></td>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>NC_000018.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>[1] \"ZPR1\"\\n</td>\n",
       "      <td>964184</td>\n",
       "      <td>964184</td>\n",
       "      <td>snv</td>\n",
       "      <td>3_prime_UTR_variant,genic_downstream_transcrip...</td>\n",
       "      <td>11</td>\n",
       "      <td>11:116778201</td>\n",
       "      <td></td>\n",
       "      <td>S</td>\n",
       "      <td></td>\n",
       "      <td>NC_000011.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>[1] \"DAB2IP\"\\n</td>\n",
       "      <td>7025486</td>\n",
       "      <td>7025486</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant,genic_upstream_transcript_variant</td>\n",
       "      <td>9</td>\n",
       "      <td>9:121660124</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000009.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>[1] \"PLCE1\"\\n</td>\n",
       "      <td>731141</td>\n",
       "      <td>731141</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>10</td>\n",
       "      <td>10:94138924</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000010.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>[1] \"IL6R\"\\n</td>\n",
       "      <td>12126142</td>\n",
       "      <td>12126142</td>\n",
       "      <td>snv</td>\n",
       "      <td>genic_downstream_transcript_variant,intron_var...</td>\n",
       "      <td>1</td>\n",
       "      <td>1:154452980</td>\n",
       "      <td></td>\n",
       "      <td>V</td>\n",
       "      <td></td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>[1] \"LRP1\"\\n</td>\n",
       "      <td>11172113</td>\n",
       "      <td>11172113</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>12</td>\n",
       "      <td>12:57133500</td>\n",
       "      <td></td>\n",
       "      <td>B</td>\n",
       "      <td></td>\n",
       "      <td>NC_000012.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>[1] \"SUGCT\"\\n</td>\n",
       "      <td>17688414</td>\n",
       "      <td>17688414</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant,genic_downstream_transcript_var...</td>\n",
       "      <td>7</td>\n",
       "      <td>7:40801371</td>\n",
       "      <td></td>\n",
       "      <td>B</td>\n",
       "      <td></td>\n",
       "      <td>NC_000007.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>[1] \"GDF7\"\\n</td>\n",
       "      <td>9306894</td>\n",
       "      <td>9306894</td>\n",
       "      <td>snv</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>2</td>\n",
       "      <td>2:20678345</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000002.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>None</td>\n",
       "      <td>1660364</td>\n",
       "      <td>1660364</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1:214232262</td>\n",
       "      <td></td>\n",
       "      <td>H</td>\n",
       "      <td></td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>None</td>\n",
       "      <td>1892971</td>\n",
       "      <td>1892971</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>11</td>\n",
       "      <td>11:102924877</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000011.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>None</td>\n",
       "      <td>77925345</td>\n",
       "      <td>77925345</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>12</td>\n",
       "      <td>12:19777560</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000012.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>[1] \"LPA\"\\n</td>\n",
       "      <td>10455872</td>\n",
       "      <td>10455872</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>6</td>\n",
       "      <td>6:160589086</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td>benign</td>\n",
       "      <td>NC_000006.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>[1] \"ERG\"\\n</td>\n",
       "      <td>2836411</td>\n",
       "      <td>2836411</td>\n",
       "      <td>snv</td>\n",
       "      <td>genic_downstream_transcript_variant,intron_var...</td>\n",
       "      <td>21</td>\n",
       "      <td>21:38447907</td>\n",
       "      <td></td>\n",
       "      <td>H</td>\n",
       "      <td></td>\n",
       "      <td>NC_000021.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>None</td>\n",
       "      <td>2954021</td>\n",
       "      <td>2954021</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>8</td>\n",
       "      <td>8:125469835</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000008.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>[1] \"ZPR1\"\\n</td>\n",
       "      <td>964184</td>\n",
       "      <td>964184</td>\n",
       "      <td>snv</td>\n",
       "      <td>3_prime_UTR_variant,genic_downstream_transcrip...</td>\n",
       "      <td>11</td>\n",
       "      <td>11:116778201</td>\n",
       "      <td></td>\n",
       "      <td>S</td>\n",
       "      <td></td>\n",
       "      <td>NC_000011.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>[1] \"LPL\"\\n</td>\n",
       "      <td>331</td>\n",
       "      <td>331</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>8</td>\n",
       "      <td>8:19962894</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000008.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>[1] \"LIPA\"\\n</td>\n",
       "      <td>1412444</td>\n",
       "      <td>1412444</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>10</td>\n",
       "      <td>10:89243170</td>\n",
       "      <td></td>\n",
       "      <td>B</td>\n",
       "      <td></td>\n",
       "      <td>NC_000010.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>[1] \"CELSR2\"\\n</td>\n",
       "      <td>12740374</td>\n",
       "      <td>12740374</td>\n",
       "      <td>snv</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>1</td>\n",
       "      <td>1:109274968</td>\n",
       "      <td></td>\n",
       "      <td>K</td>\n",
       "      <td>association</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>[1] \"LPA\"\\n</td>\n",
       "      <td>118039278</td>\n",
       "      <td>118039278</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>6</td>\n",
       "      <td>6:160564494</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000006.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>[1] \"PLCE1\"\\n</td>\n",
       "      <td>11187851</td>\n",
       "      <td>11187851</td>\n",
       "      <td>snv</td>\n",
       "      <td>genic_downstream_transcript_variant,intron_var...</td>\n",
       "      <td>10</td>\n",
       "      <td>10:94310118</td>\n",
       "      <td></td>\n",
       "      <td>D</td>\n",
       "      <td></td>\n",
       "      <td>NC_000010.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>[1] \"PLAUR\"\\n</td>\n",
       "      <td>4760</td>\n",
       "      <td>4760</td>\n",
       "      <td>snv</td>\n",
       "      <td>coding_sequence_variant,genic_downstream_trans...</td>\n",
       "      <td>19</td>\n",
       "      <td>19:43648948</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000019.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>[1] \"PCIF1\"\\n</td>\n",
       "      <td>58749629</td>\n",
       "      <td>58749629</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant,upstream_transcript_variant,gen...</td>\n",
       "      <td>20</td>\n",
       "      <td>20:45942678</td>\n",
       "      <td></td>\n",
       "      <td>D</td>\n",
       "      <td></td>\n",
       "      <td>NC_000020.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>None</td>\n",
       "      <td>7240701</td>\n",
       "      <td>7240701</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>18</td>\n",
       "      <td>18:22523104</td>\n",
       "      <td></td>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>NC_000018.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>None</td>\n",
       "      <td>247617</td>\n",
       "      <td>247617</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>16</td>\n",
       "      <td>16:56956804</td>\n",
       "      <td></td>\n",
       "      <td>M</td>\n",
       "      <td></td>\n",
       "      <td>NC_000016.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>None</td>\n",
       "      <td>143235478</td>\n",
       "      <td>143235478</td>\n",
       "      <td>delins</td>\n",
       "      <td></td>\n",
       "      <td>20</td>\n",
       "      <td>20:45927137</td>\n",
       "      <td></td>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>NC_000020.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>None</td>\n",
       "      <td>6734238</td>\n",
       "      <td>6734238</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>2:113083453</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000002.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>[1] \"VMP1\"\\n</td>\n",
       "      <td>2150879</td>\n",
       "      <td>2150879</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>17</td>\n",
       "      <td>17:59781849</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000017.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>None</td>\n",
       "      <td>17368814</td>\n",
       "      <td>17368814</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>11</td>\n",
       "      <td>11:102877965</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000011.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>[1] \"ADAMTS8\"\\n</td>\n",
       "      <td>4936098</td>\n",
       "      <td>4936098</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>11</td>\n",
       "      <td>11:130410772</td>\n",
       "      <td></td>\n",
       "      <td>D</td>\n",
       "      <td></td>\n",
       "      <td>NC_000011.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>[1] \"PCSK9\"\\n</td>\n",
       "      <td>11591147</td>\n",
       "      <td>11591147</td>\n",
       "      <td>snv</td>\n",
       "      <td>upstream_transcript_variant,coding_sequence_va...</td>\n",
       "      <td>1</td>\n",
       "      <td>1:55039974</td>\n",
       "      <td></td>\n",
       "      <td>D</td>\n",
       "      <td></td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>None</td>\n",
       "      <td>12408895</td>\n",
       "      <td>12408895</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1:9378116</td>\n",
       "      <td></td>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>[1] \"LTBP4\"\\n</td>\n",
       "      <td>34093919</td>\n",
       "      <td>34093919</td>\n",
       "      <td>snv</td>\n",
       "      <td>coding_sequence_variant,missense_variant</td>\n",
       "      <td>19</td>\n",
       "      <td>19:40611394</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td>benign,benign-likely-benign</td>\n",
       "      <td>NC_000019.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>[1] \"APOE\"\\n</td>\n",
       "      <td>429358</td>\n",
       "      <td>429358</td>\n",
       "      <td>snv</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>19</td>\n",
       "      <td>19:44908684</td>\n",
       "      <td></td>\n",
       "      <td>Y</td>\n",
       "      <td>pathogenic,risk-factor,association,conflicting...</td>\n",
       "      <td>NC_000019.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>[1] \"SUGCT\"\\n</td>\n",
       "      <td>79486579</td>\n",
       "      <td>79486579</td>\n",
       "      <td>snv</td>\n",
       "      <td>genic_downstream_transcript_variant,intron_var...</td>\n",
       "      <td>7</td>\n",
       "      <td>7:40391533</td>\n",
       "      <td></td>\n",
       "      <td>V</td>\n",
       "      <td></td>\n",
       "      <td>NC_000007.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>[1] \"LOC105378327\" \"LOC124902436\"\\n</td>\n",
       "      <td>12768538</td>\n",
       "      <td>12768538</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant,genic_upstream_transcript_variant</td>\n",
       "      <td>10</td>\n",
       "      <td>10:62646598</td>\n",
       "      <td></td>\n",
       "      <td>S</td>\n",
       "      <td></td>\n",
       "      <td>NC_000010.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>[1] \"DNM3\"\\n</td>\n",
       "      <td>2227200</td>\n",
       "      <td>2227200</td>\n",
       "      <td>snv</td>\n",
       "      <td>genic_downstream_transcript_variant,intron_var...</td>\n",
       "      <td>1</td>\n",
       "      <td>1:172364308</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>[1] \"LOC124904262\"\\n</td>\n",
       "      <td>4401144</td>\n",
       "      <td>4401144</td>\n",
       "      <td>snv</td>\n",
       "      <td>non_coding_transcript_variant</td>\n",
       "      <td>18</td>\n",
       "      <td>18:22613705</td>\n",
       "      <td></td>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>NC_000018.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>[1] \"LINC02732\"\\n</td>\n",
       "      <td>10891091</td>\n",
       "      <td>10891091</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>11</td>\n",
       "      <td>11:110358805</td>\n",
       "      <td></td>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>NC_000011.10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  genes     snp_id        uid snp_class  \\\n",
       "0                    [1] \"CDKN2B-AS1\"\\n    4977574    4977574       snv   \n",
       "1                           [1] \"LPA\"\\n   10455872   10455872       snv   \n",
       "2                        [1] \"ZNF335\"\\n    3827066    3827066       snv   \n",
       "3                        [1] \"CELSR2\"\\n   12740374   12740374       snv   \n",
       "4                                  None    7994761    7994761       snv   \n",
       "5                          [1] \"APOE\"\\n     429358     429358       snv   \n",
       "6                                  None   73015011   73015011       snv   \n",
       "7                       [1] \"ADAMTS8\"\\n    4936098    4936098       snv   \n",
       "8                        [1] \"CHRNA5\"\\n   17486278   17486278       snv   \n",
       "9                    [1] \"CDKN2B-AS1\"\\n    2891168    2891168       snv   \n",
       "10                          [1] \"LPA\"\\n  140570886  140570886       snv   \n",
       "11                   [1] \"CDKN2B-AS1\"\\n   10757274   10757274       snv   \n",
       "12                                 None     434182     434182       snv   \n",
       "13                         [1] \"ZPR1\"\\n     964184     964184       snv   \n",
       "14                       [1] \"DAB2IP\"\\n    7025486    7025486       snv   \n",
       "15                        [1] \"PLCE1\"\\n     731141     731141       snv   \n",
       "16                         [1] \"IL6R\"\\n   12126142   12126142       snv   \n",
       "17                         [1] \"LRP1\"\\n   11172113   11172113       snv   \n",
       "18                        [1] \"SUGCT\"\\n   17688414   17688414       snv   \n",
       "19                         [1] \"GDF7\"\\n    9306894    9306894       snv   \n",
       "20                                 None    1660364    1660364       snv   \n",
       "21                                 None    1892971    1892971       snv   \n",
       "22                                 None   77925345   77925345       snv   \n",
       "23                          [1] \"LPA\"\\n   10455872   10455872       snv   \n",
       "24                          [1] \"ERG\"\\n    2836411    2836411       snv   \n",
       "25                                 None    2954021    2954021       snv   \n",
       "26                         [1] \"ZPR1\"\\n     964184     964184       snv   \n",
       "27                          [1] \"LPL\"\\n        331        331       snv   \n",
       "28                         [1] \"LIPA\"\\n    1412444    1412444       snv   \n",
       "29                       [1] \"CELSR2\"\\n   12740374   12740374       snv   \n",
       "30                          [1] \"LPA\"\\n  118039278  118039278       snv   \n",
       "31                        [1] \"PLCE1\"\\n   11187851   11187851       snv   \n",
       "32                        [1] \"PLAUR\"\\n       4760       4760       snv   \n",
       "33                        [1] \"PCIF1\"\\n   58749629   58749629       snv   \n",
       "34                                 None    7240701    7240701       snv   \n",
       "35                                 None     247617     247617       snv   \n",
       "36                                 None  143235478  143235478    delins   \n",
       "37                                 None    6734238    6734238       snv   \n",
       "38                         [1] \"VMP1\"\\n    2150879    2150879       snv   \n",
       "39                                 None   17368814   17368814       snv   \n",
       "40                      [1] \"ADAMTS8\"\\n    4936098    4936098       snv   \n",
       "41                        [1] \"PCSK9\"\\n   11591147   11591147       snv   \n",
       "42                                 None   12408895   12408895       snv   \n",
       "43                        [1] \"LTBP4\"\\n   34093919   34093919       snv   \n",
       "44                         [1] \"APOE\"\\n     429358     429358       snv   \n",
       "45                        [1] \"SUGCT\"\\n   79486579   79486579       snv   \n",
       "46  [1] \"LOC105378327\" \"LOC124902436\"\\n   12768538   12768538       snv   \n",
       "47                         [1] \"DNM3\"\\n    2227200    2227200       snv   \n",
       "48                 [1] \"LOC124904262\"\\n    4401144    4401144       snv   \n",
       "49                    [1] \"LINC02732\"\\n   10891091   10891091       snv   \n",
       "\n",
       "                                            fxn_class chr        chrpos  \\\n",
       "0   genic_downstream_transcript_variant,intron_var...   9    9:22098575   \n",
       "1                                      intron_variant   6   6:160589086   \n",
       "2   downstream_transcript_variant,intron_variant,g...  20   20:45957384   \n",
       "3                                 3_prime_UTR_variant   1   1:109274968   \n",
       "4                                                      13   13:22297307   \n",
       "5            missense_variant,coding_sequence_variant  19   19:44908684   \n",
       "6                                                      19   19:11079088   \n",
       "7                                      intron_variant  11  11:130410772   \n",
       "8                                      intron_variant  15   15:78575140   \n",
       "9   genic_downstream_transcript_variant,intron_var...   9    9:22098620   \n",
       "10                                     intron_variant   6   6:160591981   \n",
       "11  intron_variant,genic_downstream_transcript_var...   9    9:22096056   \n",
       "12                                                     18   18:22702893   \n",
       "13  3_prime_UTR_variant,genic_downstream_transcrip...  11  11:116778201   \n",
       "14   intron_variant,genic_upstream_transcript_variant   9   9:121660124   \n",
       "15                                     intron_variant  10   10:94138924   \n",
       "16  genic_downstream_transcript_variant,intron_var...   1   1:154452980   \n",
       "17                                     intron_variant  12   12:57133500   \n",
       "18  intron_variant,genic_downstream_transcript_var...   7    7:40801371   \n",
       "19                                3_prime_UTR_variant   2    2:20678345   \n",
       "20                                                      1   1:214232262   \n",
       "21                                                     11  11:102924877   \n",
       "22                                                     12   12:19777560   \n",
       "23                                     intron_variant   6   6:160589086   \n",
       "24  genic_downstream_transcript_variant,intron_var...  21   21:38447907   \n",
       "25                                                      8   8:125469835   \n",
       "26  3_prime_UTR_variant,genic_downstream_transcrip...  11  11:116778201   \n",
       "27                                     intron_variant   8    8:19962894   \n",
       "28                                     intron_variant  10   10:89243170   \n",
       "29                                3_prime_UTR_variant   1   1:109274968   \n",
       "30                                     intron_variant   6   6:160564494   \n",
       "31  genic_downstream_transcript_variant,intron_var...  10   10:94310118   \n",
       "32  coding_sequence_variant,genic_downstream_trans...  19   19:43648948   \n",
       "33  intron_variant,upstream_transcript_variant,gen...  20   20:45942678   \n",
       "34                                                     18   18:22523104   \n",
       "35                                                     16   16:56956804   \n",
       "36                                                     20   20:45927137   \n",
       "37                                                      2   2:113083453   \n",
       "38                                     intron_variant  17   17:59781849   \n",
       "39                                                     11  11:102877965   \n",
       "40                                     intron_variant  11  11:130410772   \n",
       "41  upstream_transcript_variant,coding_sequence_va...   1    1:55039974   \n",
       "42                                                      1     1:9378116   \n",
       "43           coding_sequence_variant,missense_variant  19   19:40611394   \n",
       "44           missense_variant,coding_sequence_variant  19   19:44908684   \n",
       "45  genic_downstream_transcript_variant,intron_var...   7    7:40391533   \n",
       "46   intron_variant,genic_upstream_transcript_variant  10   10:62646598   \n",
       "47  genic_downstream_transcript_variant,intron_var...   1   1:172364308   \n",
       "48                      non_coding_transcript_variant  18   18:22613705   \n",
       "49                                     intron_variant  11  11:110358805   \n",
       "\n",
       "   global_population allele  \\\n",
       "0                         D   \n",
       "1                         R   \n",
       "2                         Y   \n",
       "3                         K   \n",
       "4                         D   \n",
       "5                         Y   \n",
       "6                         Y   \n",
       "7                         D   \n",
       "8                         M   \n",
       "9                         R   \n",
       "10                        Y   \n",
       "11                        N   \n",
       "12                        N   \n",
       "13                        S   \n",
       "14                        R   \n",
       "15                        R   \n",
       "16                        V   \n",
       "17                        B   \n",
       "18                        B   \n",
       "19                        R   \n",
       "20                        H   \n",
       "21                        R   \n",
       "22                        R   \n",
       "23                        R   \n",
       "24                        H   \n",
       "25                        R   \n",
       "26                        S   \n",
       "27                        R   \n",
       "28                        B   \n",
       "29                        K   \n",
       "30                        R   \n",
       "31                        D   \n",
       "32                        R   \n",
       "33                        D   \n",
       "34                        Y   \n",
       "35                        M   \n",
       "36                        N   \n",
       "37                        R   \n",
       "38                        R   \n",
       "39                        R   \n",
       "40                        D   \n",
       "41                        D   \n",
       "42                        N   \n",
       "43                        R   \n",
       "44                        Y   \n",
       "45                        V   \n",
       "46                        S   \n",
       "47                        R   \n",
       "48                        Y   \n",
       "49                        N   \n",
       "\n",
       "                                clinical_significance           acc  \n",
       "0                                         risk-factor  NC_000009.12  \n",
       "1                                              benign  NC_000006.12  \n",
       "2                                              benign  NC_000020.11  \n",
       "3                                         association  NC_000001.11  \n",
       "4                                                      NC_000013.11  \n",
       "5   pathogenic,risk-factor,association,conflicting...  NC_000019.10  \n",
       "6                                                      NC_000019.10  \n",
       "7                                                      NC_000011.10  \n",
       "8                                                      NC_000015.10  \n",
       "9                                                      NC_000009.12  \n",
       "10                                                     NC_000006.12  \n",
       "11                                        risk-factor  NC_000009.12  \n",
       "12                                                     NC_000018.10  \n",
       "13                                                     NC_000011.10  \n",
       "14                                                     NC_000009.12  \n",
       "15                                                     NC_000010.11  \n",
       "16                                                     NC_000001.11  \n",
       "17                                                     NC_000012.12  \n",
       "18                                                     NC_000007.14  \n",
       "19                                                     NC_000002.12  \n",
       "20                                                     NC_000001.11  \n",
       "21                                                     NC_000011.10  \n",
       "22                                                     NC_000012.12  \n",
       "23                                             benign  NC_000006.12  \n",
       "24                                                      NC_000021.9  \n",
       "25                                                     NC_000008.11  \n",
       "26                                                     NC_000011.10  \n",
       "27                                                     NC_000008.11  \n",
       "28                                                     NC_000010.11  \n",
       "29                                        association  NC_000001.11  \n",
       "30                                                     NC_000006.12  \n",
       "31                                                     NC_000010.11  \n",
       "32                                                     NC_000019.10  \n",
       "33                                                     NC_000020.11  \n",
       "34                                                     NC_000018.10  \n",
       "35                                                     NC_000016.10  \n",
       "36                                                     NC_000020.11  \n",
       "37                                                     NC_000002.12  \n",
       "38                                                     NC_000017.11  \n",
       "39                                                     NC_000011.10  \n",
       "40                                                     NC_000011.10  \n",
       "41                                                     NC_000001.11  \n",
       "42                                                     NC_000001.11  \n",
       "43                        benign,benign-likely-benign  NC_000019.10  \n",
       "44  pathogenic,risk-factor,association,conflicting...  NC_000019.10  \n",
       "45                                                     NC_000007.14  \n",
       "46                                                     NC_000010.11  \n",
       "47                                                     NC_000001.11  \n",
       "48                                                     NC_000018.10  \n",
       "49                                                     NC_000011.10  "
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_snp_data = []\n",
    "\n",
    "# Open a text file in write mode to log skipped rows\n",
    "with open(\"skipped_rows_log.txt\", \"w\") as log_file:\n",
    "    log_file.write(\"Details of rows skipped due to None UID:\\n\\n\")\n",
    "\n",
    "    for index, row in UID_SNP.iterrows():\n",
    "        test_snp = row['UID']\n",
    "        \n",
    "        # Check if the UID is not None\n",
    "        if test_snp is None:\n",
    "            # Print to console and write to log file\n",
    "            log_message = f\"Skipping row {index} because UID is None. Row details:\\n{row}\\n\\n\"\n",
    "            print(log_message)\n",
    "            log_file.write(log_message)\n",
    "            continue  # Skip this iteration if UID is None\n",
    "\n",
    "        try: \n",
    "            # Retrieve SNP data from the SNP database\n",
    "            search_test = rentrez.entrez_summary(db='snp', id=test_snp)\n",
    "            \n",
    "            # Filter and safely retrieve desired fields\n",
    "            filtered_search_test_dict = {\n",
    "                key: str(search_test.rx2(key)[0]) if len(search_test.rx2(key)) > 0 else None\n",
    "                for key in search_test.names if key in desired_dbSNP_values\n",
    "            }\n",
    "            \n",
    "            # Order the data to match the desired column order\n",
    "            ordered_data = {key: filtered_search_test_dict.get(key, None) for key in desired_dbSNP_values}\n",
    "            \n",
    "            # Append ordered data to the results list\n",
    "            all_snp_data.append(ordered_data)\n",
    "\n",
    "        except Exception as e:\n",
    "            error_message = f\"Error retrieving data for UID {test_snp}: {e}\\n\"\n",
    "            print(error_message)\n",
    "            log_file.write(error_message)\n",
    "\n",
    "# Convert results to DataFrame\n",
    "final_df = pd.DataFrame(all_snp_data)\n",
    "\n",
    "# Display the first 10 rows\n",
    "final_df.head(50)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "f6520708-7ff6-44ce-8cf3-55a4169a0a38",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Now let's make a copy of our RAW results:\n",
    "final_df.to_csv(os.path.join(output_path, 'compiled_SNPs_RENTREZ_RAW.csv'), index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "281f19ee-bb6b-49ab-9700-8092436a009b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>uid</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>chr</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>global_population</th>\n",
       "      <th>allele</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>acc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[1] \"CDKN2B-AS1\"\\n</td>\n",
       "      <td>4977574</td>\n",
       "      <td>4977574</td>\n",
       "      <td>snv</td>\n",
       "      <td>genic_downstream_transcript_variant,intron_var...</td>\n",
       "      <td>9</td>\n",
       "      <td>9:22098575</td>\n",
       "      <td></td>\n",
       "      <td>D</td>\n",
       "      <td>risk-factor</td>\n",
       "      <td>NC_000009.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[1] \"LPA\"\\n</td>\n",
       "      <td>10455872</td>\n",
       "      <td>10455872</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>6</td>\n",
       "      <td>6:160589086</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td>benign</td>\n",
       "      <td>NC_000006.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[1] \"ZNF335\"\\n</td>\n",
       "      <td>3827066</td>\n",
       "      <td>3827066</td>\n",
       "      <td>snv</td>\n",
       "      <td>downstream_transcript_variant,intron_variant,g...</td>\n",
       "      <td>20</td>\n",
       "      <td>20:45957384</td>\n",
       "      <td></td>\n",
       "      <td>Y</td>\n",
       "      <td>benign</td>\n",
       "      <td>NC_000020.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[1] \"CELSR2\"\\n</td>\n",
       "      <td>12740374</td>\n",
       "      <td>12740374</td>\n",
       "      <td>snv</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>1</td>\n",
       "      <td>1:109274968</td>\n",
       "      <td></td>\n",
       "      <td>K</td>\n",
       "      <td>association</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>None</td>\n",
       "      <td>7994761</td>\n",
       "      <td>7994761</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>13</td>\n",
       "      <td>13:22297307</td>\n",
       "      <td></td>\n",
       "      <td>D</td>\n",
       "      <td></td>\n",
       "      <td>NC_000013.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>[1] \"C5orf67\"\\n</td>\n",
       "      <td>464605</td>\n",
       "      <td>464605</td>\n",
       "      <td>snv</td>\n",
       "      <td>non_coding_transcript_variant</td>\n",
       "      <td>5</td>\n",
       "      <td>5:56511543</td>\n",
       "      <td></td>\n",
       "      <td>H</td>\n",
       "      <td></td>\n",
       "      <td>NC_000005.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>[1] \"ANKRD44\"     \"ANKRD44-IT1\"\\n</td>\n",
       "      <td>919433</td>\n",
       "      <td>919433</td>\n",
       "      <td>snv</td>\n",
       "      <td>genic_upstream_transcript_variant,intron_variant</td>\n",
       "      <td>2</td>\n",
       "      <td>2:197301841</td>\n",
       "      <td></td>\n",
       "      <td>V</td>\n",
       "      <td></td>\n",
       "      <td>NC_000002.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>None</td>\n",
       "      <td>530948381</td>\n",
       "      <td>530948381</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>22</td>\n",
       "      <td>22:25507825</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000022.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>[1] \"MON1A\"\\n</td>\n",
       "      <td>142344547</td>\n",
       "      <td>142344547</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>3</td>\n",
       "      <td>3:49918603</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000003.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>None</td>\n",
       "      <td>141830649</td>\n",
       "      <td>141830649</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>14</td>\n",
       "      <td>14:20526739</td>\n",
       "      <td></td>\n",
       "      <td>W</td>\n",
       "      <td></td>\n",
       "      <td>NC_000014.9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>147 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 genes     snp_id        uid snp_class  \\\n",
       "0                   [1] \"CDKN2B-AS1\"\\n    4977574    4977574       snv   \n",
       "1                          [1] \"LPA\"\\n   10455872   10455872       snv   \n",
       "2                       [1] \"ZNF335\"\\n    3827066    3827066       snv   \n",
       "3                       [1] \"CELSR2\"\\n   12740374   12740374       snv   \n",
       "4                                 None    7994761    7994761       snv   \n",
       "..                                 ...        ...        ...       ...   \n",
       "145                    [1] \"C5orf67\"\\n     464605     464605       snv   \n",
       "146  [1] \"ANKRD44\"     \"ANKRD44-IT1\"\\n     919433     919433       snv   \n",
       "147                               None  530948381  530948381       snv   \n",
       "148                      [1] \"MON1A\"\\n  142344547  142344547       snv   \n",
       "149                               None  141830649  141830649       snv   \n",
       "\n",
       "                                             fxn_class chr       chrpos  \\\n",
       "0    genic_downstream_transcript_variant,intron_var...   9   9:22098575   \n",
       "1                                       intron_variant   6  6:160589086   \n",
       "2    downstream_transcript_variant,intron_variant,g...  20  20:45957384   \n",
       "3                                  3_prime_UTR_variant   1  1:109274968   \n",
       "4                                                       13  13:22297307   \n",
       "..                                                 ...  ..          ...   \n",
       "145                      non_coding_transcript_variant   5   5:56511543   \n",
       "146   genic_upstream_transcript_variant,intron_variant   2  2:197301841   \n",
       "147                                                     22  22:25507825   \n",
       "148                                     intron_variant   3   3:49918603   \n",
       "149                                                     14  14:20526739   \n",
       "\n",
       "    global_population allele clinical_significance           acc  \n",
       "0                          D           risk-factor  NC_000009.12  \n",
       "1                          R                benign  NC_000006.12  \n",
       "2                          Y                benign  NC_000020.11  \n",
       "3                          K           association  NC_000001.11  \n",
       "4                          D                        NC_000013.11  \n",
       "..                ...    ...                   ...           ...  \n",
       "145                        H                        NC_000005.10  \n",
       "146                        V                        NC_000002.12  \n",
       "147                        R                        NC_000022.11  \n",
       "148                        R                        NC_000003.12  \n",
       "149                        W                         NC_000014.9  \n",
       "\n",
       "[147 rows x 11 columns]"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### And let's make it beautiful by rearranging and sorting each column:\n",
    "#### First, let's delete all variants that are not SNVs:\n",
    "filtered_final_df = final_df.loc[final_df['snp_class'] == 'snv'] \n",
    "filtered_final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "9e3773e8-af4e-4ae4-b3a0-61bc87730794",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>uid</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>chr</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>global_population</th>\n",
       "      <th>allele</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>acc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDKN2B-AS1</td>\n",
       "      <td>4977574</td>\n",
       "      <td>4977574</td>\n",
       "      <td>snv</td>\n",
       "      <td>genic_downstream_transcript_variant,intron_var...</td>\n",
       "      <td>9</td>\n",
       "      <td>9:22098575</td>\n",
       "      <td></td>\n",
       "      <td>D</td>\n",
       "      <td>risk-factor</td>\n",
       "      <td>NC_000009.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LPA</td>\n",
       "      <td>10455872</td>\n",
       "      <td>10455872</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>6</td>\n",
       "      <td>6:160589086</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td>benign</td>\n",
       "      <td>NC_000006.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZNF335</td>\n",
       "      <td>3827066</td>\n",
       "      <td>3827066</td>\n",
       "      <td>snv</td>\n",
       "      <td>downstream_transcript_variant,intron_variant,g...</td>\n",
       "      <td>20</td>\n",
       "      <td>20:45957384</td>\n",
       "      <td></td>\n",
       "      <td>Y</td>\n",
       "      <td>benign</td>\n",
       "      <td>NC_000020.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CELSR2</td>\n",
       "      <td>12740374</td>\n",
       "      <td>12740374</td>\n",
       "      <td>snv</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>1</td>\n",
       "      <td>1:109274968</td>\n",
       "      <td></td>\n",
       "      <td>K</td>\n",
       "      <td>association</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>None</td>\n",
       "      <td>7994761</td>\n",
       "      <td>7994761</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>13</td>\n",
       "      <td>13:22297307</td>\n",
       "      <td></td>\n",
       "      <td>D</td>\n",
       "      <td></td>\n",
       "      <td>NC_000013.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>C5orf67</td>\n",
       "      <td>464605</td>\n",
       "      <td>464605</td>\n",
       "      <td>snv</td>\n",
       "      <td>non_coding_transcript_variant</td>\n",
       "      <td>5</td>\n",
       "      <td>5:56511543</td>\n",
       "      <td></td>\n",
       "      <td>H</td>\n",
       "      <td></td>\n",
       "      <td>NC_000005.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>ANKRD44     ANKRD44-IT1</td>\n",
       "      <td>919433</td>\n",
       "      <td>919433</td>\n",
       "      <td>snv</td>\n",
       "      <td>genic_upstream_transcript_variant,intron_variant</td>\n",
       "      <td>2</td>\n",
       "      <td>2:197301841</td>\n",
       "      <td></td>\n",
       "      <td>V</td>\n",
       "      <td></td>\n",
       "      <td>NC_000002.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>None</td>\n",
       "      <td>530948381</td>\n",
       "      <td>530948381</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>22</td>\n",
       "      <td>22:25507825</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000022.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>MON1A</td>\n",
       "      <td>142344547</td>\n",
       "      <td>142344547</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>3</td>\n",
       "      <td>3:49918603</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000003.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>None</td>\n",
       "      <td>141830649</td>\n",
       "      <td>141830649</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>14</td>\n",
       "      <td>14:20526739</td>\n",
       "      <td></td>\n",
       "      <td>W</td>\n",
       "      <td></td>\n",
       "      <td>NC_000014.9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>147 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       genes     snp_id        uid snp_class  \\\n",
       "0                 CDKN2B-AS1    4977574    4977574       snv   \n",
       "1                        LPA   10455872   10455872       snv   \n",
       "2                     ZNF335    3827066    3827066       snv   \n",
       "3                     CELSR2   12740374   12740374       snv   \n",
       "4                       None    7994761    7994761       snv   \n",
       "..                       ...        ...        ...       ...   \n",
       "145                  C5orf67     464605     464605       snv   \n",
       "146  ANKRD44     ANKRD44-IT1     919433     919433       snv   \n",
       "147                     None  530948381  530948381       snv   \n",
       "148                    MON1A  142344547  142344547       snv   \n",
       "149                     None  141830649  141830649       snv   \n",
       "\n",
       "                                             fxn_class chr       chrpos  \\\n",
       "0    genic_downstream_transcript_variant,intron_var...   9   9:22098575   \n",
       "1                                       intron_variant   6  6:160589086   \n",
       "2    downstream_transcript_variant,intron_variant,g...  20  20:45957384   \n",
       "3                                  3_prime_UTR_variant   1  1:109274968   \n",
       "4                                                       13  13:22297307   \n",
       "..                                                 ...  ..          ...   \n",
       "145                      non_coding_transcript_variant   5   5:56511543   \n",
       "146   genic_upstream_transcript_variant,intron_variant   2  2:197301841   \n",
       "147                                                     22  22:25507825   \n",
       "148                                     intron_variant   3   3:49918603   \n",
       "149                                                     14  14:20526739   \n",
       "\n",
       "    global_population allele clinical_significance           acc  \n",
       "0                          D           risk-factor  NC_000009.12  \n",
       "1                          R                benign  NC_000006.12  \n",
       "2                          Y                benign  NC_000020.11  \n",
       "3                          K           association  NC_000001.11  \n",
       "4                          D                        NC_000013.11  \n",
       "..                ...    ...                   ...           ...  \n",
       "145                        H                        NC_000005.10  \n",
       "146                        V                        NC_000002.12  \n",
       "147                        R                        NC_000022.11  \n",
       "148                        R                        NC_000003.12  \n",
       "149                        W                         NC_000014.9  \n",
       "\n",
       "[147 rows x 11 columns]"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Now let's clean up la mugre of each cell in the genes column\n",
    "\n",
    "# Define a function to clean up each cell\n",
    "def clean_cell(value):\n",
    "    if isinstance(value, str):\n",
    "        # Remove `[1]`, quotation marks, and newline characters\n",
    "        cleaned_value = re.sub(r'^\\[\\d+\\]\\s*\"', '', value).replace('\"', '').strip()\n",
    "        return cleaned_value\n",
    "    return value\n",
    "\n",
    "# Apply the cleaning function to the 'genes' column using .loc\n",
    "filtered_final_df.loc[:, 'genes'] = filtered_final_df['genes'].apply(clean_cell)\n",
    "\n",
    "filtered_final_df.head(200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b71082ca-853c-4260-b7d2-0a9838ab33f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Now let's make reorder and rename the columns so they match our RENTREZ template. \n",
    "RENTREZ_rename = filtered_final_df.rename(columns={'genes': 'Gene ID (GENE_ID)', 'snp_id': 'dbSNP ID (RS)'})\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
