{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "191510b4-5c34-448d-ab60-e1beb0ca649a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Before we start, remember to activate your conda environment.\n",
    "## Write 'conda create --name myenv numpy pandas scipy matplotlib' into your terminal\n",
    "## Then activate the environment: 'conda activate myenv'\n",
    "## Remember to check that your Python interpreter is correct\n",
    "\n",
    "# First step: Import the libraries we're going to use (PANDAS, NUMPY, MATPLOTLIB)\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib as plt\n",
    "import os "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f1e8d419-84f5-41e0-980e-26e9a9850130",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DOI</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SNP ID</th>\n",
       "      <th>P-VALUE (GWAS)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [DOI, GENE, SNP ID, P-VALUE (GWAS)]\n",
       "Index: []"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Second step: We're going to read our Resources.csv and test to see if we can see it.\n",
    "SNP_file = pd.read_csv('Resources/Template.csv')    #You should always open the Data_Extraction_Code \n",
    "                                                       #folder to have a relative path similar to this one\n",
    "SNP_file.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7cbf96bf-ed42-4524-a648-a2069bb6b777",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DOI</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SNP ID</th>\n",
       "      <th>P-VALUE (GWAS)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [DOI, GENE, SNP ID, P-VALUE (GWAS)]\n",
       "Index: []"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Third step: Now let's create a second dataframe (to keep from editing the original)\n",
    "SNP_editable = SNP_file\n",
    "SNP_editable.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f8e5a280-4c04-4991-8c73-de47397fbea7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DOI</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SNP ID</th>\n",
       "      <th>P-VALUE (GWAS)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10.3340/jkns.2022.0026</td>\n",
       "      <td>BOLL</td>\n",
       "      <td>rs700651</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10.3340/jkns.2022.0026</td>\n",
       "      <td>[U]EDNRA</td>\n",
       "      <td>rs6841581</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10.3340/jkns.2022.0026</td>\n",
       "      <td>PTCH1</td>\n",
       "      <td>rs1105980</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10.3340/jkns.2022.0026</td>\n",
       "      <td>CCDC3</td>\n",
       "      <td>rs12412014</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10.3340/jkns.2022.0026</td>\n",
       "      <td>LINC00457</td>\n",
       "      <td>rs1536847</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      DOI       GENE      SNP ID  P-VALUE (GWAS)\n",
       "0  10.3340/jkns.2022.0026       BOLL    rs700651             NaN\n",
       "1  10.3340/jkns.2022.0026   [U]EDNRA   rs6841581             NaN\n",
       "2  10.3340/jkns.2022.0026      PTCH1   rs1105980             NaN\n",
       "3  10.3340/jkns.2022.0026      CCDC3  rs12412014             NaN\n",
       "4  10.3340/jkns.2022.0026  LINC00457   rs1536847             NaN"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Fourth step: Let's merge the dataframe and use SNP ID as our merge.\n",
    "# Let's read our second .csv file\n",
    "Our_SNP_List = pd.read_csv('Resources/Input_Files/IA_test.csv')\n",
    "Our_SNP_List.head()\n",
    "\n",
    "# And..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "551e8452-ecf7-4d5f-9b8f-db04c2b1db8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MERGE! \n",
    "merge_data = pd.merge(Our_SNP_List, SNP_file, on=[\"SNP ID\", \"GENE\", \"DOI\", \"P-VALUE (GWAS)\"], how='left')\n",
    "merge_data.head(10)\n",
    "\n",
    "# Now we have our own database set up as a dataframe and filled it with our columns. \n",
    "# If you merged two databases, great! If you did not, boo! Let's output the merged_data (if you did not merge two databases,\n",
    "# then ignore the output.\n",
    "output_path = 'Resources/OutputFromExtractionCode'\n",
    "merge_data.to_csv(os.path.join(output_path, 'Merged_data_1.csv'), index=False)\n",
    "\n",
    "\n",
    "#merge_data.to_csv('Merged_data_1.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "dd82b8c2-e8c2-4e49-902e-0409cd30961d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>SNP ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BOLL</td>\n",
       "      <td>rs700651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[U]EDNRA</td>\n",
       "      <td>rs6841581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PTCH1</td>\n",
       "      <td>rs1105980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CCDC3</td>\n",
       "      <td>rs12412014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LINC00457</td>\n",
       "      <td>rs1536847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>C5orf60</td>\n",
       "      <td>rs62405726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>METTL4</td>\n",
       "      <td>rs549315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>RGPD4</td>\n",
       "      <td>rs700855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>RGPD4</td>\n",
       "      <td>rs328025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MALL</td>\n",
       "      <td>rs117802391</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        GENE       SNP ID\n",
       "0       BOLL     rs700651\n",
       "1   [U]EDNRA    rs6841581\n",
       "2      PTCH1    rs1105980\n",
       "3      CCDC3   rs12412014\n",
       "4  LINC00457    rs1536847\n",
       "5    C5orf60   rs62405726\n",
       "6     METTL4     rs549315\n",
       "7      RGPD4     rs700855\n",
       "8      RGPD4     rs328025\n",
       "9       MALL  rs117802391"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Now let's drop all the data we collected and just stay with the two things that matter most to us: \n",
    "# the SNP ID and the Gene:\n",
    "SNP_ID_And_Gene_From_Merged_Data = merge_data[['GENE', 'SNP ID']]\n",
    "SNP_ID_And_Gene_From_Merged_Data.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "cfd151fe-b8d4-468b-a7b2-6e8a1a9687c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "### We'll be using this for our RENTREZ module in R'. Let's output it twice (one for keeps and one we can modify safely)\n",
    "SNP_ID_And_Gene_From_Merged_Data.to_csv(os.path.join(output_path, 'Original_Gene_And_dbSNP_ID_Merged.csv'), index=False)\n",
    "SNP_ID_And_Gene_From_Merged_Data.to_csv(os.path.join(output_path, 'Editable_Gene_And_dbSNP_ID_Merged.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "3f1d9c91-ad09-4fc8-9a3f-e61027e45130",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>SNP ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>8q11</td>\n",
       "      <td>rs10958409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>9p</td>\n",
       "      <td>rs2891168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>9p21</td>\n",
       "      <td>rs1333040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>9p21</td>\n",
       "      <td>rs10757278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>ADAMTS12</td>\n",
       "      <td>rs1364044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>ADAMTS13</td>\n",
       "      <td>rs2285489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>ADAMTS13</td>\n",
       "      <td>rs4962153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>ADAMTS13</td>\n",
       "      <td>rs739469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>ADAMTS13</td>\n",
       "      <td>rs2301612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>ADAMTS15</td>\n",
       "      <td>rs767345140</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         GENE       SNP ID\n",
       "223      8q11   rs10958409\n",
       "226        9p    rs2891168\n",
       "224     9p21     rs1333040\n",
       "225     9p21    rs10757278\n",
       "127  ADAMTS12    rs1364044\n",
       "131  ADAMTS13    rs2285489\n",
       "129  ADAMTS13    rs4962153\n",
       "128  ADAMTS13     rs739469\n",
       "130  ADAMTS13    rs2301612\n",
       "266  ADAMTS15  rs767345140"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### We can already do something pretty cool, which is determine which gene had more pathogenic SNPs.\n",
    "### We can sort it:\n",
    "SNP_Sorted_By_Gene =  SNP_ID_And_Gene_From_Merged_Data.sort_values(by='GENE')\n",
    "SNP_Sorted_By_Gene.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "8d3a0793-bc82-49fd-ad95-c142054d2c3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>Reported SNPs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VCAN</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MMP13</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MMP2</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PCNT</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RNF213</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MMP24</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>LIMK1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MMP26</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>COL1A2</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CDKN2B-AS1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ADAMTS13</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ELN</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MMP15</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>IL12B</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>THSD1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>MMP25</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>MMP16</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MMP14</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>TNF-α</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          GENE  Reported SNPs\n",
       "0         VCAN             15\n",
       "1        MMP13             11\n",
       "2         MMP2             11\n",
       "3         PCNT              9\n",
       "4       RNF213              6\n",
       "5        MMP24              6\n",
       "6        LIMK1              6\n",
       "7        MMP26              6\n",
       "8       COL1A2              5\n",
       "9   CDKN2B-AS1              5\n",
       "10    ADAMTS13              4\n",
       "11         ELN              4\n",
       "12       MMP15              3\n",
       "13         ENG              3\n",
       "14       IL12B              3\n",
       "15       THSD1              3\n",
       "16       MMP25              3\n",
       "17       MMP16              3\n",
       "18       MMP14              2\n",
       "19       TNF-α              2"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### And get the unique count for each gene:\n",
    "Unique_Genes = SNP_Sorted_By_Gene['GENE'].value_counts().reset_index()\n",
    "Unique_Genes.columns = ['GENE', 'Reported SNPs']\n",
    "Unique_Genes.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "493db22e-ff98-465f-b2ef-966c4fa59d04",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>SNP ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8q11</td>\n",
       "      <td>[rs10958409]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9p</td>\n",
       "      <td>[rs2891168]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9p21</td>\n",
       "      <td>[rs1333040, rs10757278]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ADAMTS12</td>\n",
       "      <td>[rs1364044]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ADAMTS13</td>\n",
       "      <td>[rs2285489, rs4962153, rs739469, rs2301612]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ADAMTS15</td>\n",
       "      <td>[rs767345140]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ADAMTS2</td>\n",
       "      <td>[rs11750568]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AL359922.1</td>\n",
       "      <td>[rs4977574]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ANGPTL6</td>\n",
       "      <td>[rs147149731, rs559282550]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ARHGAP32</td>\n",
       "      <td>[rs371331393]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ARHGEF11</td>\n",
       "      <td>[rs7550260, rs7550260]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ARHGEF17</td>\n",
       "      <td>[rs2298808]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ATP11 A</td>\n",
       "      <td>[rs2297801]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ATP2B1</td>\n",
       "      <td>[rs11105337]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>BET1L</td>\n",
       "      <td>[rs2280543]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>BOLL</td>\n",
       "      <td>[rs700651]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>C12orf5</td>\n",
       "      <td>[rs4766232]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>C12orf75</td>\n",
       "      <td>[rs2374513]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>C5orf60</td>\n",
       "      <td>[rs62405726]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>C9orf92</td>\n",
       "      <td>[rs12350582]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          GENE                                       SNP ID\n",
       "0         8q11                                 [rs10958409]\n",
       "1           9p                                  [rs2891168]\n",
       "2        9p21                       [rs1333040, rs10757278]\n",
       "3     ADAMTS12                                  [rs1364044]\n",
       "4     ADAMTS13  [rs2285489, rs4962153, rs739469, rs2301612]\n",
       "5     ADAMTS15                                [rs767345140]\n",
       "6      ADAMTS2                                 [rs11750568]\n",
       "7   AL359922.1                                  [rs4977574]\n",
       "8      ANGPTL6                   [rs147149731, rs559282550]\n",
       "9    ARHGAP32                                 [rs371331393]\n",
       "10    ARHGEF11                       [rs7550260, rs7550260]\n",
       "11    ARHGEF17                                  [rs2298808]\n",
       "12     ATP11 A                                  [rs2297801]\n",
       "13      ATP2B1                                 [rs11105337]\n",
       "14       BET1L                                  [rs2280543]\n",
       "15        BOLL                                   [rs700651]\n",
       "16     C12orf5                                  [rs4766232]\n",
       "17    C12orf75                                  [rs2374513]\n",
       "18     C5orf60                                 [rs62405726]\n",
       "19    C9orf92                                  [rs12350582]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### And we should group them by gene so we can see which SNPs are actually related:\n",
    "grouped_SNPs = SNP_Sorted_By_Gene.groupby('GENE')['SNP ID'].apply(list).reset_index()\n",
    "grouped_SNPs.head(20)\n",
    "\n",
    "# groupby('category') groups the data by the 'category' column.\n",
    "# apply(list) applies the list function to each group, resulting in a list of strings for each unique category.\n",
    "# reset_index() is used to turn the result back into a DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "f26c4003-dff5-4b2d-92bb-48e238852392",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Now let's output them as CSV files:\n",
    "SNP_Sorted_By_Gene.to_csv(os.path.join(output_path, 'SNP_Sorted_By_Gene.csv'), index=False)\n",
    "Unique_Genes.to_csv(os.path.join(output_path, 'Unique_Genes.csv'), index=False)\n",
    "grouped_SNPs.to_csv(os.path.join(output_path, 'grouped_SNPs.csv'), index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "d6033b25-548d-411d-9dbf-8aaa7a7eef34",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Now let's turn to our python R file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c79a7766-782e-426c-a8f8-be7dc2972d10",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "567f7951-42ed-4e02-a62e-6882f4c2104e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ae2d098-9349-4a38-a6fd-34ef1afd361a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7083f8b3-2e5e-4faf-b406-b66850232af4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f448431-8a6f-4633-8ab5-8c9643a2fc37",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3dabf6a-1a01-4818-ba9d-b77ba4b25644",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
