{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9d9b8fd4-0498-44fc-9d30-bd878bb7cfa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Now since this is a new file we need to import other modules (YAY!) including our R' emulator\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects.packages import importr\n",
    "from rpy2.robjects import pandas2ri\n",
    "\n",
    "### and let's not forget the usual gang:\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib as plt\n",
    "import re\n",
    "import os "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9b8c2979-866d-405e-a5f1-fd872a08cdb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Before we get started in an R' environment. We need to talk about something important.\n",
    "#### Rentrez does not use SNP ID to search. It uses UID. As in, they don't use 'rs2000' \n",
    "#### but rather '2000'. So we need to use our previous output and split the function.\n",
    "Output_GeneSNP_file = pd.read_csv('Resources/OutputFromExtractionCode/Editable_Gene_And_dbSNP_ID_Merged.csv')\n",
    "Output_GeneSNP_file.head(10)\n",
    "\n",
    "output_path = 'Resources/OutputFromExtractionCode'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "751ff5a8-be40-41ee-b44d-ece195aa51da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>SNP ID</th>\n",
       "      <th>UID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BOLL</td>\n",
       "      <td>rs700651</td>\n",
       "      <td>700651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[U]EDNRA</td>\n",
       "      <td>rs6841581</td>\n",
       "      <td>6841581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PTCH1</td>\n",
       "      <td>rs1105980</td>\n",
       "      <td>1105980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CCDC3</td>\n",
       "      <td>rs12412014</td>\n",
       "      <td>12412014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LINC00457</td>\n",
       "      <td>rs1536847</td>\n",
       "      <td>1536847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>C5orf60</td>\n",
       "      <td>rs62405726</td>\n",
       "      <td>62405726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>METTL4</td>\n",
       "      <td>rs549315</td>\n",
       "      <td>549315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>RGPD4</td>\n",
       "      <td>rs700855</td>\n",
       "      <td>700855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>RGPD4</td>\n",
       "      <td>rs328025</td>\n",
       "      <td>328025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MALL</td>\n",
       "      <td>rs117802391</td>\n",
       "      <td>117802391</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        GENE       SNP ID        UID\n",
       "0       BOLL     rs700651     700651\n",
       "1   [U]EDNRA    rs6841581    6841581\n",
       "2      PTCH1    rs1105980    1105980\n",
       "3      CCDC3   rs12412014   12412014\n",
       "4  LINC00457    rs1536847    1536847\n",
       "5    C5orf60   rs62405726   62405726\n",
       "6     METTL4     rs549315     549315\n",
       "7      RGPD4     rs700855     700855\n",
       "8      RGPD4     rs328025     328025\n",
       "9       MALL  rs117802391  117802391"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Now we split it. \n",
    "Split_dbSNP_ID = Output_GeneSNP_file\n",
    "Split_dbSNP_ID[['rs', 'UID']] = Output_GeneSNP_file['SNP ID'].str.split('rs', expand = True)\n",
    "Split_dbSNP_ID.head(10)\n",
    "\n",
    "#And delete the rs column:\n",
    "UID_SNP = Split_dbSNP_ID.drop(columns=['rs'])\n",
    "UID_SNP.head(10)\n",
    "\n",
    "## Now we have the UID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d59ca7d3-ffa1-4e16-b92b-8d440f53957c",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Let's import RENTREZ to our python environment\n",
    "rentrez = importr('rentrez')\n",
    "\n",
    "### You'll need to provide an email address to use Entrez tools. BTW Hot tip: \n",
    "### If you have multiple lines, using triple quotes is cleaner and easier to read:\n",
    "\n",
    "robjects.r('''\n",
    "library(rentrez)\n",
    "entrez_email <- \"kenji.ponsm@gmail.com\"\n",
    "''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a111e1f2-502d-44da-9872-d5a970fe2d4e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Searchable fields for database 'snp'\n",
      "  ALL \t All terms from all searchable fields \n",
      "  UID \t Unique number assigned to publication \n",
      "  FILT \t Limits the records \n",
      "  RS \t Clustered SNP ID (rs) \n",
      "  CHR \t chromosomes \n",
      "  GENE \t locus link symbol \n",
      "  HAN \t Submitter Handle \n",
      "  ACCN \t nucleotide accessions \n",
      "  GENE_ID \t Gene ID \n",
      "  FXN \t dbSNP Functional consequence class \n",
      "  GTYP \t Genotype info \n",
      "  SS \t Submitter ID \n",
      "  VARI \t Allele \n",
      "  SCLS \t SNP class \n",
      "  CPOS \t Chromosome base position \n",
      "  WORD \t Free text associated with record \n",
      "  SIDX \t SNP Index \n",
      "  CLIN \t Variations with clinical effects or significances \n",
      "  GMAF \t Minor Allele Frequency derived from global population (ie. 1000G) \n",
      "  VALI \t Validation status \n",
      "  CPOS_GRCH37 \t Chromosome base position on previous assembly version \n",
      "  ORGN \t Organism \n",
      "  ALFA_EUR \t ALFA European Minor Allele Frequency \n",
      "  ALFA_AFR \t ALFA African population Minor Allele Frequency \n",
      "  ALFA_ASN \t ALFA Asian population Minor Allele Frequency \n",
      "  ALFA_LAC \t ALFA Latin American 1 population Minor Allele Frequency \n",
      "  ALFA_LEN \t ALFA Latin American 2 population Minor Allele Frequency \n",
      "  ALFA_SAS \t ALFA South Asian population Minor Allele Frequency \n",
      "  ALFA_OTR \t ALFA Other population Minor Allele Frequency \n",
      "\n"
     ]
    }
   ],
   "source": [
    "### What can we look for with entrez\n",
    "snp_database = robjects.r('entrez_db_searchable(\"snp\")')\n",
    "print(snp_database)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2b86d25f-7e00-496f-90dd-3ba9b8db7d8c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "700651\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>allele_origin</th>\n",
       "      <th>global_mafs</th>\n",
       "      <th>global_population</th>\n",
       "      <th>global_samplesize</th>\n",
       "      <th>suspected</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>genes</th>\n",
       "      <th>acc</th>\n",
       "      <th>...</th>\n",
       "      <th>allele</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>chrpos_prev_assm</th>\n",
       "      <th>text</th>\n",
       "      <th>snp_id_sort</th>\n",
       "      <th>clinical_sort</th>\n",
       "      <th>cited_sort</th>\n",
       "      <th>chrpos_sort</th>\n",
       "      <th>merged_sort</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>700651</td>\n",
       "      <td>700651</td>\n",
       "      <td></td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[1] \"BOLL\"\\n</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>...</td>\n",
       "      <td>R</td>\n",
       "      <td>snv</td>\n",
       "      <td>2:197766990</td>\n",
       "      <td>2:198631714</td>\n",
       "      <td></td>\n",
       "      <td>0000700651</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0197766990</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      uid  snp_id allele_origin  \\\n",
       "0  700651  700651                 \n",
       "\n",
       "                                         global_mafs global_population  \\\n",
       "0   [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...                     \n",
       "\n",
       "  global_samplesize suspected clinical_significance         genes  \\\n",
       "0                                                    [1] \"BOLL\"\\n   \n",
       "\n",
       "            acc  ... allele snp_class       chrpos chrpos_prev_assm text  \\\n",
       "0  NC_000002.12  ...      R       snv  2:197766990      2:198631714        \n",
       "\n",
       "  snp_id_sort clinical_sort cited_sort chrpos_sort merged_sort  \n",
       "0  0000700651             0             0197766990           0  \n",
       "\n",
       "[1 rows x 32 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Let's give it a test shall we? Let's say I want the information of rs429358 (GENE: APOE)\n",
    "print(UID_SNP.iloc[0,2])\n",
    "test_snp = UID_SNP.iloc[0,2]\n",
    "search_test = rentrez.entrez_summary(db='snp', id=test_snp)\n",
    "search_test_dict = {key: str(search_test.rx2(key) [0]) for key in search_test.names}\n",
    "\n",
    "search_test_df = pd.DataFrame([search_test_dict])\n",
    "\n",
    "search_test_df.head(10)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "42a5e8ad-01e1-4263-ac18-3c39726932e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [1] \"uid\"                   \"snp_id\"                \"allele_origin\"        \n",
      " [4] \"global_mafs\"           \"global_population\"     \"global_samplesize\"    \n",
      " [7] \"suspected\"             \"clinical_significance\" \"genes\"                \n",
      "[10] \"acc\"                   \"chr\"                   \"handle\"               \n",
      "[13] \"spdi\"                  \"fxn_class\"             \"validated\"            \n",
      "[16] \"docsum\"                \"tax_id\"                \"orig_build\"           \n",
      "[19] \"upd_build\"             \"createdate\"            \"updatedate\"           \n",
      "[22] \"ss\"                    \"allele\"                \"snp_class\"            \n",
      "[25] \"chrpos\"                \"chrpos_prev_assm\"      \"text\"                 \n",
      "[28] \"snp_id_sort\"           \"clinical_sort\"         \"cited_sort\"           \n",
      "[31] \"chrpos_sort\"           \"merged_sort\"          \n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(search_test.names)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c48732cd-c951-44bb-9104-227be736a212",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>global_mafs</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>genes</th>\n",
       "      <th>acc</th>\n",
       "      <th>chr</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>allele</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>chrpos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>700651</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td></td>\n",
       "      <td>[1] \"BOLL\"\\n</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>R</td>\n",
       "      <td>snv</td>\n",
       "      <td>2:197766990</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      uid                                        global_mafs  \\\n",
       "0  700651   [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...   \n",
       "\n",
       "  clinical_significance         genes           acc chr       fxn_class  \\\n",
       "0                        [1] \"BOLL\"\\n  NC_000002.12   2  intron_variant   \n",
       "\n",
       "  allele snp_class       chrpos  \n",
       "0      R       snv  2:197766990  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### And let's filter for the actual values we want to extract: \n",
    "desired_dbSNP_values = (['genes', 'uid', 'snp_class', \n",
    "                         'fxn_class', 'chr', 'chrpos', 'global_mafs', \n",
    "                         'allele','clinical_significance', 'acc'] \n",
    "                         )\n",
    "filtered_search_test_dict = {key: str(search_test.rx2(key)[0]) for key in search_test.names if key in desired_dbSNP_values}\n",
    "filtered_search_test = pd.DataFrame([filtered_search_test_dict])\n",
    "filtered_search_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fc3d7c7d-291c-4ee1-b728-9fd3dc5c9971",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>uid</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>chr</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>global_mafs</th>\n",
       "      <th>allele</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>acc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[1] \"BOLL\"\\n</td>\n",
       "      <td>700651</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>2</td>\n",
       "      <td>2:197766990</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000002.12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          genes     uid snp_class       fxn_class chr       chrpos  \\\n",
       "0  [1] \"BOLL\"\\n  700651       snv  intron_variant   2  2:197766990   \n",
       "\n",
       "                                         global_mafs allele  \\\n",
       "0   [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      R   \n",
       "\n",
       "  clinical_significance           acc  \n",
       "0                        NC_000002.12  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_and_ordered_search_test = filtered_search_test[[col for col in desired_dbSNP_values if col in filtered_search_test.columns]]\n",
    "filtered_and_ordered_search_test\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9827951a-19a9-4922-b977-d9bf27b7921c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>uid</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>chr</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>global_mafs</th>\n",
       "      <th>allele</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>acc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[1] \"BOLL\"\\n</td>\n",
       "      <td>700651</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>2</td>\n",
       "      <td>2:197766990</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000002.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[1] \"EDNRA\"\\n</td>\n",
       "      <td>6841581</td>\n",
       "      <td>snv</td>\n",
       "      <td>2KB_upstream_variant,upstream_transcript_variant</td>\n",
       "      <td>4</td>\n",
       "      <td>4:147480038</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...</td>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>NC_000004.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>None</td>\n",
       "      <td>1105980</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>9</td>\n",
       "      <td>9:95351353</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>S</td>\n",
       "      <td></td>\n",
       "      <td>NC_000009.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>None</td>\n",
       "      <td>12412014</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>10:12869725</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>B</td>\n",
       "      <td></td>\n",
       "      <td>NC_000010.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[1] \"LINC00457\" \"LINC02343\"\\n</td>\n",
       "      <td>1536847</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>13</td>\n",
       "      <td>13:34532838</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>D</td>\n",
       "      <td></td>\n",
       "      <td>NC_000013.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>[1] \"C5orf60\"\\n</td>\n",
       "      <td>62405726</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>5</td>\n",
       "      <td>5:179642467</td>\n",
       "      <td>[1] \"1000Genomes\" \"ExAC\"        \"GnomAD\"     ...</td>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>NC_000005.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>None</td>\n",
       "      <td>549315</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>18</td>\n",
       "      <td>18:2183055</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>V</td>\n",
       "      <td></td>\n",
       "      <td>NC_000018.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>[1] \"GACAT1\"\\n</td>\n",
       "      <td>700855</td>\n",
       "      <td>snv</td>\n",
       "      <td>upstream_transcript_variant,2KB_upstream_variant</td>\n",
       "      <td>2</td>\n",
       "      <td>2:107752238</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...</td>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>NC_000002.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>None</td>\n",
       "      <td>328025</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>2:107738589</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>D</td>\n",
       "      <td></td>\n",
       "      <td>NC_000002.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>[1] \"MALL\"\\n</td>\n",
       "      <td>117802391</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>2</td>\n",
       "      <td>2:110104507</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"GnomAD...</td>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>NC_000002.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>[1] \"LINC01978\" \"LINC01979\"\\n</td>\n",
       "      <td>57851800</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>17</td>\n",
       "      <td>17:79922572</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>V</td>\n",
       "      <td></td>\n",
       "      <td>NC_000017.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>[1] \"RREB1\"\\n</td>\n",
       "      <td>9505086</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>6</td>\n",
       "      <td>6:7231953</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>NC_000006.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>[1] \"DST\"\\n</td>\n",
       "      <td>117021265</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant,genic_upstream_transcript_variant</td>\n",
       "      <td>6</td>\n",
       "      <td>6:56763223</td>\n",
       "      <td>[1] \"1000Genomes\" \"GnomAD\"      \"KOREAN\"      ...</td>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>NC_000006.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>None</td>\n",
       "      <td>5005908</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>2:153147166</td>\n",
       "      <td>[1] \"1000Genomes\"    \"Estonian\"       \"GENOME...</td>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>NC_000002.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>[1] \"FOXP1\"\\n</td>\n",
       "      <td>878118</td>\n",
       "      <td>snv</td>\n",
       "      <td>genic_upstream_transcript_variant,genic_downst...</td>\n",
       "      <td>3</td>\n",
       "      <td>3:71197077</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>D</td>\n",
       "      <td></td>\n",
       "      <td>NC_000003.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>[1] \"LINC01344\"\\n</td>\n",
       "      <td>12033118</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>1</td>\n",
       "      <td>1:182260612</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>H</td>\n",
       "      <td></td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>[1] \"RBMS3\"\\n</td>\n",
       "      <td>1979271</td>\n",
       "      <td>snv</td>\n",
       "      <td>genic_upstream_transcript_variant,intron_variant</td>\n",
       "      <td>3</td>\n",
       "      <td>3:29565914</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>NC_000003.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>None</td>\n",
       "      <td>189432614</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>2:136051665</td>\n",
       "      <td>[1] \"1000Genomes\" \"GnomAD\"      \"KOREAN\"      ...</td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000002.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>None</td>\n",
       "      <td>3848296</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>16</td>\n",
       "      <td>16:82516943</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000016.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>[1] \"RUFY1\"\\n</td>\n",
       "      <td>4075890</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant,genic_upstream_transcript_variant</td>\n",
       "      <td>5</td>\n",
       "      <td>5:179570372</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>B</td>\n",
       "      <td></td>\n",
       "      <td>NC_000005.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>None</td>\n",
       "      <td>4350902</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>3:184634412</td>\n",
       "      <td>[1] \"1000Genomes\"    \"Estonian\"       \"GENOME...</td>\n",
       "      <td>B</td>\n",
       "      <td></td>\n",
       "      <td>NC_000003.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>[1] \"PLEKHA1\"\\n</td>\n",
       "      <td>10510110</td>\n",
       "      <td>snv</td>\n",
       "      <td>non_coding_transcript_variant,genic_downstream...</td>\n",
       "      <td>10</td>\n",
       "      <td>10:122432914</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>NC_000010.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>[1] \"PFKP-DT\"      \"LOC124902536\"\\n</td>\n",
       "      <td>58183624</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant,upstream_transcript_variant,2KB...</td>\n",
       "      <td>10</td>\n",
       "      <td>10:3065025</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>NC_000010.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>[1] \"TRIM22\"\\n</td>\n",
       "      <td>7480654</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>11</td>\n",
       "      <td>11:5701609</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>B</td>\n",
       "      <td></td>\n",
       "      <td>NC_000011.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>None</td>\n",
       "      <td>4411883</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>3:94830842</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...</td>\n",
       "      <td>K</td>\n",
       "      <td></td>\n",
       "      <td>NC_000003.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>None</td>\n",
       "      <td>74585958</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>3:134054518</td>\n",
       "      <td>[1] \"1000Genomes\"    \"GnomAD\"         \"KOREAN...</td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000003.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>None</td>\n",
       "      <td>9509543</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>13</td>\n",
       "      <td>13:21118265</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>NC_000013.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>None</td>\n",
       "      <td>11672303</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>19</td>\n",
       "      <td>19:33235469</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>NC_000019.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>[1] \"FTO\"\\n</td>\n",
       "      <td>9302654</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant,genic_downstream_transcript_var...</td>\n",
       "      <td>16</td>\n",
       "      <td>16:53975633</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>NC_000016.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>[1] \"SLFN11\"\\n</td>\n",
       "      <td>77814639</td>\n",
       "      <td>snv</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>17</td>\n",
       "      <td>17:35351808</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"GENOME...</td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000017.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>[1] \"EIF4H\"\\n</td>\n",
       "      <td>150664966</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>7</td>\n",
       "      <td>7:74179827</td>\n",
       "      <td>[1] \"1000Genomes\" \"ALSPAC\"      \"Estonian\"   ...</td>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>NC_000007.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>[1] \"EIF4H\"\\n</td>\n",
       "      <td>150664966</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>7</td>\n",
       "      <td>7:74179827</td>\n",
       "      <td>[1] \"1000Genomes\" \"ALSPAC\"      \"Estonian\"   ...</td>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>NC_000007.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>None</td>\n",
       "      <td>4774715</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>15</td>\n",
       "      <td>15:54848006</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>NC_000015.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>None</td>\n",
       "      <td>4774715</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>15</td>\n",
       "      <td>15:54848006</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>NC_000015.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>[1] \"TNIK\"\\n</td>\n",
       "      <td>11925024</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant,genic_downstream_transcript_var...</td>\n",
       "      <td>3</td>\n",
       "      <td>3:171296278</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>H</td>\n",
       "      <td></td>\n",
       "      <td>NC_000003.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>[1] \"TNIK\"\\n</td>\n",
       "      <td>1231</td>\n",
       "      <td>snv</td>\n",
       "      <td>genic_downstream_transcript_variant,intron_var...</td>\n",
       "      <td>3</td>\n",
       "      <td>3:171313444</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>W</td>\n",
       "      <td></td>\n",
       "      <td>NC_000003.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>[1] \"ELN\"\\n</td>\n",
       "      <td>8326</td>\n",
       "      <td>snv</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>7</td>\n",
       "      <td>7:74069359</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>N</td>\n",
       "      <td>benign</td>\n",
       "      <td>NC_000007.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>[1] \"LIMK1\"\\n</td>\n",
       "      <td>710968</td>\n",
       "      <td>snv</td>\n",
       "      <td>2KB_upstream_variant,upstream_transcript_variant</td>\n",
       "      <td>7</td>\n",
       "      <td>7:74083398</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000007.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>[1] \"LIMK1\"\\n</td>\n",
       "      <td>6460071</td>\n",
       "      <td>snv</td>\n",
       "      <td>upstream_transcript_variant,2KB_upstream_variant</td>\n",
       "      <td>7</td>\n",
       "      <td>7:74082866</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000007.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>[1] \"MMP9\"        \"SLC12A5-AS1\"\\n</td>\n",
       "      <td>20544</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant,3_prime_UTR_variant</td>\n",
       "      <td>20</td>\n",
       "      <td>20:46016371</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>B</td>\n",
       "      <td>benign</td>\n",
       "      <td>NC_000020.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>[1] \"LOXL2\"\\n</td>\n",
       "      <td>1010156</td>\n",
       "      <td>snv</td>\n",
       "      <td>coding_sequence_variant,synonymous_variant</td>\n",
       "      <td>8</td>\n",
       "      <td>8:23333428</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...</td>\n",
       "      <td>H</td>\n",
       "      <td></td>\n",
       "      <td>NC_000008.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>[1] \"KLK8\"\\n</td>\n",
       "      <td>1722561</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>19</td>\n",
       "      <td>19:50996987</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000019.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>[1] \"KLK8\"\\n</td>\n",
       "      <td>1701946</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>19</td>\n",
       "      <td>19:51000268</td>\n",
       "      <td>[1] \"1000Genomes\"    \"Estonian\"       \"FINRIS...</td>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>NC_000019.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>None</td>\n",
       "      <td>767603</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>14</td>\n",
       "      <td>14:60631965</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>H</td>\n",
       "      <td></td>\n",
       "      <td>NC_000014.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>[1] \"IL12B\"\\n</td>\n",
       "      <td>3181216</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>5</td>\n",
       "      <td>5:159325970</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>W</td>\n",
       "      <td></td>\n",
       "      <td>NC_000005.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>[1] \"LTA\"          \"TNF\"          \"LOC10028732...</td>\n",
       "      <td>1799964</td>\n",
       "      <td>snv</td>\n",
       "      <td>upstream_transcript_variant,500B_downstream_va...</td>\n",
       "      <td>6</td>\n",
       "      <td>6:31574531</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>NC_000006.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>[1] \"IL1A\"\\n</td>\n",
       "      <td>17561</td>\n",
       "      <td>snv</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>2</td>\n",
       "      <td>2:112779646</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>M</td>\n",
       "      <td></td>\n",
       "      <td>NC_000002.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>None</td>\n",
       "      <td>318216</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>16</td>\n",
       "      <td>16:64225535</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>NC_000016.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>[1] \"IL12B\"\\n</td>\n",
       "      <td>2195940</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>5</td>\n",
       "      <td>5:159317344</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>H</td>\n",
       "      <td></td>\n",
       "      <td>NC_000005.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>[1] \"COL1A2\"\\n</td>\n",
       "      <td>42524</td>\n",
       "      <td>snv</td>\n",
       "      <td>coding_sequence_variant,missense_variant</td>\n",
       "      <td>7</td>\n",
       "      <td>7:94413927</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>N</td>\n",
       "      <td>benign,uncertain-significance</td>\n",
       "      <td>NC_000007.14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                genes        uid snp_class  \\\n",
       "0                                        [1] \"BOLL\"\\n     700651       snv   \n",
       "1                                       [1] \"EDNRA\"\\n    6841581       snv   \n",
       "2                                                None    1105980       snv   \n",
       "3                                                None   12412014       snv   \n",
       "4                       [1] \"LINC00457\" \"LINC02343\"\\n    1536847       snv   \n",
       "5                                     [1] \"C5orf60\"\\n   62405726       snv   \n",
       "6                                                None     549315       snv   \n",
       "7                                      [1] \"GACAT1\"\\n     700855       snv   \n",
       "8                                                None     328025       snv   \n",
       "9                                        [1] \"MALL\"\\n  117802391       snv   \n",
       "10                      [1] \"LINC01978\" \"LINC01979\"\\n   57851800       snv   \n",
       "11                                      [1] \"RREB1\"\\n    9505086       snv   \n",
       "12                                        [1] \"DST\"\\n  117021265       snv   \n",
       "13                                               None    5005908       snv   \n",
       "14                                      [1] \"FOXP1\"\\n     878118       snv   \n",
       "15                                  [1] \"LINC01344\"\\n   12033118       snv   \n",
       "16                                      [1] \"RBMS3\"\\n    1979271       snv   \n",
       "17                                               None  189432614       snv   \n",
       "18                                               None    3848296       snv   \n",
       "19                                      [1] \"RUFY1\"\\n    4075890       snv   \n",
       "20                                               None    4350902       snv   \n",
       "21                                    [1] \"PLEKHA1\"\\n   10510110       snv   \n",
       "22                [1] \"PFKP-DT\"      \"LOC124902536\"\\n   58183624       snv   \n",
       "23                                     [1] \"TRIM22\"\\n    7480654       snv   \n",
       "24                                               None    4411883       snv   \n",
       "25                                               None   74585958       snv   \n",
       "26                                               None    9509543       snv   \n",
       "27                                               None   11672303       snv   \n",
       "28                                        [1] \"FTO\"\\n    9302654       snv   \n",
       "29                                     [1] \"SLFN11\"\\n   77814639       snv   \n",
       "30                                      [1] \"EIF4H\"\\n  150664966       snv   \n",
       "31                                      [1] \"EIF4H\"\\n  150664966       snv   \n",
       "32                                               None    4774715       snv   \n",
       "33                                               None    4774715       snv   \n",
       "34                                       [1] \"TNIK\"\\n   11925024       snv   \n",
       "35                                       [1] \"TNIK\"\\n       1231       snv   \n",
       "36                                        [1] \"ELN\"\\n       8326       snv   \n",
       "37                                      [1] \"LIMK1\"\\n     710968       snv   \n",
       "38                                      [1] \"LIMK1\"\\n    6460071       snv   \n",
       "39                  [1] \"MMP9\"        \"SLC12A5-AS1\"\\n      20544       snv   \n",
       "40                                      [1] \"LOXL2\"\\n    1010156       snv   \n",
       "41                                       [1] \"KLK8\"\\n    1722561       snv   \n",
       "42                                       [1] \"KLK8\"\\n    1701946       snv   \n",
       "43                                               None     767603       snv   \n",
       "44                                      [1] \"IL12B\"\\n    3181216       snv   \n",
       "45  [1] \"LTA\"          \"TNF\"          \"LOC10028732...    1799964       snv   \n",
       "46                                       [1] \"IL1A\"\\n      17561       snv   \n",
       "47                                               None     318216       snv   \n",
       "48                                      [1] \"IL12B\"\\n    2195940       snv   \n",
       "49                                     [1] \"COL1A2\"\\n      42524       snv   \n",
       "\n",
       "                                            fxn_class chr        chrpos  \\\n",
       "0                                      intron_variant   2   2:197766990   \n",
       "1    2KB_upstream_variant,upstream_transcript_variant   4   4:147480038   \n",
       "2                                                       9    9:95351353   \n",
       "3                                                      10   10:12869725   \n",
       "4                                      intron_variant  13   13:34532838   \n",
       "5                                      intron_variant   5   5:179642467   \n",
       "6                                                      18    18:2183055   \n",
       "7    upstream_transcript_variant,2KB_upstream_variant   2   2:107752238   \n",
       "8                                                       2   2:107738589   \n",
       "9                                      intron_variant   2   2:110104507   \n",
       "10                                     intron_variant  17   17:79922572   \n",
       "11                                     intron_variant   6     6:7231953   \n",
       "12   intron_variant,genic_upstream_transcript_variant   6    6:56763223   \n",
       "13                                                      2   2:153147166   \n",
       "14  genic_upstream_transcript_variant,genic_downst...   3    3:71197077   \n",
       "15                                     intron_variant   1   1:182260612   \n",
       "16   genic_upstream_transcript_variant,intron_variant   3    3:29565914   \n",
       "17                                                      2   2:136051665   \n",
       "18                                                     16   16:82516943   \n",
       "19   intron_variant,genic_upstream_transcript_variant   5   5:179570372   \n",
       "20                                                      3   3:184634412   \n",
       "21  non_coding_transcript_variant,genic_downstream...  10  10:122432914   \n",
       "22  intron_variant,upstream_transcript_variant,2KB...  10    10:3065025   \n",
       "23                                     intron_variant  11    11:5701609   \n",
       "24                                                      3    3:94830842   \n",
       "25                                                      3   3:134054518   \n",
       "26                                                     13   13:21118265   \n",
       "27                                                     19   19:33235469   \n",
       "28  intron_variant,genic_downstream_transcript_var...  16   16:53975633   \n",
       "29                                3_prime_UTR_variant  17   17:35351808   \n",
       "30                                     intron_variant   7    7:74179827   \n",
       "31                                     intron_variant   7    7:74179827   \n",
       "32                                                     15   15:54848006   \n",
       "33                                                     15   15:54848006   \n",
       "34  intron_variant,genic_downstream_transcript_var...   3   3:171296278   \n",
       "35  genic_downstream_transcript_variant,intron_var...   3   3:171313444   \n",
       "36                                3_prime_UTR_variant   7    7:74069359   \n",
       "37   2KB_upstream_variant,upstream_transcript_variant   7    7:74083398   \n",
       "38   upstream_transcript_variant,2KB_upstream_variant   7    7:74082866   \n",
       "39                 intron_variant,3_prime_UTR_variant  20   20:46016371   \n",
       "40         coding_sequence_variant,synonymous_variant   8    8:23333428   \n",
       "41                                     intron_variant  19   19:50996987   \n",
       "42                                     intron_variant  19   19:51000268   \n",
       "43                                                     14   14:60631965   \n",
       "44                                     intron_variant   5   5:159325970   \n",
       "45  upstream_transcript_variant,500B_downstream_va...   6    6:31574531   \n",
       "46           missense_variant,coding_sequence_variant   2   2:112779646   \n",
       "47                                                     16   16:64225535   \n",
       "48                                     intron_variant   5   5:159317344   \n",
       "49           coding_sequence_variant,missense_variant   7    7:94413927   \n",
       "\n",
       "                                          global_mafs allele  \\\n",
       "0    [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      R   \n",
       "1    [1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...      N   \n",
       "2    [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      S   \n",
       "3    [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      B   \n",
       "4    [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      D   \n",
       "5    [1] \"1000Genomes\" \"ExAC\"        \"GnomAD\"     ...      N   \n",
       "6    [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      V   \n",
       "7    [1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...      Y   \n",
       "8    [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      D   \n",
       "9    [1] \"1000Genomes\"    \"ALSPAC\"         \"GnomAD...      Y   \n",
       "10   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      V   \n",
       "11   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      Y   \n",
       "12  [1] \"1000Genomes\" \"GnomAD\"      \"KOREAN\"      ...      Y   \n",
       "13   [1] \"1000Genomes\"    \"Estonian\"       \"GENOME...      N   \n",
       "14   [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      D   \n",
       "15   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      H   \n",
       "16   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      N   \n",
       "17  [1] \"1000Genomes\" \"GnomAD\"      \"KOREAN\"      ...      R   \n",
       "18   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      R   \n",
       "19   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      B   \n",
       "20   [1] \"1000Genomes\"    \"Estonian\"       \"GENOME...      B   \n",
       "21   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      N   \n",
       "22   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      Y   \n",
       "23   [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      B   \n",
       "24   [1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...      K   \n",
       "25   [1] \"1000Genomes\"    \"GnomAD\"         \"KOREAN...      R   \n",
       "26   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      N   \n",
       "27   [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      Y   \n",
       "28   [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      Y   \n",
       "29   [1] \"1000Genomes\"    \"ALSPAC\"         \"GENOME...      R   \n",
       "30   [1] \"1000Genomes\" \"ALSPAC\"      \"Estonian\"   ...      Y   \n",
       "31   [1] \"1000Genomes\" \"ALSPAC\"      \"Estonian\"   ...      Y   \n",
       "32   [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      Y   \n",
       "33   [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      Y   \n",
       "34   [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      H   \n",
       "35   [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      W   \n",
       "36   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      N   \n",
       "37   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      R   \n",
       "38   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      R   \n",
       "39   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      B   \n",
       "40   [1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...      H   \n",
       "41   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      R   \n",
       "42   [1] \"1000Genomes\"    \"Estonian\"       \"FINRIS...      N   \n",
       "43   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      H   \n",
       "44   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      W   \n",
       "45   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      Y   \n",
       "46   [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      M   \n",
       "47   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      N   \n",
       "48   [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      H   \n",
       "49   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      N   \n",
       "\n",
       "            clinical_significance           acc  \n",
       "0                                  NC_000002.12  \n",
       "1                                  NC_000004.12  \n",
       "2                                  NC_000009.12  \n",
       "3                                  NC_000010.11  \n",
       "4                                  NC_000013.11  \n",
       "5                                  NC_000005.10  \n",
       "6                                  NC_000018.10  \n",
       "7                                  NC_000002.12  \n",
       "8                                  NC_000002.12  \n",
       "9                                  NC_000002.12  \n",
       "10                                 NC_000017.11  \n",
       "11                                 NC_000006.12  \n",
       "12                                 NC_000006.12  \n",
       "13                                 NC_000002.12  \n",
       "14                                 NC_000003.12  \n",
       "15                                 NC_000001.11  \n",
       "16                                 NC_000003.12  \n",
       "17                                 NC_000002.12  \n",
       "18                                 NC_000016.10  \n",
       "19                                 NC_000005.10  \n",
       "20                                 NC_000003.12  \n",
       "21                                 NC_000010.11  \n",
       "22                                 NC_000010.11  \n",
       "23                                 NC_000011.10  \n",
       "24                                 NC_000003.12  \n",
       "25                                 NC_000003.12  \n",
       "26                                 NC_000013.11  \n",
       "27                                 NC_000019.10  \n",
       "28                                 NC_000016.10  \n",
       "29                                 NC_000017.11  \n",
       "30                                 NC_000007.14  \n",
       "31                                 NC_000007.14  \n",
       "32                                 NC_000015.10  \n",
       "33                                 NC_000015.10  \n",
       "34                                 NC_000003.12  \n",
       "35                                 NC_000003.12  \n",
       "36                         benign  NC_000007.14  \n",
       "37                                 NC_000007.14  \n",
       "38                                 NC_000007.14  \n",
       "39                         benign  NC_000020.11  \n",
       "40                                 NC_000008.11  \n",
       "41                                 NC_000019.10  \n",
       "42                                 NC_000019.10  \n",
       "43                                  NC_000014.9  \n",
       "44                                 NC_000005.10  \n",
       "45                                 NC_000006.12  \n",
       "46                                 NC_000002.12  \n",
       "47                                 NC_000016.10  \n",
       "48                                 NC_000005.10  \n",
       "49  benign,uncertain-significance  NC_000007.14  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_snp_data = []\n",
    "\n",
    "# Open a text file in write mode to log skipped rows\n",
    "with open(\"skipped_rows_log.txt\", \"w\") as log_file:\n",
    "    log_file.write(\"Details of rows skipped due to None UID:\\n\\n\")\n",
    "\n",
    "    for index, row in UID_SNP.iterrows():\n",
    "        test_snp = row['UID']\n",
    "        \n",
    "        # Check if the UID is not None\n",
    "        if test_snp is None:\n",
    "            # Print to console and write to log file\n",
    "            log_message = f\"Skipping row {index} because UID is None. Row details:\\n{row}\\n\\n\"\n",
    "            print(log_message)\n",
    "            log_file.write(log_message)\n",
    "            continue  # Skip this iteration if UID is None\n",
    "\n",
    "        try: \n",
    "            # Retrieve SNP data from the SNP database\n",
    "            search_test = rentrez.entrez_summary(db='snp', id=test_snp)\n",
    "            \n",
    "            # Filter and safely retrieve desired fields\n",
    "            filtered_search_test_dict = {\n",
    "                key: str(search_test.rx2(key)[0]) if len(search_test.rx2(key)) > 0 else None\n",
    "                for key in search_test.names if key in desired_dbSNP_values\n",
    "            }\n",
    "            \n",
    "            # Order the data to match the desired column order\n",
    "            ordered_data = {key: filtered_search_test_dict.get(key, None) for key in desired_dbSNP_values}\n",
    "            \n",
    "            # Append ordered data to the results list\n",
    "            all_snp_data.append(ordered_data)\n",
    "\n",
    "        except Exception as e:\n",
    "            error_message = f\"Error retrieving data for UID {test_snp}: {e}\\n\"\n",
    "            print(error_message)\n",
    "            log_file.write(error_message)\n",
    "\n",
    "# Convert results to DataFrame\n",
    "final_df = pd.DataFrame(all_snp_data)\n",
    "\n",
    "# Display the first 50 rows\n",
    "final_df.head(50)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f6520708-7ff6-44ce-8cf3-55a4169a0a38",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Now let's make a copy of our RAW results:\n",
    "final_df.to_csv(os.path.join(output_path, 'compiled_SNPs_RENTREZ_RAW.csv'), index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "281f19ee-bb6b-49ab-9700-8092436a009b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>uid</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>chr</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>global_mafs</th>\n",
       "      <th>allele</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>acc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[1] \"BOLL\"\\n</td>\n",
       "      <td>700651</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>2</td>\n",
       "      <td>2:197766990</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000002.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[1] \"EDNRA\"\\n</td>\n",
       "      <td>6841581</td>\n",
       "      <td>snv</td>\n",
       "      <td>2KB_upstream_variant,upstream_transcript_variant</td>\n",
       "      <td>4</td>\n",
       "      <td>4:147480038</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...</td>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>NC_000004.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>None</td>\n",
       "      <td>1105980</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>9</td>\n",
       "      <td>9:95351353</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>S</td>\n",
       "      <td></td>\n",
       "      <td>NC_000009.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>None</td>\n",
       "      <td>12412014</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>10:12869725</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>B</td>\n",
       "      <td></td>\n",
       "      <td>NC_000010.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[1] \"LINC00457\" \"LINC02343\"\\n</td>\n",
       "      <td>1536847</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>13</td>\n",
       "      <td>13:34532838</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>D</td>\n",
       "      <td></td>\n",
       "      <td>NC_000013.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284</th>\n",
       "      <td>[1] \"THSD1\"\\n</td>\n",
       "      <td>186046951</td>\n",
       "      <td>snv</td>\n",
       "      <td>coding_sequence_variant,missense_variant</td>\n",
       "      <td>13</td>\n",
       "      <td>13:52397661</td>\n",
       "      <td>[1] \"1000Genomes\"   \"ExAC\"          \"GnomAD\"  ...</td>\n",
       "      <td>S</td>\n",
       "      <td></td>\n",
       "      <td>NC_000013.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285</th>\n",
       "      <td>[1] \"TMEM132B\"\\n</td>\n",
       "      <td>61940807</td>\n",
       "      <td>snv</td>\n",
       "      <td>coding_sequence_variant,genic_downstream_trans...</td>\n",
       "      <td>12</td>\n",
       "      <td>12:125654210</td>\n",
       "      <td>[1] \"1000Genomes\"   \"ALSPAC\"        \"Estonian...</td>\n",
       "      <td>M</td>\n",
       "      <td></td>\n",
       "      <td>NC_000012.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>[1] \"LOX\"    \"SRFBP1\"\\n</td>\n",
       "      <td>10519694</td>\n",
       "      <td>snv</td>\n",
       "      <td>genic_downstream_transcript_variant,intron_var...</td>\n",
       "      <td>5</td>\n",
       "      <td>5:122071524</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>NC_000005.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>[1] \"LOXL2\"\\n</td>\n",
       "      <td>1010156</td>\n",
       "      <td>snv</td>\n",
       "      <td>coding_sequence_variant,synonymous_variant</td>\n",
       "      <td>8</td>\n",
       "      <td>8:23333428</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...</td>\n",
       "      <td>H</td>\n",
       "      <td></td>\n",
       "      <td>NC_000008.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288</th>\n",
       "      <td>None</td>\n",
       "      <td>1746048</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>10:44280376</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...</td>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>NC_000010.11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>289 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             genes        uid snp_class  \\\n",
       "0                     [1] \"BOLL\"\\n     700651       snv   \n",
       "1                    [1] \"EDNRA\"\\n    6841581       snv   \n",
       "2                             None    1105980       snv   \n",
       "3                             None   12412014       snv   \n",
       "4    [1] \"LINC00457\" \"LINC02343\"\\n    1536847       snv   \n",
       "..                             ...        ...       ...   \n",
       "284                  [1] \"THSD1\"\\n  186046951       snv   \n",
       "285               [1] \"TMEM132B\"\\n   61940807       snv   \n",
       "286        [1] \"LOX\"    \"SRFBP1\"\\n   10519694       snv   \n",
       "287                  [1] \"LOXL2\"\\n    1010156       snv   \n",
       "288                           None    1746048       snv   \n",
       "\n",
       "                                             fxn_class chr        chrpos  \\\n",
       "0                                       intron_variant   2   2:197766990   \n",
       "1     2KB_upstream_variant,upstream_transcript_variant   4   4:147480038   \n",
       "2                                                        9    9:95351353   \n",
       "3                                                       10   10:12869725   \n",
       "4                                       intron_variant  13   13:34532838   \n",
       "..                                                 ...  ..           ...   \n",
       "284           coding_sequence_variant,missense_variant  13   13:52397661   \n",
       "285  coding_sequence_variant,genic_downstream_trans...  12  12:125654210   \n",
       "286  genic_downstream_transcript_variant,intron_var...   5   5:122071524   \n",
       "287         coding_sequence_variant,synonymous_variant   8    8:23333428   \n",
       "288                                                     10   10:44280376   \n",
       "\n",
       "                                           global_mafs allele  \\\n",
       "0     [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      R   \n",
       "1     [1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...      N   \n",
       "2     [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      S   \n",
       "3     [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      B   \n",
       "4     [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      D   \n",
       "..                                                 ...    ...   \n",
       "284  [1] \"1000Genomes\"   \"ExAC\"          \"GnomAD\"  ...      S   \n",
       "285   [1] \"1000Genomes\"   \"ALSPAC\"        \"Estonian...      M   \n",
       "286   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      Y   \n",
       "287   [1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...      H   \n",
       "288   [1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...      Y   \n",
       "\n",
       "    clinical_significance           acc  \n",
       "0                          NC_000002.12  \n",
       "1                          NC_000004.12  \n",
       "2                          NC_000009.12  \n",
       "3                          NC_000010.11  \n",
       "4                          NC_000013.11  \n",
       "..                    ...           ...  \n",
       "284                        NC_000013.11  \n",
       "285                        NC_000012.12  \n",
       "286                        NC_000005.10  \n",
       "287                        NC_000008.11  \n",
       "288                        NC_000010.11  \n",
       "\n",
       "[289 rows x 10 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### And let's make it beautiful by rearranging and sorting each column:\n",
    "#### First, let's delete all variants that are not SNVs:\n",
    "filtered_final_df = final_df.loc[final_df['snp_class'] == 'snv'] \n",
    "filtered_final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9e3773e8-af4e-4ae4-b3a0-61bc87730794",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>uid</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>chr</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>global_mafs</th>\n",
       "      <th>allele</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>acc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BOLL</td>\n",
       "      <td>700651</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>2</td>\n",
       "      <td>2:197766990</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000002.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EDNRA</td>\n",
       "      <td>6841581</td>\n",
       "      <td>snv</td>\n",
       "      <td>2KB_upstream_variant,upstream_transcript_variant</td>\n",
       "      <td>4</td>\n",
       "      <td>4:147480038</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...</td>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>NC_000004.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>None</td>\n",
       "      <td>1105980</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>9</td>\n",
       "      <td>9:95351353</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>S</td>\n",
       "      <td></td>\n",
       "      <td>NC_000009.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>None</td>\n",
       "      <td>12412014</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>10:12869725</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>B</td>\n",
       "      <td></td>\n",
       "      <td>NC_000010.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LINC00457 LINC02343</td>\n",
       "      <td>1536847</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>13</td>\n",
       "      <td>13:34532838</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>D</td>\n",
       "      <td></td>\n",
       "      <td>NC_000013.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>LOC105376234</td>\n",
       "      <td>4979583</td>\n",
       "      <td>snv</td>\n",
       "      <td>genic_downstream_transcript_variant,intron_var...</td>\n",
       "      <td>9</td>\n",
       "      <td>9:115809149</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...</td>\n",
       "      <td>H</td>\n",
       "      <td></td>\n",
       "      <td>NC_000009.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>OLFML2A</td>\n",
       "      <td>79134766</td>\n",
       "      <td>snv</td>\n",
       "      <td>upstream_transcript_variant,missense_variant,c...</td>\n",
       "      <td>9</td>\n",
       "      <td>9:124799444</td>\n",
       "      <td>[1] \"1000Genomes\"    \"Estonian\"       \"GnomAD...</td>\n",
       "      <td>D</td>\n",
       "      <td></td>\n",
       "      <td>NC_000009.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>LOC105369370</td>\n",
       "      <td>76855873</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>11</td>\n",
       "      <td>11:69339687</td>\n",
       "      <td>[1] \"1000Genomes\" \"ALSPAC\"      \"GnomAD\"     ...</td>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>NC_000011.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>ARHGAP32</td>\n",
       "      <td>371331393</td>\n",
       "      <td>snv</td>\n",
       "      <td>stop_gained,genic_downstream_transcript_varian...</td>\n",
       "      <td>11</td>\n",
       "      <td>11:128969377</td>\n",
       "      <td>[1] \"GnomAD\"     \"GoESP\"      \"PAGE_STUDY\" \"TO...</td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "      <td>NC_000011.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>CD163L1</td>\n",
       "      <td>138525217</td>\n",
       "      <td>snv</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>12</td>\n",
       "      <td>12:7375596</td>\n",
       "      <td>[1] \"1000Genomes\"   \"ExAC\"          \"GnomAD\" ...</td>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>NC_000012.12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>200 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   genes        uid snp_class  \\\n",
       "0                   BOLL     700651       snv   \n",
       "1                  EDNRA    6841581       snv   \n",
       "2                   None    1105980       snv   \n",
       "3                   None   12412014       snv   \n",
       "4    LINC00457 LINC02343    1536847       snv   \n",
       "..                   ...        ...       ...   \n",
       "195         LOC105376234    4979583       snv   \n",
       "196              OLFML2A   79134766       snv   \n",
       "197         LOC105369370   76855873       snv   \n",
       "198             ARHGAP32  371331393       snv   \n",
       "199              CD163L1  138525217       snv   \n",
       "\n",
       "                                             fxn_class chr        chrpos  \\\n",
       "0                                       intron_variant   2   2:197766990   \n",
       "1     2KB_upstream_variant,upstream_transcript_variant   4   4:147480038   \n",
       "2                                                        9    9:95351353   \n",
       "3                                                       10   10:12869725   \n",
       "4                                       intron_variant  13   13:34532838   \n",
       "..                                                 ...  ..           ...   \n",
       "195  genic_downstream_transcript_variant,intron_var...   9   9:115809149   \n",
       "196  upstream_transcript_variant,missense_variant,c...   9   9:124799444   \n",
       "197                                     intron_variant  11   11:69339687   \n",
       "198  stop_gained,genic_downstream_transcript_varian...  11  11:128969377   \n",
       "199                            splice_acceptor_variant  12    12:7375596   \n",
       "\n",
       "                                           global_mafs allele  \\\n",
       "0     [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      R   \n",
       "1     [1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...      N   \n",
       "2     [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      S   \n",
       "3     [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      B   \n",
       "4     [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      D   \n",
       "..                                                 ...    ...   \n",
       "195   [1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...      H   \n",
       "196   [1] \"1000Genomes\"    \"Estonian\"       \"GnomAD...      D   \n",
       "197   [1] \"1000Genomes\" \"ALSPAC\"      \"GnomAD\"     ...      Y   \n",
       "198  [1] \"GnomAD\"     \"GoESP\"      \"PAGE_STUDY\" \"TO...      R   \n",
       "199   [1] \"1000Genomes\"   \"ExAC\"          \"GnomAD\" ...      Y   \n",
       "\n",
       "    clinical_significance           acc  \n",
       "0                          NC_000002.12  \n",
       "1                          NC_000004.12  \n",
       "2                          NC_000009.12  \n",
       "3                          NC_000010.11  \n",
       "4                          NC_000013.11  \n",
       "..                    ...           ...  \n",
       "195                        NC_000009.12  \n",
       "196                        NC_000009.12  \n",
       "197                        NC_000011.10  \n",
       "198                        NC_000011.10  \n",
       "199                        NC_000012.12  \n",
       "\n",
       "[200 rows x 10 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Now let's clean up la mugre of each cell in the genes column\n",
    "\n",
    "# Define a function to clean up each cell\n",
    "def clean_cell(value):\n",
    "    if isinstance(value, str):\n",
    "        # Remove `[1]`, quotation marks, and newline characters\n",
    "        cleaned_value = re.sub(r'^\\[\\d+\\]\\s*\"', '', value).replace('\"', '').strip()\n",
    "        return cleaned_value\n",
    "    return value\n",
    "\n",
    "# Apply the cleaning function to the 'genes' column using .loc\n",
    "filtered_final_df.loc[:, 'genes'] = filtered_final_df['genes'].apply(clean_cell)\n",
    "\n",
    "filtered_final_df.head(200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "91e34fd6-15e9-4f15-ad5e-2f8c15b1e9e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID (GENE_ID)</th>\n",
       "      <th>dbSNP ID (RS)</th>\n",
       "      <th>SNP Class (SCLS)</th>\n",
       "      <th>Functional consequence class (FXN)</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Chromosome base position (CPOS)</th>\n",
       "      <th>Minor Allele Frequency from global population (GMAF)</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Associated clinical phenotypes (CLIN)</th>\n",
       "      <th>Nucleotide Accessions (ACCN)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BOLL</td>\n",
       "      <td>700651</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>2</td>\n",
       "      <td>2:197766990</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>A/G</td>\n",
       "      <td></td>\n",
       "      <td>NC_000002.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EDNRA</td>\n",
       "      <td>6841581</td>\n",
       "      <td>snv</td>\n",
       "      <td>2KB_upstream_variant,upstream_transcript_variant</td>\n",
       "      <td>4</td>\n",
       "      <td>4:147480038</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...</td>\n",
       "      <td>A/T/C/G</td>\n",
       "      <td></td>\n",
       "      <td>NC_000004.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>None</td>\n",
       "      <td>1105980</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>9</td>\n",
       "      <td>9:95351353</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>G/C</td>\n",
       "      <td></td>\n",
       "      <td>NC_000009.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>None</td>\n",
       "      <td>12412014</td>\n",
       "      <td>snv</td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>10:12869725</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>C/G/T</td>\n",
       "      <td></td>\n",
       "      <td>NC_000010.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LINC00457 LINC02343</td>\n",
       "      <td>1536847</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>13</td>\n",
       "      <td>13:34532838</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>A/G/T</td>\n",
       "      <td></td>\n",
       "      <td>NC_000013.11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene ID (GENE_ID) dbSNP ID (RS) SNP Class (SCLS)  \\\n",
       "0                 BOLL        700651              snv   \n",
       "1                EDNRA       6841581              snv   \n",
       "2                 None       1105980              snv   \n",
       "3                 None      12412014              snv   \n",
       "4  LINC00457 LINC02343       1536847              snv   \n",
       "\n",
       "                 Functional consequence class (FXN) Chromosome  \\\n",
       "0                                    intron_variant          2   \n",
       "1  2KB_upstream_variant,upstream_transcript_variant          4   \n",
       "2                                                            9   \n",
       "3                                                           10   \n",
       "4                                    intron_variant         13   \n",
       "\n",
       "  Chromosome base position (CPOS)  \\\n",
       "0                     2:197766990   \n",
       "1                     4:147480038   \n",
       "2                      9:95351353   \n",
       "3                     10:12869725   \n",
       "4                     13:34532838   \n",
       "\n",
       "  Minor Allele Frequency from global population (GMAF)   Allele  \\\n",
       "0   [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...        A/G   \n",
       "1   [1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...    A/T/C/G   \n",
       "2   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...        G/C   \n",
       "3   [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      C/G/T   \n",
       "4   [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      A/G/T   \n",
       "\n",
       "  Associated clinical phenotypes (CLIN) Nucleotide Accessions (ACCN)  \n",
       "0                                                       NC_000002.12  \n",
       "1                                                       NC_000004.12  \n",
       "2                                                       NC_000009.12  \n",
       "3                                                       NC_000010.11  \n",
       "4                                                       NC_000013.11  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a copy to avoid the SettingWithCopyWarning\n",
    "filtered_final_df = filtered_final_df.copy()\n",
    "\n",
    "# Rename columns\n",
    "filtered_final_df.rename(columns={'genes': 'Gene ID (GENE_ID)', 'uid': 'dbSNP ID (RS)',\n",
    "                                  'snp_class': 'SNP Class (SCLS)', 'fxn_class': 'Functional consequence class (FXN)', \n",
    "                                  'chr': 'Chromosome', 'chrpos': 'Chromosome base position (CPOS)', \n",
    "                                  'global_mafs': 'Minor Allele Frequency from global population (GMAF)',\n",
    "                                  'clinical_significance': 'Associated clinical phenotypes (CLIN)',\n",
    "                                  'acc': 'Nucleotide Accessions (ACCN)', 'allele': 'Allele'                                 \n",
    "                                 }, inplace=True)\n",
    "\n",
    "# IUPAC mapping dictionary for ambiguity codes\n",
    "iupac_mapping = {\n",
    "    \"R\": \"A/G\", \"Y\": \"C/T\", \"S\": \"G/C\", \"W\": \"A/T\",\n",
    "    \"K\": \"G/T\", \"M\": \"A/C\", \"B\": \"C/G/T\", \"D\": \"A/G/T\",\n",
    "    \"H\": \"A/C/T\", \"V\": \"A/C/G\", \"N\": \"A/T/C/G\"\n",
    "}\n",
    "\n",
    "# Define a function to translate IUPAC codes in the 'allele' column\n",
    "def translate_allele(allele):\n",
    "    if isinstance(allele, str):  # Check if the value is a string\n",
    "        # Replace each IUPAC code with its corresponding alleles\n",
    "        return \"/\".join([iupac_mapping.get(a, a) for a in allele])\n",
    "    return allele\n",
    "\n",
    "# Assume filtered_final_df is your DataFrame and 'allele' is the column with IUPAC codes\n",
    "# Apply the translation function to the 'allele' column\n",
    "filtered_final_df['Allele'] = filtered_final_df['Allele'].apply(translate_allele)\n",
    "\n",
    "# Display the updated DataFrame\n",
    "filtered_final_df.head(5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6498f193-b198-486d-a815-a39041c0796c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene Symbol: BOLL, Ensembl Gene ID: ENSG00000152430, Gene Type: protein_coding\n",
      "Gene Symbol: EDNRA, Ensembl Gene ID: ENSG00000151617, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'LINC00457 LINC02343' in Ensembl.\n",
      "Gene Symbol: C5orf60, Ensembl Gene ID: ENSG00000204661, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: GACAT1, Ensembl Gene ID: ENSG00000232991, Gene Type: lncRNA\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: MALL, Ensembl Gene ID: ENSG00000144063, Gene Type: protein_coding\n",
      "No gene found for symbol 'LINC01978 LINC01979' in Ensembl.\n",
      "Gene Symbol: RREB1, Ensembl Gene ID: ENSG00000124782, Gene Type: protein_coding\n",
      "Gene Symbol: DST, Ensembl Gene ID: ENSG00000151914, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: FOXP1, Ensembl Gene ID: ENSG00000114861, Gene Type: protein_coding\n",
      "Gene Symbol: LINC01344, Ensembl Gene ID: ENSG00000228918, Gene Type: lncRNA\n",
      "Gene Symbol: RBMS3, Ensembl Gene ID: ENSG00000144642, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: RUFY1, Ensembl Gene ID: ENSG00000176783, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: PLEKHA1, Ensembl Gene ID: ENSG00000107679, Gene Type: protein_coding\n",
      "No gene found for symbol 'PFKP-DT      LOC124902536' in Ensembl.\n",
      "Gene Symbol: TRIM22, Ensembl Gene ID: ENSG00000132274, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: FTO, Ensembl Gene ID: ENSG00000140718, Gene Type: protein_coding\n",
      "Gene Symbol: SLFN11, Ensembl Gene ID: ENSG00000172716, Gene Type: protein_coding\n",
      "Gene Symbol: EIF4H, Ensembl Gene ID: ENSG00000106682, Gene Type: protein_coding\n",
      "Gene Symbol: EIF4H, Ensembl Gene ID: ENSG00000106682, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: TNIK, Ensembl Gene ID: ENSG00000154310, Gene Type: protein_coding\n",
      "Gene Symbol: TNIK, Ensembl Gene ID: ENSG00000154310, Gene Type: protein_coding\n",
      "Gene Symbol: ELN, Ensembl Gene ID: ENSG00000049540, Gene Type: protein_coding\n",
      "Gene Symbol: LIMK1, Ensembl Gene ID: ENSG00000106683, Gene Type: protein_coding\n",
      "Gene Symbol: LIMK1, Ensembl Gene ID: ENSG00000106683, Gene Type: protein_coding\n",
      "No gene found for symbol 'MMP9        SLC12A5-AS1' in Ensembl.\n",
      "Gene Symbol: LOXL2, Ensembl Gene ID: ENSG00000134013, Gene Type: protein_coding\n",
      "Gene Symbol: KLK8, Ensembl Gene ID: ENSG00000129455, Gene Type: protein_coding\n",
      "Gene Symbol: KLK8, Ensembl Gene ID: ENSG00000129455, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: IL12B, Ensembl Gene ID: ENSG00000113302, Gene Type: protein_coding\n",
      "No gene found for symbol 'LTA          TNF          LOC100287329' in Ensembl.\n",
      "Gene Symbol: IL1A, Ensembl Gene ID: ENSG00000115008, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: IL12B, Ensembl Gene ID: ENSG00000113302, Gene Type: protein_coding\n",
      "Gene Symbol: COL1A2, Ensembl Gene ID: ENSG00000164692, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: CDKN2B-AS1, Ensembl Gene ID: ENSG00000240498, Gene Type: lncRNA\n",
      "Gene Symbol: CDKN2B-AS1, Ensembl Gene ID: ENSG00000240498, Gene Type: lncRNA\n",
      "Gene Symbol: BET1L, Ensembl Gene ID: ENSG00000177951, Gene Type: protein_coding\n",
      "No gene found for symbol 'MRPL23 H19' in Ensembl.\n",
      "No gene found for symbol 'MRPL23 H19' in Ensembl.\n",
      "Gene Symbol: TNFRSF13B, Ensembl Gene ID: ENSG00000240505, Gene Type: protein_coding\n",
      "Gene Symbol: LIMK1, Ensembl Gene ID: ENSG00000106683, Gene Type: protein_coding\n",
      "Gene Symbol: MMP2, Ensembl Gene ID: ENSG00000087245, Gene Type: protein_coding\n",
      "Gene Symbol: MMP2, Ensembl Gene ID: ENSG00000087245, Gene Type: protein_coding\n",
      "No gene found for symbol 'LTA TNF' in Ensembl.\n",
      "Gene Symbol: ARHGEF17, Ensembl Gene ID: ENSG00000110237, Gene Type: protein_coding\n",
      "No gene found for symbol 'MMP24           MMP24-AS1-EDEM2' in Ensembl.\n",
      "No gene found for symbol 'MMP24           MMP24-AS1-EDEM2' in Ensembl.\n",
      "Gene Symbol: MMP26, Ensembl Gene ID: ENSG00000167346, Gene Type: protein_coding\n",
      "No gene found for symbol 'MMP24           MMP24-AS1-EDEM2' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: MMP13, Ensembl Gene ID: ENSG00000137745, Gene Type: protein_coding\n",
      "No gene found for symbol 'MMP24           MMP24-AS1-EDEM2' in Ensembl.\n",
      "Gene Symbol: MMP2, Ensembl Gene ID: ENSG00000087245, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: MMP26, Ensembl Gene ID: ENSG00000167346, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: MMP26, Ensembl Gene ID: ENSG00000167346, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'MMP24           MMP24-AS1-EDEM2' in Ensembl.\n",
      "Gene Symbol: MMP17, Ensembl Gene ID: ENSG00000198598, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'MMP24           MMP24-AS1-EDEM2' in Ensembl.\n",
      "Gene Symbol: MMP26, Ensembl Gene ID: ENSG00000167346, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: MMP27, Ensembl Gene ID: ENSG00000137675, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'MMP1   WTAPP1' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: MMP16, Ensembl Gene ID: ENSG00000156103, Gene Type: protein_coding\n",
      "Gene Symbol: MMP7, Ensembl Gene ID: ENSG00000137673, Gene Type: protein_coding\n",
      "Gene Symbol: USB1, Ensembl Gene ID: ENSG00000103005, Gene Type: protein_coding\n",
      "Gene Symbol: MMP14, Ensembl Gene ID: ENSG00000157227, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: MMP16, Ensembl Gene ID: ENSG00000156103, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: MMP25, Ensembl Gene ID: ENSG00000008516, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: MMP3, Ensembl Gene ID: ENSG00000149968, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: MMP16, Ensembl Gene ID: ENSG00000156103, Gene Type: protein_coding\n",
      "No gene found for symbol 'MMP25     MMP25-AS1' in Ensembl.\n",
      "Gene Symbol: MMP26, Ensembl Gene ID: ENSG00000167346, Gene Type: protein_coding\n",
      "Gene Symbol: MMP26, Ensembl Gene ID: ENSG00000167346, Gene Type: protein_coding\n",
      "No gene found for symbol 'MMP1   WTAPP1' in Ensembl.\n",
      "Gene Symbol: MMP14, Ensembl Gene ID: ENSG00000157227, Gene Type: protein_coding\n",
      "Gene Symbol: MMP13, Ensembl Gene ID: ENSG00000137745, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: ENG, Ensembl Gene ID: ENSG00000106991, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: IGSF9B, Ensembl Gene ID: ENSG00000080854, Gene Type: protein_coding\n",
      "Gene Symbol: TIGAR, Ensembl Gene ID: ENSG00000078237, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: ATP11A, Ensembl Gene ID: ENSG00000068650, Gene Type: protein_coding\n",
      "Gene Symbol: ZNF469, Ensembl Gene ID: ENSG00000225614, Gene Type: protein_coding\n",
      "No gene found for symbol 'MYH13        LOC107985004' in Ensembl.\n",
      "Gene Symbol: SLC47A1, Ensembl Gene ID: ENSG00000142494, Gene Type: protein_coding\n",
      "No gene found for symbol 'IL12B        LOC105377683' in Ensembl.\n",
      "Gene Symbol: SIRT3, Ensembl Gene ID: ENSG00000142082, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: CDKN2B-AS1, Ensembl Gene ID: ENSG00000240498, Gene Type: lncRNA\n",
      "Gene Symbol: ATP2B1, Ensembl Gene ID: ENSG00000070961, Gene Type: protein_coding\n",
      "Gene Symbol: MIRLET7A1HG, Ensembl Gene ID: ENSG00000269929, Gene Type: lncRNA\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: ADAMTS12, Ensembl Gene ID: ENSG00000151388, Gene Type: protein_coding\n",
      "Gene Symbol: ADAMTS13, Ensembl Gene ID: ENSG00000160323, Gene Type: protein_coding\n",
      "No gene found for symbol 'ADAMTS13 CACFD1' in Ensembl.\n",
      "Gene Symbol: ADAMTS13, Ensembl Gene ID: ENSG00000160323, Gene Type: protein_coding\n",
      "Gene Symbol: ADAMTS13, Ensembl Gene ID: ENSG00000160323, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: CDKN2B-AS1, Ensembl Gene ID: ENSG00000240498, Gene Type: lncRNA\n",
      "Gene Symbol: CDKN2B-AS1, Ensembl Gene ID: ENSG00000240498, Gene Type: lncRNA\n",
      "Gene Symbol: CDKN2B-AS1, Ensembl Gene ID: ENSG00000240498, Gene Type: lncRNA\n",
      "Gene Symbol: CDKN2B-AS1, Ensembl Gene ID: ENSG00000240498, Gene Type: lncRNA\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: ENG, Ensembl Gene ID: ENSG00000106991, Gene Type: protein_coding\n",
      "Gene Symbol: JDP2, Ensembl Gene ID: ENSG00000140044, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: KMT2D, Ensembl Gene ID: ENSG00000167548, Gene Type: protein_coding\n",
      "Gene Symbol: COL3A1, Ensembl Gene ID: ENSG00000168542, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: VCAN, Ensembl Gene ID: ENSG00000038427, Gene Type: protein_coding\n",
      "Gene Symbol: LIMK1, Ensembl Gene ID: ENSG00000106683, Gene Type: protein_coding\n",
      "Gene Symbol: STARD13, Ensembl Gene ID: ENSG00000133121, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: SERPINA3, Ensembl Gene ID: ENSG00000196136, Gene Type: protein_coding\n",
      "Gene Symbol: ARHGEF11, Ensembl Gene ID: ENSG00000132694, Gene Type: protein_coding\n",
      "Gene Symbol: NOS3, Ensembl Gene ID: ENSG00000164867, Gene Type: protein_coding\n",
      "Gene Symbol: TLR4, Ensembl Gene ID: ENSG00000136869, Gene Type: protein_coding\n",
      "Gene Symbol: VCAN, Ensembl Gene ID: ENSG00000038427, Gene Type: protein_coding\n",
      "Gene Symbol: VCAN, Ensembl Gene ID: ENSG00000038427, Gene Type: protein_coding\n",
      "Gene Symbol: VCAN, Ensembl Gene ID: ENSG00000038427, Gene Type: protein_coding\n",
      "Gene Symbol: VCAN, Ensembl Gene ID: ENSG00000038427, Gene Type: protein_coding\n",
      "Gene Symbol: VCAN, Ensembl Gene ID: ENSG00000038427, Gene Type: protein_coding\n",
      "Gene Symbol: VCAN, Ensembl Gene ID: ENSG00000038427, Gene Type: protein_coding\n",
      "No gene found for symbol 'VCAN     VCAN-AS1' in Ensembl.\n",
      "No gene found for symbol 'VCAN     VCAN-AS1' in Ensembl.\n",
      "No gene found for symbol 'VCAN     VCAN-AS1' in Ensembl.\n",
      "No gene found for symbol 'VCAN     VCAN-AS1' in Ensembl.\n",
      "Gene Symbol: VCAN, Ensembl Gene ID: ENSG00000038427, Gene Type: protein_coding\n",
      "Gene Symbol: VCAN, Ensembl Gene ID: ENSG00000038427, Gene Type: protein_coding\n",
      "Gene Symbol: CDKN2B-AS1, Ensembl Gene ID: ENSG00000240498, Gene Type: lncRNA\n",
      "Gene Symbol: CDKN2B-AS1, Ensembl Gene ID: ENSG00000240498, Gene Type: lncRNA\n",
      "Gene Symbol: SLC2A9, Ensembl Gene ID: ENSG00000109667, Gene Type: protein_coding\n",
      "Gene Symbol: TOX, Ensembl Gene ID: ENSG00000198846, Gene Type: protein_coding\n",
      "Gene Symbol: MMP2, Ensembl Gene ID: ENSG00000087245, Gene Type: protein_coding\n",
      "Gene Symbol: VCAN, Ensembl Gene ID: ENSG00000038427, Gene Type: protein_coding\n",
      "Gene Symbol: VCAN, Ensembl Gene ID: ENSG00000038427, Gene Type: protein_coding\n",
      "No gene found for symbol 'IL12B        LOC105377683' in Ensembl.\n",
      "No gene found for symbol 'IL12A     IL12A-AS1' in Ensembl.\n",
      "No gene found for symbol 'IL12A     IL12A-AS1' in Ensembl.\n",
      "Gene Symbol: SMAD3, Ensembl Gene ID: ENSG00000166949, Gene Type: protein_coding\n",
      "Gene Symbol: TNF, Ensembl Gene ID: ENSG00000232810, Gene Type: protein_coding\n",
      "Gene Symbol: IFNG, Ensembl Gene ID: ENSG00000111537, Gene Type: protein_coding\n",
      "No gene found for symbol 'IL6     IL6-AS1' in Ensembl.\n",
      "No gene found for symbol 'IL10 IL19' in Ensembl.\n",
      "No gene found for symbol 'IL10 IL19' in Ensembl.\n",
      "Gene Symbol: PRDM2, Ensembl Gene ID: ENSG00000116731, Gene Type: protein_coding\n",
      "No gene found for symbol 'GBA' in Ensembl.\n",
      "Gene Symbol: FMO4, Ensembl Gene ID: ENSG00000076258, Gene Type: protein_coding\n",
      "Gene Symbol: RNF144A, Ensembl Gene ID: ENSG00000151692, Gene Type: protein_coding\n",
      "Gene Symbol: LINC02991, Ensembl Gene ID: ENSG00000196758, Gene Type: lncRNA\n",
      "Gene Symbol: LINC01237, Ensembl Gene ID: ENSG00000233806, Gene Type: lncRNA\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: TCF24, Ensembl Gene ID: ENSG00000261787, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: LINGO2, Ensembl Gene ID: ENSG00000174482, Gene Type: protein_coding\n",
      "No gene found for symbol 'TMOD1        LOC105376168' in Ensembl.\n",
      "Gene Symbol: SUSD1, Ensembl Gene ID: ENSG00000106868, Gene Type: protein_coding\n",
      "No gene found for symbol 'LOC105376234' in Ensembl.\n",
      "Gene Symbol: OLFML2A, Ensembl Gene ID: ENSG00000185585, Gene Type: protein_coding\n",
      "No gene found for symbol 'LOC105369370' in Ensembl.\n",
      "Gene Symbol: ARHGAP32, Ensembl Gene ID: ENSG00000134909, Gene Type: protein_coding\n",
      "Gene Symbol: CD163L1, Ensembl Gene ID: ENSG00000177675, Gene Type: protein_coding\n",
      "Gene Symbol: SLC2A14, Ensembl Gene ID: ENSG00000173262, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: LINC02130, Ensembl Gene ID: ENSG00000260009, Gene Type: lncRNA\n",
      "No gene found for symbol 'DYNLRB2-AS1  LOC105371357' in Ensembl.\n",
      "Gene Symbol: SCARF1, Ensembl Gene ID: ENSG00000074660, Gene Type: protein_coding\n",
      "Gene Symbol: MINK1, Ensembl Gene ID: ENSG00000141503, Gene Type: protein_coding\n",
      "Gene Symbol: SLC47A1, Ensembl Gene ID: ENSG00000142494, Gene Type: protein_coding\n",
      "No gene found for symbol 'NAPA     KPTN     NAPA-AS1' in Ensembl.\n",
      "Gene Symbol: DSCAM, Ensembl Gene ID: ENSG00000171587, Gene Type: protein_coding\n",
      "No gene found for symbol 'LRRC3    LRRC3-DT' in Ensembl.\n",
      "Gene Symbol: SLC5A4-AS1, Ensembl Gene ID: ENSG00000242082, Gene Type: lncRNA\n",
      "Gene Symbol: LIMK1, Ensembl Gene ID: ENSG00000106683, Gene Type: protein_coding\n",
      "Gene Symbol: MMP2, Ensembl Gene ID: ENSG00000087245, Gene Type: protein_coding\n",
      "Gene Symbol: MMP2, Ensembl Gene ID: ENSG00000087245, Gene Type: protein_coding\n",
      "Gene Symbol: MMP2, Ensembl Gene ID: ENSG00000087245, Gene Type: protein_coding\n",
      "No gene found for symbol 'LTA TNF' in Ensembl.\n",
      "Gene Symbol: MMP2, Ensembl Gene ID: ENSG00000087245, Gene Type: protein_coding\n",
      "Gene Symbol: MMP2, Ensembl Gene ID: ENSG00000087245, Gene Type: protein_coding\n",
      "Gene Symbol: LIMK1, Ensembl Gene ID: ENSG00000106683, Gene Type: protein_coding\n",
      "Gene Symbol: KLK8, Ensembl Gene ID: ENSG00000129455, Gene Type: protein_coding\n",
      "Gene Symbol: RNF213, Ensembl Gene ID: ENSG00000173821, Gene Type: protein_coding\n",
      "Gene Symbol: ELN, Ensembl Gene ID: ENSG00000049540, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: CDKN2B-AS1, Ensembl Gene ID: ENSG00000240498, Gene Type: lncRNA\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: CDKN2B-AS1, Ensembl Gene ID: ENSG00000240498, Gene Type: lncRNA\n",
      "Gene Symbol: COL1A2, Ensembl Gene ID: ENSG00000164692, Gene Type: protein_coding\n",
      "No gene found for symbol 'LTA          TNF          LOC100287329' in Ensembl.\n",
      "No gene found for symbol 'MMP9        SLC12A5-AS1' in Ensembl.\n",
      "Gene Symbol: MMP9, Ensembl Gene ID: ENSG00000100985, Gene Type: protein_coding\n",
      "Gene Symbol: COL1A2, Ensembl Gene ID: ENSG00000164692, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: C14orf39, Ensembl Gene ID: ENSG00000179008, Gene Type: protein_coding\n",
      "No gene found for symbol 'SZRD1   SPATA21' in Ensembl.\n",
      "Gene Symbol: C14orf39, Ensembl Gene ID: ENSG00000179008, Gene Type: protein_coding\n",
      "Gene Symbol: ENG, Ensembl Gene ID: ENSG00000106991, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: AGMO, Ensembl Gene ID: ENSG00000187546, Gene Type: protein_coding\n",
      "Gene Symbol: AGMO, Ensembl Gene ID: ENSG00000187546, Gene Type: protein_coding\n",
      "Gene Symbol: ARHGEF11, Ensembl Gene ID: ENSG00000132694, Gene Type: protein_coding\n",
      "Gene Symbol: IQSEC1, Ensembl Gene ID: ENSG00000144711, Gene Type: protein_coding\n",
      "Gene Symbol: JDP2, Ensembl Gene ID: ENSG00000140044, Gene Type: protein_coding\n",
      "Gene Symbol: JDP2, Ensembl Gene ID: ENSG00000140044, Gene Type: protein_coding\n",
      "Gene Symbol: JDP2, Ensembl Gene ID: ENSG00000140044, Gene Type: protein_coding\n",
      "Gene Symbol: JDP2, Ensembl Gene ID: ENSG00000140044, Gene Type: protein_coding\n",
      "Gene Symbol: JDP2, Ensembl Gene ID: ENSG00000140044, Gene Type: protein_coding\n",
      "Gene Symbol: COL1A2, Ensembl Gene ID: ENSG00000164692, Gene Type: protein_coding\n",
      "Gene Symbol: COL1A2, Ensembl Gene ID: ENSG00000164692, Gene Type: protein_coding\n",
      "Gene Symbol: ELN, Ensembl Gene ID: ENSG00000049540, Gene Type: protein_coding\n",
      "Gene Symbol: ELN, Ensembl Gene ID: ENSG00000049540, Gene Type: protein_coding\n",
      "No gene found for symbol 'IL6     IL6-AS1' in Ensembl.\n",
      "Gene Symbol: RNF213, Ensembl Gene ID: ENSG00000173821, Gene Type: protein_coding\n",
      "Gene Symbol: RNF213, Ensembl Gene ID: ENSG00000173821, Gene Type: protein_coding\n",
      "Gene Symbol: RNF213, Ensembl Gene ID: ENSG00000173821, Gene Type: protein_coding\n",
      "Gene Symbol: RNF213, Ensembl Gene ID: ENSG00000173821, Gene Type: protein_coding\n",
      "Gene Symbol: RNF213, Ensembl Gene ID: ENSG00000173821, Gene Type: protein_coding\n",
      "Gene Symbol: RNF213, Ensembl Gene ID: ENSG00000173821, Gene Type: protein_coding\n",
      "Gene Symbol: RNF213, Ensembl Gene ID: ENSG00000173821, Gene Type: protein_coding\n",
      "Gene Symbol: RNF213, Ensembl Gene ID: ENSG00000173821, Gene Type: protein_coding\n",
      "Gene Symbol: RNF213, Ensembl Gene ID: ENSG00000173821, Gene Type: protein_coding\n",
      "No gene found for symbol 'RNF213     RNF213-AS1' in Ensembl.\n",
      "No gene found for symbol 'RNF213     RNF213-AS1' in Ensembl.\n",
      "No gene found for symbol 'RNF213     RNF213-AS1' in Ensembl.\n",
      "No gene found for symbol 'RNF213     RNF213-AS1' in Ensembl.\n",
      "No gene found for symbol 'None' in Ensembl.\n",
      "Gene Symbol: ADAMTS15, Ensembl Gene ID: ENSG00000166106, Gene Type: protein_coding\n",
      "No gene found for symbol 'SHFL    ANGPTL6' in Ensembl.\n",
      "Gene Symbol: ANGPTL6, Ensembl Gene ID: ENSG00000130812, Gene Type: protein_coding\n",
      "Gene Symbol: PCNT, Ensembl Gene ID: ENSG00000160299, Gene Type: protein_coding\n",
      "Gene Symbol: PCNT, Ensembl Gene ID: ENSG00000160299, Gene Type: protein_coding\n",
      "Gene Symbol: PCNT, Ensembl Gene ID: ENSG00000160299, Gene Type: protein_coding\n",
      "Gene Symbol: PCNT, Ensembl Gene ID: ENSG00000160299, Gene Type: protein_coding\n",
      "Gene Symbol: PCNT, Ensembl Gene ID: ENSG00000160299, Gene Type: protein_coding\n",
      "Gene Symbol: PCNT, Ensembl Gene ID: ENSG00000160299, Gene Type: protein_coding\n",
      "Gene Symbol: PCNT, Ensembl Gene ID: ENSG00000160299, Gene Type: protein_coding\n",
      "Gene Symbol: PCNT, Ensembl Gene ID: ENSG00000160299, Gene Type: protein_coding\n",
      "Gene Symbol: PCNT, Ensembl Gene ID: ENSG00000160299, Gene Type: protein_coding\n",
      "Gene Symbol: RNF213, Ensembl Gene ID: ENSG00000173821, Gene Type: protein_coding\n",
      "Gene Symbol: RNF213, Ensembl Gene ID: ENSG00000173821, Gene Type: protein_coding\n",
      "Gene Symbol: RNF213, Ensembl Gene ID: ENSG00000173821, Gene Type: protein_coding\n",
      "No gene found for symbol 'RNF213     RNF213-AS1' in Ensembl.\n",
      "Gene Symbol: THSD1, Ensembl Gene ID: ENSG00000136114, Gene Type: protein_coding\n",
      "Gene Symbol: THSD1, Ensembl Gene ID: ENSG00000136114, Gene Type: protein_coding\n",
      "Gene Symbol: THSD1, Ensembl Gene ID: ENSG00000136114, Gene Type: protein_coding\n",
      "Gene Symbol: TMEM132B, Ensembl Gene ID: ENSG00000139364, Gene Type: protein_coding\n",
      "No gene found for symbol 'LOX    SRFBP1' in Ensembl.\n",
      "Gene Symbol: LOXL2, Ensembl Gene ID: ENSG00000134013, Gene Type: protein_coding\n",
      "No gene found for symbol 'None' in Ensembl.\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "\n",
    "# Function to get Gene ID and Gene Type from Ensembl REST API\n",
    "def get_gene_info_via_ensembl(gene_symbol, species=\"homo_sapiens\"):\n",
    "    # Construct the URL for Ensembl's REST API\n",
    "    url = f\"https://rest.ensembl.org/lookup/symbol/{species}/{gene_symbol}\"\n",
    "    \n",
    "    # Set the headers to request JSON format\n",
    "    headers = {\"Content-Type\": \"application/json\"}\n",
    "    \n",
    "    try:\n",
    "        # Send the GET request to Ensembl's API\n",
    "        response = requests.get(url, headers=headers)\n",
    "        \n",
    "        # Check if the response is successful\n",
    "        if response.status_code == 200:\n",
    "            data = response.json()\n",
    "            \n",
    "            # Extract Gene ID and Gene Type\n",
    "            gene_id = data.get(\"id\", \"Gene ID not available\")\n",
    "            gene_type = data.get(\"biotype\", \"Gene type not available\")\n",
    "            \n",
    "            # Check if Gene Type is None and replace it\n",
    "            if gene_type is None:\n",
    "                gene_type = \"No gene reported\"\n",
    "            \n",
    "            print(f\"Gene Symbol: {gene_symbol}, Ensembl Gene ID: {gene_id}, Gene Type: {gene_type}\")\n",
    "            return gene_id, gene_type\n",
    "        elif response.status_code == 400:\n",
    "            print(f\"No gene found for symbol '{gene_symbol}' in Ensembl.\")\n",
    "            return \"No gene found\", \"Intergenic (Review)\"\n",
    "        else:\n",
    "            print(f\"Failed to retrieve data for {gene_symbol}. HTTP status: {response.status_code}\")\n",
    "            return None, \"Error retrieving data\"\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred for gene {gene_symbol}: {e}\")\n",
    "        return None, \"Error retrieving data\"\n",
    "\n",
    "# Assume filtered_final_df is your existing DataFrame with the gene symbols in a column named 'Gene Symbol'\n",
    "# Replace 'Gene Symbol' with the name of your actual column containing the gene symbols\n",
    "filtered_final_df[['Ensembl Gene ID', 'Gene Type']] = filtered_final_df['Gene ID (GENE_ID)'].apply(\n",
    "    lambda gene: pd.Series(get_gene_info_via_ensembl(gene))\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "34268edb-3400-4069-887d-34e3b2dc92ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "####Now let's output this table.\n",
    "filtered_final_df.to_csv(os.path.join(output_path, 'RENTREZ_NCBI_Compiled.csv'), index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "27fdd931-a77e-4d29-923f-a4e5e6f2a07d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID (GENE_ID)</th>\n",
       "      <th>Reported SNPs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RNF213</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>VCAN</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDKN2B-AS1</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PCNT</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MMP2</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MMP24           MMP24-AS1-EDEM2</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>JDP2</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MMP26</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>LIMK1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>RNF213     RNF213-AS1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>COL1A2</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ELN</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>VCAN     VCAN-AS1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>THSD1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>MMP16</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ADAMTS13</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>KLK8</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SLC47A1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MMP13</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Gene ID (GENE_ID)  Reported SNPs\n",
       "0                            RNF213             13\n",
       "1                              VCAN             11\n",
       "2                        CDKN2B-AS1             11\n",
       "3                              PCNT              9\n",
       "4                              MMP2              9\n",
       "5   MMP24           MMP24-AS1-EDEM2              6\n",
       "6                              JDP2              6\n",
       "7                             MMP26              6\n",
       "8                             LIMK1              6\n",
       "9             RNF213     RNF213-AS1              5\n",
       "10                           COL1A2              5\n",
       "11                              ELN              4\n",
       "12                VCAN     VCAN-AS1              4\n",
       "13                            THSD1              3\n",
       "14                              ENG              3\n",
       "15                            MMP16              3\n",
       "16                         ADAMTS13              3\n",
       "17                             KLK8              3\n",
       "18                          SLC47A1              2\n",
       "19                            MMP13              2"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Again unique genes\n",
    "Unique_Genes = filtered_final_df['Gene ID (GENE_ID)'].value_counts().reset_index()\n",
    "Unique_Genes.columns = ['Gene ID (GENE_ID)', 'Reported SNPs']\n",
    "Unique_Genes.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "56fe4f9f-7b10-400d-8c3c-d48dd5a2fa9a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID (GENE_ID)</th>\n",
       "      <th>dbSNP ID (RS)</th>\n",
       "      <th>SNP Class (SCLS)</th>\n",
       "      <th>Functional consequence class (FXN)</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Chromosome base position (CPOS)</th>\n",
       "      <th>Minor Allele Frequency from global population (GMAF)</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Associated clinical phenotypes (CLIN)</th>\n",
       "      <th>Nucleotide Accessions (ACCN)</th>\n",
       "      <th>Ensembl Gene ID</th>\n",
       "      <th>Gene Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BOLL</td>\n",
       "      <td>700651</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>2</td>\n",
       "      <td>2:197766990</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>A/G</td>\n",
       "      <td></td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>ENSG00000152430</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EDNRA</td>\n",
       "      <td>6841581</td>\n",
       "      <td>snv</td>\n",
       "      <td>2KB_upstream_variant,upstream_transcript_variant</td>\n",
       "      <td>4</td>\n",
       "      <td>4:147480038</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...</td>\n",
       "      <td>A/T/C/G</td>\n",
       "      <td></td>\n",
       "      <td>NC_000004.12</td>\n",
       "      <td>ENSG00000151617</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>C5orf60</td>\n",
       "      <td>62405726</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>5</td>\n",
       "      <td>5:179642467</td>\n",
       "      <td>[1] \"1000Genomes\" \"ExAC\"        \"GnomAD\"     ...</td>\n",
       "      <td>A/T/C/G</td>\n",
       "      <td></td>\n",
       "      <td>NC_000005.10</td>\n",
       "      <td>ENSG00000204661</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MALL</td>\n",
       "      <td>117802391</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>2</td>\n",
       "      <td>2:110104507</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"GnomAD...</td>\n",
       "      <td>C/T</td>\n",
       "      <td></td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>ENSG00000144063</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>RREB1</td>\n",
       "      <td>9505086</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>6</td>\n",
       "      <td>6:7231953</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>C/T</td>\n",
       "      <td></td>\n",
       "      <td>NC_000006.12</td>\n",
       "      <td>ENSG00000124782</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>DST</td>\n",
       "      <td>117021265</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant,genic_upstream_transcript_variant</td>\n",
       "      <td>6</td>\n",
       "      <td>6:56763223</td>\n",
       "      <td>[1] \"1000Genomes\" \"GnomAD\"      \"KOREAN\"      ...</td>\n",
       "      <td>C/T</td>\n",
       "      <td></td>\n",
       "      <td>NC_000006.12</td>\n",
       "      <td>ENSG00000151914</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>FOXP1</td>\n",
       "      <td>878118</td>\n",
       "      <td>snv</td>\n",
       "      <td>genic_upstream_transcript_variant,genic_downst...</td>\n",
       "      <td>3</td>\n",
       "      <td>3:71197077</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...</td>\n",
       "      <td>A/G/T</td>\n",
       "      <td></td>\n",
       "      <td>NC_000003.12</td>\n",
       "      <td>ENSG00000114861</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>RBMS3</td>\n",
       "      <td>1979271</td>\n",
       "      <td>snv</td>\n",
       "      <td>genic_upstream_transcript_variant,intron_variant</td>\n",
       "      <td>3</td>\n",
       "      <td>3:29565914</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>A/T/C/G</td>\n",
       "      <td></td>\n",
       "      <td>NC_000003.12</td>\n",
       "      <td>ENSG00000144642</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>RUFY1</td>\n",
       "      <td>4075890</td>\n",
       "      <td>snv</td>\n",
       "      <td>intron_variant,genic_upstream_transcript_variant</td>\n",
       "      <td>5</td>\n",
       "      <td>5:179570372</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>C/G/T</td>\n",
       "      <td></td>\n",
       "      <td>NC_000005.10</td>\n",
       "      <td>ENSG00000176783</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>PLEKHA1</td>\n",
       "      <td>10510110</td>\n",
       "      <td>snv</td>\n",
       "      <td>non_coding_transcript_variant,genic_downstream...</td>\n",
       "      <td>10</td>\n",
       "      <td>10:122432914</td>\n",
       "      <td>[1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...</td>\n",
       "      <td>A/T/C/G</td>\n",
       "      <td></td>\n",
       "      <td>NC_000010.11</td>\n",
       "      <td>ENSG00000107679</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Gene ID (GENE_ID) dbSNP ID (RS) SNP Class (SCLS)  \\\n",
       "0               BOLL        700651              snv   \n",
       "1              EDNRA       6841581              snv   \n",
       "5            C5orf60      62405726              snv   \n",
       "9               MALL     117802391              snv   \n",
       "11             RREB1       9505086              snv   \n",
       "12               DST     117021265              snv   \n",
       "14             FOXP1        878118              snv   \n",
       "16             RBMS3       1979271              snv   \n",
       "19             RUFY1       4075890              snv   \n",
       "21           PLEKHA1      10510110              snv   \n",
       "\n",
       "                   Functional consequence class (FXN) Chromosome  \\\n",
       "0                                      intron_variant          2   \n",
       "1    2KB_upstream_variant,upstream_transcript_variant          4   \n",
       "5                                      intron_variant          5   \n",
       "9                                      intron_variant          2   \n",
       "11                                     intron_variant          6   \n",
       "12   intron_variant,genic_upstream_transcript_variant          6   \n",
       "14  genic_upstream_transcript_variant,genic_downst...          3   \n",
       "16   genic_upstream_transcript_variant,intron_variant          3   \n",
       "19   intron_variant,genic_upstream_transcript_variant          5   \n",
       "21  non_coding_transcript_variant,genic_downstream...         10   \n",
       "\n",
       "   Chromosome base position (CPOS)  \\\n",
       "0                      2:197766990   \n",
       "1                      4:147480038   \n",
       "5                      5:179642467   \n",
       "9                      2:110104507   \n",
       "11                       6:7231953   \n",
       "12                      6:56763223   \n",
       "14                      3:71197077   \n",
       "16                      3:29565914   \n",
       "19                     5:179570372   \n",
       "21                    10:122432914   \n",
       "\n",
       "   Minor Allele Frequency from global population (GMAF)   Allele  \\\n",
       "0    [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...        A/G   \n",
       "1    [1] \"1000Genomes\"    \"ALSPAC\"         \"Chilea...    A/T/C/G   \n",
       "5    [1] \"1000Genomes\" \"ExAC\"        \"GnomAD\"     ...    A/T/C/G   \n",
       "9    [1] \"1000Genomes\"    \"ALSPAC\"         \"GnomAD...        C/T   \n",
       "11   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...        C/T   \n",
       "12  [1] \"1000Genomes\" \"GnomAD\"      \"KOREAN\"      ...        C/T   \n",
       "14   [1] \"1000Genomes\"    \"ALSPAC\"         \"Daghes...      A/G/T   \n",
       "16   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...    A/T/C/G   \n",
       "19   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...      C/G/T   \n",
       "21   [1] \"1000Genomes\"    \"ALSPAC\"         \"Estoni...    A/T/C/G   \n",
       "\n",
       "   Associated clinical phenotypes (CLIN) Nucleotide Accessions (ACCN)  \\\n",
       "0                                                        NC_000002.12   \n",
       "1                                                        NC_000004.12   \n",
       "5                                                        NC_000005.10   \n",
       "9                                                        NC_000002.12   \n",
       "11                                                       NC_000006.12   \n",
       "12                                                       NC_000006.12   \n",
       "14                                                       NC_000003.12   \n",
       "16                                                       NC_000003.12   \n",
       "19                                                       NC_000005.10   \n",
       "21                                                       NC_000010.11   \n",
       "\n",
       "    Ensembl Gene ID       Gene Type  \n",
       "0   ENSG00000152430  protein_coding  \n",
       "1   ENSG00000151617  protein_coding  \n",
       "5   ENSG00000204661  protein_coding  \n",
       "9   ENSG00000144063  protein_coding  \n",
       "11  ENSG00000124782  protein_coding  \n",
       "12  ENSG00000151914  protein_coding  \n",
       "14  ENSG00000114861  protein_coding  \n",
       "16  ENSG00000144642  protein_coding  \n",
       "19  ENSG00000176783  protein_coding  \n",
       "21  ENSG00000107679  protein_coding  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Assuming filtered_final_df is your DataFrame\n",
    "# Filter rows where Gene Type is 'protein-coding'\n",
    "protein_coding_df = filtered_final_df[filtered_final_df['Gene Type'] == 'protein_coding']\n",
    "\n",
    "# Display the filtered DataFrame\n",
    "protein_coding_df.head(10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ece973ac-c826-40f2-bc67-4c6425468950",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Protein-coding genes have been saved to 'Resources/OutputFromExtractionCode'\n",
      "['BOLL', 'EDNRA', 'C5orf60', 'MALL', 'RREB1']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Assuming filtered_final_df is your DataFrame with gene symbols and gene types\n",
    "\n",
    "# Filter rows with 'protein-coding' in the Gene Type column\n",
    "protein_coding_df = filtered_final_df[filtered_final_df['Gene Type'] == 'protein_coding']\n",
    "\n",
    "# Extract the 'Gene ID (GENE_ID)' column as a list\n",
    "protein_coding_genes = protein_coding_df['Gene ID (GENE_ID)'].tolist()\n",
    "\n",
    "# Define the output path for the text file\n",
    "output_path_txt = f'{output_path}/protein_coding_genes.txt'  # You can set this to any desired path\n",
    "\n",
    "# Save the list to a text file with each gene on a new line\n",
    "with open(output_path_txt, \"w\") as file:\n",
    "    for gene in protein_coding_genes:\n",
    "        file.write(f\"{gene}\\n\")\n",
    "\n",
    "print(f\"Protein-coding genes have been saved to '{output_path}'\")\n",
    "\n",
    "# Display the first few items in the list for verification\n",
    "print(protein_coding_genes[:5])  # Shows the first 5 genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a96b6db4-e627-4137-9f16-6c570948ea9d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd6a42f0-ec8f-45cd-bfe8-8a23f2aeb042",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cd44ea7-fadf-46ae-8af5-6c0ab20727b6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
